---
title: "XistPaper_Figure06"
output: html_document
author: "Jasper Panten"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/../../"))
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

This script reproduces all plots for Figure 6 and associated supplementary figures in PaperName.
We first load the necessary libraries and some helper functions: 

```{r load_libraries, message=F}

setwd(paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/../../"))

library(tidyverse)
library(ggplot2)
library(scran)
library(scater)
library(DESeq2)
library(patchwork)

source("./Scripts/General/auxiliary.R")
source("./Scripts/General/ase_functions.R")

escape_colors <- setNames(c("lightblue", "darkgrey", "darkgreen", "orange"), nm = c("Embryo-specific", "NPC-specific", "Facultative", "Constitutive"))

```

Now, read in the SCE, remove non-X genes and exclude lowly expressed genes

```{r load_data}

### 
data <- readRDS("./ProcessedData/merged_dataset_invivo.rds")
data$Experiment <- ifelse(unlist(lapply(str_split(data$Sample, "_"), function(x){x[[1]]})) == "AAFCTG2M5", "Experiment4", data$Experiment)

exp2_late <- colnames(data)
exp2_late <- gsub("AACMTNTM5_EmbryoJF1Rtta2_23s001680-1-1_Kneuss_lane1s", "", exp2_late) %in% 
  c("12B", "1A", "1B", "2B", "3A", "4B", "5A", "5B", "6A", "6B", "7A", "7B", "9B") & 
  grepl("AACMTNTM5_EmbryoJF1Rtta2_23s001680-1-1_Kneuss_lane1s", exp2_late)

data$EmbryoStage <- case_when(
  data$Experiment == "Experiment1" & colnames(data) != "Sample1"  ~ "Morula",
  data$Experiment == "Experiment1" & colnames(data) == "Sample1"  ~ "E3.5",
  data$Experiment == "Experiment2" & !exp2_late ~ "E3.5",
  data$Experiment == "Experiment2" & exp2_late ~ "E4.5",
  data$Experiment == "Experiment3" ~ "E4.5", 
  data$Experiment == "Experiment4" ~ "E4.5"
)

data_before_filtering <-  data[seqnames(data) == "X", ]
data_before_filtering <- data_before_filtering[rownames(data_before_filtering) != "Xist", ]

data_full <- data

data <- data[rowSums(counts(data)) / ncol(data) > 10, ]

```

Check that embryos are correctly sexed by comparing X / Y linked expression, and confirm allelic mapping:  

```{r read_qc}

data_with_autosomes <- data_full
data <- data[seqnames(data) == "X", ]
data_sexing <- data_with_autosomes[seqnames(data_with_autosomes) %in% c("X", "Y")]

# look at X/Y coverage
data.frame(
  Experiment = data_with_autosomes$Experiment, 
  Ycov = colSums(counts(data_with_autosomes[seqnames(rowRanges(data_with_autosomes)) == "Y", ])) / colSums(counts(data_with_autosomes)), 
  Xcov = colSums(counts_inactive(data_with_autosomes[seqnames(rowRanges(data_with_autosomes)) == "X", ])) / colSums(counts_inactive(data_with_autosomes))
) %>%
  ggplot(aes(Xcov, Ycov)) + geom_point(aes(col = Experiment)) + 
    theme_bw() + xlab("Fraction of X-linked reads (Xi only)") + ylab("Fraction of Y-linked reads")
ggplot2::ggsave("./Plots/Fig6/Exp2_X_Y_sexing_plot.pdf")

sample_show = 24
sample_show = 10
colnames(data_before_filtering)[[sample_show]]

data.frame(
  counts_total = counts(data_before_filtering[,sample_show])[[1]], 
  counts_allelic = (counts_active(data_before_filtering[,sample_show]) +  counts_inactive(data_before_filtering[,sample_show]))[[1]]
) %>%
  ggplot(aes(x = counts_total + 1, y = counts_allelic + 1)) + geom_point() + scale_x_log10() + scale_y_log10() + 
  theme_paper() + 
  geom_vline(xintercept = 10, linetype = "dashed") + xlab("Total reads mapped") + ylab("Allele-specific reads mapped") + 
  geom_abline(linetype = "dashed")
ggplot2::ggsave("./Plots/Fig6/Exp2_mapped_reads_sample1.pdf", height = 7, width = 7)

sample_show = 10
colnames(data_before_filtering)[[sample_show]]

# data.frame(
#   chromosome = as.character(seqnames(rowRanges(data_with_autosomes))) == "X", 
#   total = (counts_active(data_with_autosomes[,sample_show]) + counts_inactive(data_with_autosomes[,sample_show]))[[1]], 
#   ase =  (counts_active(data_with_autosomes[,sample_show]) / (counts_active(data_with_autosomes[,sample_show]) + counts_inactive(data_with_autosomes[,sample_show])))[[1]]
# ) %>%
#   ggplot(aes(x = ase, fill = chromosome)) + geom_density() + theme_paper() + 
#   ylab("Density") + xlab("JF1 / (B6 + JF1)") + 
#     scale_fill_manual(values = c("grey", "black"))
# ggplot2::ggsave("./Plots/Fig6/Exp2_density_ase_sample1.pdf", height = 7, width = 7)

data.frame(
  Experiment = data_with_autosomes$Experiment, 
  Total = colSums(counts(data_with_autosomes)), 
  Y = colSums(counts(data_with_autosomes[seqnames(rowRanges(data_with_autosomes)) == "Y", ])), 
  Xact = colSums(counts_active(data_with_autosomes[seqnames(rowRanges(data_with_autosomes)) == "X", ])), 
  Xinact = colSums(counts_inactive(data_with_autosomes[seqnames(rowRanges(data_with_autosomes)) == "X", ])) , 
  Xist_total = as.numeric(counts(data_with_autosomes["Xist", ])), 
  Xist_inact = as.numeric(counts_inactive(data_with_autosomes["Xist", ])), 
  Xist_act = as.numeric(counts_active(data_with_autosomes["Xist", ])), 
  RTTA_reads = data_with_autosomes$RttaCounts
) -> test

write.csv(test, "./ProcessedData/embryo_sexing_overview.csv")
write.csv(cbind(test[,1:2], test[,-c(1:2)] / test$Total * 1e6), "./ProcessedData/embryo_sexing_overview_cpm.csv")

test %>% 
  rownames_to_column("Sample") %>%
  mutate(Sample = as.numeric(factor(Sample))) %>%
  pivot_longer(-c("Experiment", "Sample", "Total")) %>%
  ggplot(aes(x = Sample, y = value / Total, fill = Experiment)) + geom_bar(stat = "identity") + 
    facet_wrap(~name, scales = "free") + coord_flip() + xlab("Fraction of total reads")
ggplot2::ggsave("./Plots/Fig6/cross_embryo_counts.pdf", width = 20, height = 20)

test %>% 
  rownames_to_column("Sample") %>%
  mutate(Sample = as.numeric(factor(Sample))) %>%
  pivot_longer(-c("Experiment", "Sample", "Total", "Xact", "Xinact")) %>%
  ggplot(aes(x = (Xact + Xinact) / Total, y = value / Total, col = Experiment)) + geom_point() + 
    facet_wrap(~name, scales = "free") + xlab("Fraction X reads") + ylab("Fraction reads (faceted variable)")
ggplot2::ggsave("./Plots/Fig6/cross_embryo_counts_scatter.pdf", width = 20, height = 20)

### y counts closer look

all_y_counts <- counts(data_with_autosomes[seqnames(rowRanges(data_with_autosomes)) == "Y", ])
all_y_counts <- all_y_counts[rowSums(all_y_counts) > 10, ]

data.frame(
  Sex = as.numeric(as.numeric(all_y_counts["Ddx3y", ]) > 0), 
  row.names = data$Sample
) -> colData_annotation

pdf("./Plots/Fig6/y_counts.pdf", width = 12, height = 12)
pheatmap::pheatmap(log10(all_y_counts + 1), annotation_col = colData_annotation)
dev.off()

data$Sex <- ifelse(colData_annotation$Sex == 0, "female", "male")
data_with_autosomes$Sex <- ifelse(colData_annotation$Sex == 0, "female", "male")

sort(as.numeric(counts(data_with_autosomes)["Xist",]))

samples_out <- c("AACYJJ7M5_E45doxSept23_23s002684-1-1_loda_lane1B30F1", 
                 "AACYJJ7M5_E45doxSept23_23s002684-1-1_loda_lane1B30F1")

# test %>%
#   rownames_to_column("Sample") %>%
#   ggplot(aes(Total, Xinact, shape = Experiment, col = Xist_total > 50)) + geom_point() + 
#     theme_bw() + xlab("Total reads") + ylab("Fraction of Xi reads") + ggrepel::geom_text_repel(aes(label = Sample), size = 3)
# 
# test %>%
#   ggplot(aes(Y / Total, Xinact / Total, col = Experiment)) + geom_point() + 
#     theme_bw() + xlab("Y fraction") + ylab("Xi fraction")
# 
# test %>%
#   ggplot(aes(Xinact, Xist_total, col = Experiment)) + geom_point() + 
#     theme_bw() + xlab("Inactive X") + ylab("Fraction Xist (total) reads")
# 
# test %>%
#   ggplot(aes(Xist_inact, Xist_act, col = Experiment)) + geom_point() + 
#     theme_bw() + xlab("Xi Xist reads") + ylab("Xa Xist reads") + 
#     xlim(c(0, 4000)) + ylim(c(0, 1000))
# 
# samples_out <- test %>% dplyr::filter((Xinact / Total) < 0.0005 | grepl("lane1A10B4|lane1B30F1", rownames(.)))
# 
# test %>%
#   rownames_to_column("Sample") %>%
#   ggplot(aes(Total, Xinact, shape = Experiment, col = Sample %in% rownames(samples_out))) + geom_point() + 
#     theme_bw() + xlab("Total reads") + ylab("Fraction of Xi reads")
# 
# test %>%
#   rownames_to_column("Sample") %>%
#   ggplot(aes(Total, Xinact, shape = RTTA_reads, col = Sample %in% rownames(samples_out))) + geom_point() + 
#     theme_bw() + xlab("Total reads") + ylab("Fraction of Xi reads")

```

```{r xist_vs_escape}

### 
data <- data[,data$Sex != "male" & data$Experiment != "Experiment123" & !(data$Sample %in% samples_out) ]
data <- data[rowMeans(counts(data)) > 10, ]
data_here <- data
data_with_autosomes <- data_with_autosomes[ , data_with_autosomes$Sex != "male" & data_with_autosomes$Experiment != "Experiment123" & 
                                            !(data_with_autosomes$Sample %in% samples_out)]
ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))

### 
data_here_filtered <- data_here

fastq_path <- "/omics/groups/OE0538/internal/users/panten/projects/AntoniaESCProject/invivo_data/data/"

geo_metadata <- data.frame(
  SampleName = data_here_filtered$Sample, 
  # CloneNameShort = data_here_filtered$Sample, 
  Organism = "mouse", 
  Strain = "JF1 x C57BL6/J (Xptet/Y; R26rtTA/WT)", 
  # CellLine = "neural progenitor cell",
  Genotype = ifelse(data_here$RttaGenotype == "Rtta-", "Xptet", "Xptet / R26rtTA"),
  Treatment = "Doxycycline 24h", 
  # Time = "untreated", 
  Experiment = data_here$Experiment, 
  EmbryoStage = data_here$EmbryoStage,
  Molecule = "total RNA", 
  SeqType = "paired end", 
  InstrumentModel = "NextSeq 2000", 
  fastq1 = paste0(fastq_path, data_here_filtered$Sample, "_1.fq.gz"), 
  fastq2 = paste0(fastq_path, data_here_filtered$Sample, "_2.fq.gz"), 
  fastq1_fileonly = paste0(data_here_filtered$Sample, "_1.fq.gz"), 
  fastq2_fileonly = paste0(data_here_filtered$Sample, "_2.fq.gz"), 
  processed_1 = "embryo_total_counts.txt", 
  processed_2 = "embryo_genome1_counts.txt", 
  processed_3 = "embryo_genome2_counts.txt", 
  processed_4 = "embryo_total_counts_normalized.txt"
)

write_csv(geo_metadata, "./ProcessedData/sample_overview_invivo_used_embryos.csv")

data_upload <- data_with_autosomes

counts(data_upload) %>% 
  rownames_to_column("Gene") %>%
  write_delim("../../invivo_data/invivo_geo_upload_processed/processed_data/invivo_total_counts.txt", delim = "\t")
counts_active(data_upload) %>% 
  rownames_to_column("Gene") %>%
  write_delim("../../invivo_data/invivo_geo_upload_processed/processed_data/invivo_genome1_counts.txt", delim = "\t")
counts_inactive(data_upload) %>% 
  rownames_to_column("Gene") %>%
  write_delim("../../invivo_data/invivo_geo_upload_processed/processed_data/invivo_genome2_counts.txt", delim = "\t")

# data.frame(
#   Sample = data$ShortSampleName, 
#   Expression = as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6), 
#   rtta_genotype = data_here$RttaGenotype
# ) %>%
#   ggplot(aes(x = Sample, y = Expression, fill = rtta_genotype)) + 
#   stat_summary(stat = "mean", geom = "bar", col = "black") + 
#   coord_flip() + theme_paper() + 
#   xlab("") + ylab("Expression (CPM)") + geom_jitter(size = 3, width = 0.1) + 
#   # scale_y_continuous(expand = c(0, 0), limits = c(0, 320000)) + 
#   theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
#   ggtitle("Xist expression")
# ggsave("./Plots/Fig5/exp2_xist_expression.pdf")

# Distribution d-scores
# ratios %>%
#   t() %>% data.frame() %>%
#   add_column(Sample = data$ShortSampleName) %>%
#   pivot_longer(-c("Sample")) %>%
#   ggplot(aes(x = Sample, y = value)) + 
#   geom_boxplot(col = "black", outlier.color = "grey") + 
#   theme_paper() + ylab("ASE (d-score)") + 
#   theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
#   labs(fill = "Escape Category") + xlab("") + coord_flip()
# ggsave("./Plots/Fig5/exp2_d_score_distribution.pdf")

size_factors <- colSums(counts(data_with_autosomes)) / colSums(counts(data_with_autosomes))[[1]]

data.frame(
  Sample = data_here$ShortSampleName, 
  Expression = as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6)
) %>% pull(Expression, name = Sample) -> sample_to_xist

sample_to_rtta <- colData(data_here) %>% data.frame() %>% pull(RttaGenotype, name = ShortSampleName)

sample_to_experiment <- colData(data_here) %>% data.frame() %>% pull(Experiment, name = ShortSampleName)

# ratios %>%
#   t() %>% data.frame() %>%
#   add_column(Sample = data_here$ShortSampleName) %>%
#   pivot_longer(-c("Sample")) %>%
#   group_by(Sample) %>% 
#   summarize(value = mean(value, na.rm = T)) %>%
#   mutate(XistExpression = sample_to_xist[as.character(Sample)]) %>%
#   mutate(Rtta_Genotype = sample_to_rtta[as.character(Sample)]) %>%
#   mutate(Experiment = sample_to_experiment[as.character(Sample)]) %>%
#   ggplot(aes(x = XistExpression, y = value, col = Rtta_Genotype)) + 
#     geom_point(size = 3) + scale_x_log10() + theme_bw(base_size = 20) + 
#     xlab("Xist expression (CPM)") + ylab("Mean escape") + 
#     geom_smooth(method = "lm") + scale_color_manual(values = c("grey", "darkblue")) + 
#     facet_wrap(~Experiment)
# ggplot2::ggsave("./Plots/Fig6/exp2_xist_vs_escape.pdf")

ratios %>%
  t() %>% data.frame() %>%
  add_column(Sample = data_here$ShortSampleName) %>%
  pivot_longer(-c("Sample")) %>%
  group_by(Sample) %>% 
  summarize(value = mean(value, na.rm = T)) %>% 
  pull(value, name = Sample) -> average_escape

```

```{r global_analysis}

### total expression analysis
library(DESeq2)

deseq_dataset <- DESeqDataSetFromMatrix(countData = counts(data_with_autosomes[seqnames(data_with_autosomes) != "X", ]), 
                                        colData = colData(data_with_autosomes[seqnames(data_with_autosomes) != "X", ]), design = ~ 1)
deseq_dataset <- estimateSizeFactors(deseq_dataset)
deseq_dataset <- logNormCounts(deseq_dataset)
deseq_dataset_hvgs <- getTopHVGs(deseq_dataset)

pca_total <- prcomp(t(assays(deseq_dataset[deseq_dataset_hvgs[1:2000], ])[["logcounts"]]))

data.frame(
  PC1 = pca_total$x[,1], 
  PC2 = pca_total$x[,2], 
  average_escape = average_escape[as.character(deseq_dataset$ShortSampleName)],
  Experiment = sample_to_experiment[as.character(deseq_dataset$ShortSampleName)]
) -> test_here

# test_here %>% 
#   ggplot(aes(x = PC1, y = PC2, fill = average_escape)) + 
#   geom_point(size = 10, pch = 21) + 
#   scale_fill_gradient(low = "white", high = "red") + theme_bw() + 
#   labs(fill = "Average Escape") + 
#   facet_wrap(~Experiment)
# ggplot2::ggsave("./Plots/Fig6/exp2_pca_average_escape.pdf")

# data.frame(
#   PC1 = pca_total$x[,1], 
#   PC2 = pca_total$x[,2], 
#   xist_expression = log(as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6) + 1), 
#   average_escape = average_escape[deseq_dataset$ShortSampleName]
# ) %>% 
#   ggplot(aes(x = PC1, y = PC2, fill = xist_expression)) + 
#   geom_point(size = 10, pch = 21) + viridis::scale_fill_viridis() + 
#   theme_bw() + 
#   labs(fill = "Xist (CPM)")
# ggplot2::ggsave("./Plots/Fig6/exp2_pca_xist.pdf")

data.frame(
  PC1 = pca_total$x[,1], 
  PC2 = pca_total$x[,2], 
  Experiment = deseq_dataset$Experiment, 
  xist_expression = log(as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6) + 1), 
  average_escape = average_escape[deseq_dataset$ShortSampleName], 
  EmbryoStage = deseq_dataset$EmbryoStage
) %>% 
  mutate(EmbryoStageComputational = case_when(
    PC1 > -55 ~ "Late Blastocyst", 
    PC1 < -105 ~ "Morula", 
    .default = "Early Blastocyst"
  )) -> testytest

testytest %>%
  ggplot(aes(x = PC1, y = PC2, fill = Experiment)) + 
    geom_point(size = 10, pch = 21) + 
    theme_paper(textsize = 30) + 
    labs(fill = "Experiment") + 
    #scale_fill_manual(values = c("beige", "brown", "grey")) +
    xlab(paste0("PC1 (", round(100 * (pca_total$sdev / sum(pca_total$sdev))[[1]], digits = 2), "%)")) + 
    ylab(paste0("PC1 (", round(100 * (pca_total$sdev / sum(pca_total$sdev))[[2]], digits = 2), "%)"))
ggplot2::ggsave("./Plots/Fig6/exp2_pca_experiment.pdf", height = 7, width = 12)

testytest %>%
  ggplot(aes(x = PC1, y = PC2, fill = EmbryoStage)) + 
    geom_point(size = 10, pch = 21) + 
    theme_paper(textsize = 30) + 
    labs(fill = "Experiment") + scale_fill_manual(values = c("beige", "brown", "grey")) +
    xlab(paste0("PC1 (", round(100 * (pca_total$sdev / sum(pca_total$sdev))[[1]], digits = 2), "%)")) + 
    ylab(paste0("PC1 (", round(100 * (pca_total$sdev / sum(pca_total$sdev))[[2]], digits = 2), "%)"))
ggplot2::ggsave("./Plots/Fig6/exp2_pca_stage_computational.pdf", height = 7, width = 12)

rotations_fulldataset <- data.frame(pca_total$rotation)
head(rotations_fulldataset[order(rotations_fulldataset$PC1, decreasing = F), 1:5], n = 30)
head(rotations_fulldataset[order(rotations_fulldataset$PC2, decreasing = T), 1:5], n = 30)

# data.frame(
#   PC1 = pca_total$x[,1], 
#   PC2 = pca_total$x[,2], 
#   xist_expression = log(as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6) + 1), 
#   average_escape = average_escape[as.character(deseq_dataset$ShortSampleName)], 
#   rtta_genotype = sample_to_rtta[as.character(deseq_dataset$ShortSampleName)]
# ) %>% 
#   ggplot(aes(x = PC1, y = PC2, fill = rtta_genotype)) + 
#   geom_point(size = 10, pch = 21) + 
#   theme_paper() + 
#   labs(fill = "Genotype")
# ggsave("./Plots/Fig6/exp2_pca_genotype.pdf")

data.frame(
  PC1 = pca_total$x[,1], 
  PC2 = pca_total$x[,2], 
  Stage = testytest$EmbryoStage,
  xist_expression = log(as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6) + 1), 
  average_escape = average_escape[as.character(deseq_dataset$ShortSampleName)], 
  rtta_genotype = sample_to_rtta[as.character(deseq_dataset$ShortSampleName)]
) -> plot_data

plot_data %>% 
  ggplot(aes(x = PC1, y = average_escape, fill = rtta_genotype)) + geom_point(pch = 21, size = 12) + 
  scale_fill_manual(values = c("lightgrey", "darkblue")) + theme_paper(textsize = 30) + 
  xlab("PC1 (Embryo Stage)") + ylab("Average Allelic Ratio (Xi / (Xi + Xa))") + labs(fill = "Genotype") + 
  ylim(c(0, 0.5))
ggplot2::ggsave("./Plots/Fig6/exp2_scatter_genotype_rtta.pdf", height = 6, width = 9)

plot_data %>% 
  ggplot(aes(x = PC1, y = average_escape, fill = Stage)) + geom_point(pch = 21, size = 12) + 
  scale_fill_manual(values = c("beige", "brown", "grey")) + theme_paper(textsize = 30) + 
  xlab("PC1 (Embryo Stage)") + ylab("Average Allelic Ratio (Xi / (Xi + Xa))") + labs(fill = "Stage") + 
  ylim(c(0, 0.5))
ggplot2::ggsave("./Plots/Fig6/exp2_scatter_stage.pdf", height = 6, width = 9)

plot_data %>% 
  ggplot(aes(x = PC1, y = average_escape, fill = xist_expression)) + geom_point(pch = 21, size = 12) + 
  scale_fill_viridis_b() + theme_paper(textsize = 30) + 
  xlab("PC1 (Embryo Stage)") + ylab("Average Allelic Ratio (Xi / (Xi + Xa))") + labs(fill = "Xist") + 
  ylim(c(0, 0.5))
ggplot2::ggsave("./Plots/Fig6/exp2_scatter_xist_differentition.pdf", height = 6, width = 9)

plot_data %>% rownames_to_column("sample") %>% pull(PC1, name = sample) -> sample_to_pc1
data_here$embryo_stage <- case_when(
  sample_to_pc1[as.character(data_here$Sample)] > -55 ~ "Late",
  sample_to_pc1[as.character(data_here$Sample)] < -50 & sample_to_pc1[as.character(data_here$Sample)] > -105  ~ "Early",
  .default = "Morula"
)

# plot_data %>% 
#   ggplot(aes(x = PC1, y = xist_expression, fill = rtta_genotype, col = rtta_genotype)) + geom_point(pch = 21, size = 10) + 
#   theme_paper(textsize = 30) + 
#   xlab("PC1 (Embryo Stage)") + ylab("Average Allelic Ratio (Xi / (Xi + Xa))") + labs(fill = "Xist") + 
#   geom_smooth(method = "lm")
 
# plot_data %>% 
#   rownames_to_column("Sample") %>%
#   mutate(DevStage = data_here[,.$Sample]$embryo_stage) %>%
#   mutate(Experiment = data_here[,.$Sample]$Experiment) %>%
#   ggplot(aes(x = PC1, y = average_escape, fill = Experiment)) + geom_point(pch = 21, size = 7) + theme_bw() + 
#   xlab("PC1 (Embryo Stage)") + ylab("Average Allelic Ratio (Xi / (Xi + Xa))") + labs(fill = "Genotype")
# 
# plot_data %>% 
#   rownames_to_column("Sample") %>%
#   mutate(DevStage = data_here[,.$Sample]$embryo_stage) %>%
#   mutate(Experiment = data_here[,.$Sample]$Experiment) %>%
#   ggplot(aes(x = PC1, y = average_escape, fill = DevStage)) + geom_point(pch = 21, size = 7) + theme_bw() + 
#   xlab("PC1 (Embryo Stage)") + ylab("Average Allelic Ratio (Xi / (Xi + Xa))") + labs(fill = "Genotype")
# ggplot2::ggsave("./Plots/Fig6/exp2_scatter_genotype_differentition_devstage.pdf")

# data_here$embryo_pc <- sample_to_pc1[data_here$Sample]

##### remove all unnecessary samples

data_here <- data_here[,data_here$embryo_stage != "Morula"]
# data_here$RttaGenotype <- ifelse(data_here$RttaGenotype == "?", "Rtta-", data_here$RttaGenotype)

library(patchwork)

data_here <- data_here[seqnames(rowRanges(data_here)) == "X", ]
data_here <- data_here[rowSums(counts(data_here)) / ncol(data_here) > 50, ]

ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))

# For Xist-expression, we need a scaling factor per sample to account for differences in sequencing depth
size_factors <- colSums(counts(data)) / colSums(counts(data))[[1]]

# check that Xist-overexpression works by plotting Xist counts per million (CPM) per sample
data.frame(
  Condition = data_here$RttaGenotype,
  Stage = data_here$EmbryoStage, 
  Expression = as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6)
) %>%
  mutate(Condition = factor(ifelse(Condition == "Rtta-", "Rtta (negative)", "Rtta (positive)"), 
                            levels = c("Rtta (positive)", "Rtta (negative)"))) %>%
  ggplot(aes(x = Condition, y = Expression)) + 
    stat_summary(stat = "mean", geom = "bar", position = "dodge", fill = "grey") + 
    coord_flip() + xlab("") + ylab("Xist expression (CPM)") + 
    geom_jitter(size = 5, width = .1) + 
    scale_y_continuous(expand = c(0, 0), limits = c(0, 50000)) + 
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
    theme_paper(textsize = 30) + facet_wrap(~Stage) + stat_compare_means(label.y = 40000, method = "wilcox.test")
ggplot2::ggsave("./Plots/Fig6/xist_barplot.pdf", width = 12, height = 8)

# For the next heatmap, we want to highlight a few genes
genes_show <- c("Xist", "Mecp2", "Kdm6a", "Kdm5c")
genes_of_choice <- setNames(rep("", nrow(ratios)), rownames(ratios))
genes_of_choice[genes_show] <- genes_show

# Heatmap showing d-scores of all genes across Xist-OE
# ratios %>%
#   t() %>% data.frame(check.names = F) %>%
#   add_column("Condition" = data_here$ConditionClean) %>%
#   add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
#   pivot_longer(-c(Condition, Clone)) %>%
#   add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
#   mutate(Condition = factor(Condition, levels = rev(c("Control", "Dox (3d)", "Dox (7d)", "Dox (14d)", "Dox (21d)")))) %>%
#   ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
#     scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
#     scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
#     scale_y_discrete(labels = genes_of_choice, position = "right") + 
#     theme_paper() + 
#     theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
#     xlab("") + 
#     ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
#     scale_x_discrete(labels = genes_of_choice, position = "bottom") + 
#     labs(fill = "d-score") + guides(fill = "none")
# ggsave("./Plots/Fig1/heatmap_ordered.pdf", width = 8, height = 8)

# Get escapees that are consistent across E6 samples
escapees <- apply(ratios, 2, function(x){x > 0.1})
escapees_per_clone <- apply(escapees, 2, function(x){rownames(escapees)[x]})
# e6_escapees <- Reduce("intersect", escapees_per_clone[3:5])

# define escapee sets: 
# E3.5: 
escapees_control_early <- apply(ratios[ , data_here$EmbryoStage == "E3.5" & data_here$RttaGenotype == "Rtta-"], 2, function(x){x > 0.1})
escapees_control_early <- apply(escapees_control_early, 2, function(x){rownames(escapees)[x]})
escapees_control_early <- names(table(do.call("c", escapees_control_early)))[table(do.call("c", escapees_control_early)) > 0]
escapees_control_early <- escapees_control_early[!escapees_control_early %in% names(rowMeans(ratios, na.rm = T))[(rowMeans(ratios, na.rm = T)) > 0.9]]
length(escapees_control_early)

# E4.5
escapees_control_late <- apply(ratios[ , data_here$EmbryoStage == "E4.5" & data_here$RttaGenotype == "Rtta-"], 2, function(x){x > 0.1})
escapees_control_late <- apply(escapees_control_late, 2, function(x){rownames(escapees)[x]})
escapees_control_late <- names(table(do.call("c", escapees_control_late)))[table(do.call("c", escapees_control_late)) > 9]
escapees_control_late <- escapees_control_late[!escapees_control_late %in% names(rowMeans(ratios, na.rm = T))[(rowMeans(ratios, na.rm = T)) > 0.9]]
length(escapees_control_late)

escapees_control <- escapees_control_late

# For the supplement, also plot stratified by replicates: 
ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Sample" = data_here$ShortSampleName) %>%
  add_column("Stage" = data_here$embryo_stage) %>%
  add_column("Genotype" = data_here$RttaGenotype) %>%
  pivot_longer(-c(Sample, Stage, Genotype)) %>%
  mutate(name = factor(name, levels = unique(.$name))) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) -> to_plot

genes_show <- to_plot %>% arrange(position) %>% pull(position, name = name)
genes_show <- genes_show[unique(names(genes_show))]

genes_show <- c("Xist", "Mecp2", "Kdm6a", "Kdm5c")
genes_of_choice <- setNames(rep("", nrow(ratios)), rownames(ratios))
genes_of_choice[genes_show] <- genes_show

to_plot %>%
  mutate(Genotype = factor(ifelse(Genotype == "Rtta-", "Rtta (negative)", "Rtta (positive)"), 
                            levels = c("Rtta (positive)", "Rtta (negative)"))) %>%
  mutate(value = ifelse(is.na(value), 0, value)) %>%
  # dplyr::filter(Stage == "Late") %>%
  dplyr::filter(Genotype == "Rtta (negative)") %>%
  group_by(Genotype, Stage, name, position) %>%
  summarize(value = median(value)) %>%
  mutate(Stage = factor(Stage, levels = rev(c("Early", "Late")))) %>%
  ggplot(aes(reorder(name, position), y = Stage, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    theme_paper(textsize = 30) + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    labs(color = "Allelic ratio \n (Xi / (Xi + Xa))") + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 8)) + 
    scale_x_discrete(labels = genes_of_choice, position = "top") + 
    guides(fill = "none")
ggplot2::ggsave("./Plots/Fig6/heatmap_ordered_all_genes_by_position.pdf", width = 20, height = 5)

to_plot %>%
  mutate(Genotype = factor(ifelse(Genotype == "Rtta-", "Rtta (negative)", "Rtta (positive)"), 
                            levels = c("Rtta (positive)", "Rtta (negative)"))) %>%
  mutate(value = ifelse(is.na(value), 0, value)) %>%
  # dplyr::filter(Stage == "Late") %>%
  dplyr::filter(Genotype == "Rtta (negative)") %>%
  group_by(Genotype, Stage, name, position) %>%
  summarize(value = median(value)) %>%
  mutate(Stage = factor(Stage, levels = rev(c("Early", "Late")))) %>%
  group_by(Stage, value > 0.1) %>%
  summarize(n = n())

## 

## define escapee categoryes 
# compare to NPC data: 
npc_escape <- read.csv("./ProcessedData/excel_files/escape_metaanalysis_final.escape_per_clone.csv")
npc_escape$E6_merged <- ifelse(npc_escape$E6...Rep1 == "Escapee" & npc_escape$E6...Rep2 == "Escapee" & npc_escape$E6...Rep3 == "Escapee",
                               "Escapee", "Silenced")
e6_escapees <- npc_escape %>% dplyr::filter(E6_merged == "Escapee") %>% pull(symbol)

escape_comparison <- data.frame(
  Category = rowData(data_here)[escapees_control, "EscapeAnnotation"], 
  E6_escape = (npc_escape %>% pull(E6_merged, name = symbol))[escapees_control]
) %>% 
  mutate(Category_New = case_when(
    Category == "Unassigned" ~ "Embryo-specific", 
    Category == "silenced" & E6_escape == "Silenced" ~ "Embryo-specific", 
    Category == "silenced" & E6_escape == "Escapee" ~ "NPC-specific", 
    Category == "facultative" ~ "Facultative", 
    Category == "constitutive" ~ "Constitutive"
  ))

sum(table(escape_comparison$Category))
table(escape_comparison$Category)
table(escape_comparison$Category_New)
table(escape_comparison$Category, escape_comparison$E6_escape)

rowData(data_here)[escapees_control, ] %>% data.frame() %>% 
  dplyr::filter(EscapeAnnotation == "Unassigned")

ratios[escapees_control, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Sample" = data_here$ShortSampleName) %>%
  add_column("Stage" = data_here$embryo_stage) %>%
  add_column("Genotype" = data_here$RttaGenotype) %>%
  pivot_longer(-c(Sample, Stage, Genotype)) %>%
  mutate(name = factor(name, levels = unique(.$name))) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) -> to_plot

genes_show <- to_plot %>% arrange(position) %>% pull(position, name = name)
genes_show <- genes_show[unique(names(genes_show))]

p1 <- to_plot %>%
  dplyr::filter(Stage == "Late") %>%
  mutate(value = ifelse(is.na(value), 0, value)) %>%
  mutate(Genotype = factor(ifelse(Genotype == "Rtta-", "Rtta (negative)", "Rtta (positive)"), 
                          levels = c("Rtta (positive)", "Rtta (negative)"))) %>% 
  group_by(Genotype, name, position) %>%
  summarize(value = mean(value)) %>%
  ggplot(aes(reorder(name, position), y = Genotype, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    theme_paper(textsize = 30) + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    labs(color = "Allelic ratio \n (Xi / (Xi + Xa))") + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 8)) + 
    scale_x_discrete(position = "top") + 
    guides(fill = "none")

# Plot escapee annotation at the bottom of this: 
p2 <- data.frame(Gene = names(genes_show),
                 Annotation = escape_comparison[names(genes_show), ]$Category_New, 
                 Position = genes_show) %>%
  ggplot(aes(x = reorder(Gene, Position), y = 1, fill = Annotation)) + theme_void() + geom_tile(col = "black") + 
    scale_fill_manual(values = escape_colors)

p1 / p2 + plot_layout(heights = c(4, 0.2))
# ggplot2::ggsave("./Plots/Fig6/heatmap_ordered_only_escapees_by_position.pdf", width = 20, height = 5)

to_plot %>%
  dplyr::filter(Stage == "Late") %>%
  mutate(value = ifelse(is.na(value), 0, value)) %>%
  mutate(Genotype = factor(ifelse(Genotype == "Rtta-", "Rtta (negative)", "Rtta (positive)"), 
                          levels = c("Rtta (positive)", "Rtta (negative)"))) %>%
  group_by(Genotype, name, position) %>%
  summarize(value = mean(value)) %>%
  pivot_wider(values_from = value, names_from = Genotype) %>%
  mutate(escape_difference = `Rtta (positive)` - `Rtta (negative)`) %>%
  ggplot(aes(reorder(name, position), y = "1", fill = escape_difference, col = escape_difference)) + geom_tile(width = 0.9) +
    scale_fill_gradient2() + 
    scale_colour_gradient2() + 
    theme_paper(textsize = 30) + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    labs(color = "Allelic ratio \n difference") + 
    #scale_x_discrete(position = "bottom") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 8)) + 
    theme(axis.text.x = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank()) + 
    guides(fill = "none") -> p3
ggplot2::ggsave("./Plots/Fig6/heatmap_ordered_only_escapees_by_position_difference.pdf", width = 20, height = 2)

p1 / p2 / p3 + plot_layout(heights = c(1, 0.2, 0.5))
ggplot2::ggsave("./Plots/Fig6/heatmap_ordered_only_escapees_by_position.pdf", width = 20, height = 5)

# Quantify for genes that escape in E6 in the Control sample
# ratios[escapees_control, ] %>%
#   t() %>% data.frame() %>%
#   add_column("Sample" = data_here$ShortSampleName) %>%
#   add_column("Stage" = data_here$embryo_stage) %>%
#   add_column("Genotype" = data_here$RttaGenotype) %>%
#   pivot_longer(-c(Sample, Stage, Genotype)) %>%
#   add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
#   mutate(escape_status = factor(escape_status, c("silenced", "facultative", "constitutive"))) %>%
#   dplyr::filter(!is.na(escape_status)) %>%
#   ggplot(aes(x = value, y = Sample, fill = factor(escape_status, levels = rev(levels(escape_status))))) + 
#   geom_boxplot(col = "black", outlier.color = "grey") + 
#   theme_paper() + xlab("Allelic ratio (Xi / (Xi + Xa))") + ylab("") + 
#   theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
#   scale_fill_manual(values = c("orange", "darkgreen", "grey")) + 
#   labs(fill = "Escape Category") + facet_grid(rows = vars(Stage, Genotype), scales = "free_y", space = "free_y")
# ggplot2::ggsave("./Plots/Fig6/escapee_silencing_boxplots.pdf", width = 9, height = 6)

ratios[escapees_control, ] %>%
  t() %>% data.frame(fix.empty.names = F) %>%
  add_column("Sample" = data_here$ShortSampleName) %>%
  add_column("Stage" = data_here$embryo_stage) %>%
  add_column("Genotype" = data_here$RttaGenotype) %>%
  pivot_longer(-c(Sample, Stage, Genotype)) %>%
  dplyr::filter(Stage == "Late") %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = ifelse(escape_status == "Unassigned", "Embryo-specific", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "silenced" & !(name %in% e6_escapees), "Embryo-specific", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "silenced" & (name %in% e6_escapees), "NPC-specific", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "facultative", "Facultative", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "constitutive", "Constitutive", escape_status)) %>%
  mutate(escape_status = factor(escape_status, c("Embryo-specific", "NPC-specific", "Facultative", "Constitutive"))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  group_by(Stage, Genotype, name, escape_status) %>%
  summarize(value = mean(value, na.rm = T)) %>%
  mutate(Genotype = factor(Genotype, levels = c("Rtta+", "Rtta-"))) %>%
  mutate(Stage = paste(Stage, " Blastocyst")) %>%
  ggplot(aes(x = value, y = Genotype, fill = factor(escape_status, levels = rev(levels(escape_status))))) + 
    geom_boxplot(col = "black", outlier.color = "grey") + 
    theme_paper() + xlab("Allelic ratio (Xi / (Xi + Xa))") + ylab("") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    scale_fill_manual(values = c("orange", "darkgreen", "grey", "lightblue")) + 
    labs(fill = "Escape Category")
ggplot2::ggsave("./Plots/Fig6/escapee_silencing_boxplots_grouped.pdf", width = 9, height = 6)

ratios[escapees_control, ] %>%
  t() %>% data.frame(fix.empty.names = F) %>%
  add_column("Sample" = data_here$ShortSampleName) %>%
  add_column("Stage" = data_here$embryo_stage) %>%
  add_column("Genotype" = data_here$RttaGenotype) %>%
  pivot_longer(-c(Sample, Stage, Genotype)) %>%
  dplyr::filter(Stage == "Late") %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = ifelse(escape_status == "Unassigned", "Embryo-specific", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "silenced" & !(name %in% e6_escapees), "Embryo-specific", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "silenced" & (name %in% e6_escapees), "NPC-specific", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "facultative", "Facultative", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "constitutive", "Constitutive", escape_status)) %>%
  mutate(escape_status = factor(escape_status, c("Embryo-specific", "NPC-specific", "Facultative", "Constitutive"))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  group_by(Stage, Genotype, name, escape_status) %>%
  summarize(value = mean(value, na.rm = T)) %>%
  mutate(Genotype = factor(Genotype, levels = c("Rtta+", "Rtta-"))) %>%
  mutate(Stage = paste(Stage, " Blastocyst")) %>%
  ungroup() %>%
  compare_means(value ~ Genotype, data = ., paired = T)

ratios[escapees_control, ] %>%
  t() %>% data.frame(fix.empty.names = F) %>%
  add_column("Sample" = data_here$ShortSampleName) %>%
  add_column("Stage" = data_here$embryo_stage) %>%
  add_column("Genotype" = data_here$RttaGenotype) %>%
  pivot_longer(-c(Sample, Stage, Genotype)) %>%
  dplyr::filter(Stage == "Late") %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = ifelse(escape_status == "Unassigned", "Embryo-specific", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "silenced" & !(name %in% e6_escapees), "Embryo-specific", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "silenced" & (name %in% e6_escapees), "NPC-specific", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "facultative", "Facultative", escape_status)) %>%
  mutate(escape_status = ifelse(escape_status == "constitutive", "Constitutive", escape_status)) %>%
  mutate(escape_status = factor(escape_status, c("Embryo-specific", "NPC-specific", "Facultative", "Constitutive"))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  group_by(Stage, Genotype, name, escape_status) %>%
  summarize(value = mean(value, na.rm = T)) %>%
  mutate(Genotype = factor(Genotype, levels = c("Rtta+", "Rtta-"))) %>%
  mutate(Stage = paste(Stage, " Blastocyst")) %>%
  compare_means(value ~ Genotype, group.by = c("escape_status"), data = ., paired = T)

# ratios %>%
#   t() %>% data.frame() %>%
#   add_column("Sample" = data_here$ShortSampleName) %>%
#   add_column("Stage" = data_here$embryo_stage) %>%
#   add_column("Genotype" = data_here$RttaGenotype) %>%
#   pivot_longer(-c(Sample, Stage, Genotype)) %>%
#   add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
#   mutate(escape_status = factor(escape_status, c("silenced", "facultative", "constitutive"))) %>%
#   dplyr::filter(!is.na(escape_status)) %>%
#   group_by(Sample, Stage, Genotype, escape_status) %>%
#   summarize(value = median(value, na.rm = T)) %>%
#   ggplot(aes(x = value, y = Sample, col = escape_status)) + 
#     geom_point() + 
#     theme_paper() + xlab("Allelic ratio (Xi / (Xi + Xa))") + ylab("") + 
#     theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
#     labs(fill = "Escape Category") + facet_grid(rows = vars(Stage, Genotype), scales = "free_y", space = "free_y") + 
#     scale_color_manual(values = escape_colors)
# ggplot2::ggsave("./Plots/Fig6/escapee_silencing_boxplots_summarized.pdf", width = 9, height = 6)

# ratios[escapees_control, ] %>%
#   t() %>% data.frame() %>%
#   add_column("Sample" = data_here$ShortSampleName) %>%
#   add_column("Stage" = data_here$embryo_stage) %>%
#   add_column("Genotype" = data_here$RttaGenotype) %>%
#   pivot_longer(-c(Sample, Stage, Genotype)) %>%
#   add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
#   mutate(escape_status = factor(escape_status, c("silenced", "facultative", "constitutive"))) %>%
#   dplyr::filter(!is.na(escape_status)) %>%
#   group_by(Sample, Stage, Genotype, escape_status) %>%
#   summarize(value = median(value, na.rm = T)) %>%
#   ggplot(aes(x = value, y = Genotype, fill = escape_status)) +
#     geom_boxplot() +
#     theme_paper() + xlab("Allelic ratio (Xi / (Xi + Xa))") + ylab("") +
#     theme(axis.text.x=element_text(angle = 45, hjust = 1)) +
#     labs(fill = "Escape Category") + facet_grid(rows = vars(Stage, Genotype), scales = "free_y", space = "free_y") +
#     scale_fill_manual(values = escape_colors)
# ggplot2::ggsave("./Plots/Fig6/escapee_silencing_boxplots_summarized_boxplot.pdf", width = 9, height = 6)

### plot individual genes: 

ratios_here <- ratios[escapees_control, ]
ratios_here[(counts_inactive(data_here[escapees_control, ]) + counts_active(data_here[escapees_control, ])) <= 10] <- NA

plot_gene <- function(gene, save = F){
  data_per_gene <- data.frame(
    ASE = as.numeric(ratios_here[gene, ]), 
    Stage = data_here$embryo_stage,
    Genotype = data_here$RttaGenotype
  ) 
  p1 = data_per_gene %>% 
    mutate(Genotype = factor(Genotype, levels = c("Rtta+", "Rtta-"))) %>%
    ggplot(aes(x = Genotype, y = ASE)) + geom_boxplot(outlier.color = NA, width = 0.4) + geom_jitter(width = 0.1, pch = 21, size = 3, fill = "grey") + coord_flip() + 
    facet_wrap(~Stage, ncol = 1) + ylim(c(0, .75)) + theme_paper(textsize = 20) + 
    xlab("") + ggtitle(gene)
  if (save){
    ggsave(paste0("./Plots/Fig6/gene_plots/", gene, "_escape_exp2.pdf"))
  }
  return(p1)
}

### check genome-wide: 

ratios_here %>% 
  rownames_to_column("Gene") %>%
  pivot_longer(-c(Gene)) %>%
  add_column("Sample" = colData(data_here)[.$name, ]$ShortSampleName) %>%
  add_column("Stage" =  colData(data_here)[.$name, ]$embryo_stage) %>%
  add_column("Genotype" =  colData(data_here)[.$name, ]$RttaGenotype) %>%
  # dplyr::filter(Sample != "AACYJJ7M5_E45doxSept23_23s002684-1-1_loda_lane1A10B4") %>%
  dplyr::filter(Stage == "Late") %>%
  dplyr::filter(!is.na(value)) %>%
  group_by(Gene) %>%
  mutate(n_minus = sum(Genotype == "Rtta-"), n_plus = sum(Genotype == "Rtta+")) %>%
  arrange(n_minus + n_plus) -> summarized_here

# which genes do we exclude: 
summarized_here %>% 
  dplyr::filter(n_minus < 3 | n_plus < 3) %>% pull(Gene) %>% unique()

summarized_here %>%
  # dplyr::filter(n_minus >= 3 & n_plus >= 3) %>%
  dplyr::filter(Gene != "Xist") %>%
  dplyr::group_by(Gene) %>%
  group_modify( ~ {
    data.frame(
     mean_rtta_minus = mean(.$value[.$Genotype == "Rtta-"]), 
     mean_rtta_plus = mean(.$value[.$Genotype == "Rtta+"])
   )
  } ) %>% dplyr::rename("gene_name" = "Gene") %>%
  right_join(data.frame(rowData(data_here))[,c("gene_name", "EscapeAnnotation")]) %>% dplyr::rename("Gene" = "gene_name") %>%
  dplyr::filter(!is.na(mean_rtta_minus)) %>% ungroup() %>% 
  dplyr::mutate(dscore_difference = mean_rtta_minus - mean_rtta_plus) -> difference_results
table(difference_results$dscore_difference < -.0)
table(difference_results$dscore_difference > .0)

summarized_here %>%
  dplyr::filter(n_minus >= 3 & n_plus >= 3) %>%
  dplyr::filter(Gene != "Xist") %>%
  group_by(Gene) %>%
  group_modify( ~ {
    data.frame(
     pval = t.test(.$value ~ .$Genotype)$p.value, 
     mean_rtta_minus = mean(.$value[.$Genotype == "Rtta-"]), 
     mean_rtta_plus = mean(.$value[.$Genotype == "Rtta+"])
   )
  } ) %>% dplyr::rename("gene_name" = "Gene") %>%
  right_join(data.frame(rowData(data_here))[,c("gene_name", "EscapeAnnotation")]) %>% rename("Gene" = "gene_name") %>%
  dplyr::filter(!is.na(mean_rtta_minus)) %>% ungroup() -> t.test_results

t.test_results %>% 
  mutate(padj = p.adjust(pval, method = "BH")) %>%
  mutate(escape_diff = mean_rtta_plus - mean_rtta_minus) %>%
  mutate(significant = padj < 0.1) %>% 
  dplyr::filter(significant)

t.test_results %>% 
  mutate(padj = p.adjust(pval, method = "BH")) %>%
  mutate(escape_diff = mean_rtta_plus - mean_rtta_minus) %>%
  mutate(significant = padj < 0.1) %>% {
  ggplot(data = ., aes(x = mean_rtta_minus, mean_rtta_plus, col = significant)) + 
    geom_point(size = 5) + 
    geom_abline() + coord_fixed() + 
    theme_paper() + scale_color_manual(values = c("grey", "red")) + 
    ggrepel::geom_text_repel(aes(label = Gene), data = .[.$escape_diff < -0.1, ], size = 5) + 
    xlab("RTTA- (mean allelic ratio Xi / (Xi + Xa))") + ylab("RTTA+ (mean allelic ratio Xi / (Xi + Xa))") + 
    xlim(c(0, 0.7)) + ylim(c(0, 0.7))
}
ggplot2::ggsave("./Plots/Fig6/significant_reduction_scatterplot.pdf", width = 9, height = 6)

t.test_results %>% 
  mutate(EscapeAnnotation = ifelse(EscapeAnnotation == "Unassigned", "Embryo-specific", EscapeAnnotation)) %>%
  mutate(EscapeAnnotation = ifelse(EscapeAnnotation == "silenced" & !(Gene %in% e6_escapees), "Embryo-specific", EscapeAnnotation)) %>%
  mutate(EscapeAnnotation = ifelse(EscapeAnnotation == "silenced" & (Gene %in% e6_escapees), "NPC-specific", EscapeAnnotation)) %>%
  mutate(EscapeAnnotation = ifelse(EscapeAnnotation == "constitutive", "Constitutive", EscapeAnnotation)) %>%
  mutate(EscapeAnnotation = ifelse(EscapeAnnotation == "facultative", "Facultative", EscapeAnnotation)) %>%
  mutate(padj = p.adjust(pval)) %>%
  mutate(escape_diff = mean_rtta_plus - mean_rtta_minus) %>%
  mutate(significant = pval < 0.1) %>% {
  ggplot(data = ., aes(x = mean_rtta_minus, mean_rtta_plus, col = EscapeAnnotation)) + 
    geom_point(size = 4) + 
    geom_abline() + coord_fixed() + 
    theme_paper() + 
    scale_color_manual(values = c("Constitutive" = "orange", "Facultative" = "darkgreen", 
                                  "NPC-specific" = "grey", "Embryo-specific" = "lightblue")) + 
    ggrepel::geom_text_repel(aes(label = Gene), data = .[ .$escape_diff < -0.1 | .$EscapeAnnotation == "Embryo-specific", ], size = 5) + 
    xlab("RTTA- (mean allelic ratio Xi / (Xi + Xa))") + ylab("RTTA+ (mean allelic ratio Xi / (Xi + Xa))") + 
    xlim(c(0, 0.8)) + ylim(c(0, 0.8))
}
ggplot2::ggsave("./Plots/Fig6/significant_reduction_scatterplot_categories.pdf", width = 9, height = 6)

plot_genes <- c("Atp6ap2", "Kdm6a", "Slc25a5", "Lamp2", "Fmr1", "Mecp2", "G6pdx", "Kif4", "Rps4x", 
                # "Chic1", 
                "Rnf12", "Atrx", 
                # "Atp7a", 
                "Gla", "Fgd1", "Kdm5c", "Huwe1", "Pdha")

ratios_here[plot_genes, ] %>% 
  rownames_to_column("Gene") %>%
  pivot_longer(-c(Gene)) %>%
  mutate("Sample" = colData(data_here)[.$name, ]$ShortSampleName) %>%
  mutate("Stage" =  colData(data_here)[.$name, ]$embryo_stage) %>%
  mutate("Genotype" =  colData(data_here)[.$name, ]$RttaGenotype) %>%
  mutate("Stage_PC" = sample_to_pc1[name]) %>%
  ggplot(aes(x = Stage, y = value, fill = Genotype)) + 
    geom_jitter(pch = 21, size = 3, position = position_jitterdodge(jitter.width = 0.05, dodge.width = 0.5)) + 
    facet_wrap(~Gene, scales = "free_y") + geom_hline(yintercept = 0.5, linetype = 'dashed') + 
    theme_bw()

ratios_here[plot_genes, ] %>% 
  rownames_to_column("Gene") %>%
  pivot_longer(-c(Gene)) %>%
  mutate("Sample" = colData(data_here)[.$name, ]$ShortSampleName) %>%
  mutate("Stage" =  colData(data_here)[.$name, ]$embryo_stage) %>%
  mutate("Genotype" =  colData(data_here)[.$name, ]$RttaGenotype) %>%
  mutate("Stage_PC" = sample_to_pc1[name]) %>%
  dplyr::filter(Gene == "Kdm5c") %>% arrange(value)

ratios_here[plot_genes, ] %>% 
  rownames_to_column("Gene") %>%
  pivot_longer(-c(Gene)) %>%
  mutate("Sample" = colData(data_here)[.$name, ]$ShortSampleName) %>%
  mutate("Stage" =  colData(data_here)[.$name, ]$embryo_stage) %>%
  mutate("Genotype" =  colData(data_here)[.$name, ]$RttaGenotype) %>%
  mutate("Stage_PC" = sample_to_pc1[name]) %>%
  dplyr::filter(Stage == "Late") %>%
  ggplot(aes(x = Stage_PC, y = value, fill = Genotype)) + 
    geom_point(pch = 21, size = 3, position = position_jitterdodge(jitter.width = 0.05, dodge.width = 0.5)) + 
    facet_wrap(~Gene, scales = "free_y") + geom_hline(yintercept = 0.5, linetype = 'dashed') + 
    theme_bw()

# ratios_here[plot_genes, ] %>% 
#   # data.frame() %>%
#   rownames_to_column("Gene") %>%
#   pivot_longer(-c(Gene)) %>%
#   mutate("Sample" = colData(data_here)[.$name, ]$ShortSampleName) %>%
#   mutate("Stage" =  colData(data_here)[.$name, ]$embryo_stage) %>%
#   mutate("Genotype" =  colData(data_here)[.$name, ]$RttaGenotype) %>%
#   ggplot(aes(x = Genotype, y = Gene, fill = value)) + geom_tile() + 
#     scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
#     facet_wrap(~Stage)

# ratios_here[plot_genes, ] %>% 
#   # data.frame() %>%
#   rownames_to_column("Gene") %>%
#   pivot_longer(-c(Gene)) %>%
#   mutate("Sample" = colData(data_here)[.$name, ]$ShortSampleName) %>%
#   mutate("Stage" =  colData(data_here)[.$name, ]$embryo_stage) %>%
#   mutate("Genotype" =  colData(data_here)[.$name, ]$RttaGenotype) %>%
#   ggplot(aes(x = Stage, y = value, fill = Genotype)) + 
#     geom_jitter(pch = 21, size = 3, position = position_jitterdodge(jitter.width = 0.05, dodge.width = 0.5)) + 
#     facet_wrap(~Gene, scales = "free_y") + geom_hline(yintercept = 0.5, linetype = 'dashed') + 
#     theme_bw()
# ggplot2::ggsave("./Plots/Fig6/escapee_specific_genes.pdf", width = 9, height = 6)

agnese_genes <- readxl::read_excel("./ProcessedData/in vivo_targets.xlsx")

plot_genes_2 <- agnese_genes$Target
plot_genes_2 <- plot_genes_2[plot_genes_2 %in% rownames(ratios_here)]
ratios_here[plot_genes_2, ] %>% 
  rownames_to_column("Gene") %>%
  pivot_longer(-c(Gene)) %>%
  mutate("Sample" = colData(data_here)[.$name, ]$ShortSampleName) %>%
  mutate("Stage" =  colData(data_here)[.$name, ]$embryo_stage) %>%
  mutate("Genotype" =  colData(data_here)[.$name, ]$RttaGenotype) %>%
  ggplot(aes(x = Stage, y = value, fill = Genotype)) + 
    geom_jitter(pch = 21, size = 3, position = position_jitterdodge(jitter.width = 0.05, dodge.width = 0.5)) + 
    facet_wrap(~Gene, scales = "free_y") + geom_hline(yintercept = 0.5, linetype = 'dashed') + 
    theme_bw()
ggplot2::ggsave("./Plots/Fig6/escapee_specific_genes_2.pdf", width = 9, height = 6)

### 

# plot_gene2 <- function(gene, save = F){
#   data_per_gene <- data.frame(
#     ASE = as.numeric(ratios_here[gene, ]), 
#     Stage = data_here$embryo_stage,
#     Genotype = data_here$RttaGenotype, 
#     diff_pc1 = data_here$embryo_pc
#   ) 
#   p1 <- data_per_gene %>% 
#     mutate(Genotype = factor(Genotype, levels = c("Rtta+", "Rtta-"))) %>%
#     ggplot(aes(x = diff_pc1, y = ASE, fill = Genotype)) + geom_point(pch = 21, size = 4) + 
#       theme_paper(textsize = 20) + geom_smooth(method = "lm", se = F, aes(col = Genotype)) + 
#       xlab("") + ggtitle(gene) + ylim(c(0, 1))
#   if (save){
#     ggsave(paste0("./Plots/Fig5/gene_plots/", gene, "_escape_exp2_curve_2.pdf"))
#   }
#   return(p1)
# }
# 
# plot_gene2("Atp6ap2", save = T)
# plot_gene2("Kdm6a", save = T)
# plot_gene2("Slc25a5", save = T)
# plot_gene2("Lamp2", save = T)
# plot_gene2("Fmr1", save = T)
# plot_gene2("Mecp2", save = T)
# plot_gene2("G6pdx", save = T)
# plot_gene2("Kif4", save = T)
# plot_gene2("Rps4x", save = T)
# plot_gene2("Chic1", save = T)
# plot_gene2("Rnf12", save = T)
# plot_gene2("Atrx", save = T)
# plot_gene2("Atp7a", save = T)
# plot_gene2("Gla", save = T)
# plot_gene2("Fgd1", save = T)
# plot_gene2("Kdm5c", save = T)
# plot_gene2("Huwe1", save = T)
# plot_gene2("Pdha", save = T)


```

Export mean expression levels and d-scores per condition: 

```{r asd}

# all_genes <- genes(EnsDb.Mmusculus.v79) %>% data.frame() %>% dplyr::filter(seqnames == "X")
# all_genes <- setNames(all_genes$end - all_genes$start, nm = all_genes$gene_name) / 1e3
# 
# export_df <- data.frame(
#   Gene = rownames(data_here),
#   Counts_Active_Early = rowSums(counts_active(data_here[,data_here$embryo_stage == "Early"])),
#   Counts_Inactive_Early = rowSums(counts_inactive(data_here[,data_here$embryo_stage == "Early"])),
#   Counts_Active_Late = rowSums(counts_active(data_here[,data_here$embryo_stage == "Late"])),
#   Counts_Inactive_Late = rowSums(counts_inactive(data_here[,data_here$embryo_stage == "Late"]))
# ) %>%
#   mutate(RPK_Early = (Counts_Active_Early + Counts_Inactive_Early) / all_genes[.$Gene] ) %>%
#   mutate(RPK_Late = (Counts_Active_Late + Counts_Inactive_Late) / all_genes[.$Gene] ) %>%
#   mutate(Ratios_Early = Counts_Inactive_Early / (Counts_Active_Early + Counts_Inactive_Early)) %>%
#   mutate(Ratios_Late = Counts_Inactive_Late / (Counts_Active_Late + Counts_Inactive_Late))
# write.csv(export_df, "./ProcessedData/export_invivo.csv")
  
```

```{r autosomal_late}

# deseq_dataset <- deseq_dataset[,colnames(data_here)]
# deseq_dataset$embryo_stage <- factor(data_here$embryo_stage)
# deseq_dataset_here <- deseq_dataset[,deseq_dataset$embryo_stage == "Late"]
# 
# deseq_dataset_here <- estimateSizeFactors(deseq_dataset_here)
# 
# deseq_dataset_here$RttaGenotype <- factor(ifelse(deseq_dataset_here$RttaGenotype == "Rtta+", "Rtta_pos", "Rtta_neg"))
# design(deseq_dataset_here) = ~ RttaGenotype
# deseq_dataset_here <- DESeq(deseq_dataset_here)
# 
# deseq_autosome_results <- results(deseq_dataset_here) %>% data.frame() %>% dplyr::arrange(pvalue)
# deseq_autosome_results %>% ggplot(aes(x = baseMean, y = log2FoldChange, col = padj < 0.1)) + geom_point() + scale_x_log10()

```