---
title: "xist_paper_figure2_astrocytes"
output: html_document
author: "Jasper Panten"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

This script reproduces all plots for Figure 2 and associated supplementary figures. Preprocessing of raw RNA sequencing data is done in ./Preprocessing/..., that yields allele-specific expression counts which are combined into a SingleCellExperiment (SCE) object in ./Preprocessing/generate_sce_astrocytes.R. 

```{r load_libraries}

setwd(paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/../../"))

library(tidyverse)
library(ggplot2)
library(scran)
library(scater)
library(DESeq2)
library(patchwork)

source("./Scripts/Rscripts/auxiliary.R")

theme_paper <- function(){
  theme(panel.background = element_blank(),
        panel.grid = element_line(colour= "grey92"), 
        panel.grid.minor = element_line(size = rel(0.5)),
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1), 
        text = element_text(size = 20))
}

counts_active <- function(sce){assays(sce)[["counts_active"]]}
counts_inactive <- function(sce){assays(sce)[["counts_inactive"]]}

escape_colors <- setNames(c("grey", "darkgreen", "orange"), nm = c("NPC-specific", "Facultative", "Constitutive"))

```

Now, read in the SCE, remove non-X genes and exclude lowly expressed genes

```{r load_data}

data <- readRDS("../ProcessedData/merged_dataset_astrocytes.rds")

## get escape annotation from previous dataset: 
data_npcs <- readRDS("../ProcessedData/merged_dataset_paper.rds")

rowData(data) <- cbind(rowData(data)[, 1:6],
                       EscapeAnnotation = rowData(data_npcs[rownames(data_npcs), ])$EscapeAnnotation, 
                       EscapeAnnotation2 = rowData(data[rownames(data), ])$EscapeAnnotation)

### export counts for geo: 
counts(data) %>% rownames_to_column("Gene") %>% write_delim("../ProcessedData/geo_astrocytes/astrocytes_total_counts.txt", delim = "\t")
counts_active(data) %>% rownames_to_column("Gene") %>% write_delim("../ProcessedData/geo_astrocytes/astrocytes_active_counts.txt", delim = "\t")
counts_inactive(data) %>% rownames_to_column("Gene") %>% write_delim("../ProcessedData/geo_astrocytes/astrocytes_inactive_counts.txt", delim = "\t")

data_before_filtering <-  data[seqnames(data) == "X", ]

data <- data[(rowSums(counts_inactive(data)) + rowSums(counts_active(data))) / ncol(data) > 10, ]

data_with_autosomes <- data
data <- data[seqnames(data) == "X", ]
data$Replicates <- str_extract(data$SampleName, "R1|R2")

convert_names <- list(
  "Astro_dox3" = "3d dox", 
  "Astro_dox7" = "7d dox", 
  "Astro_dox3_wo" = "3d dox + washout", 
  "Astro_dox7_wo" = "7d dox + washout", 
  "Astro_no_dox3" = "3d no dox", 
  "Astro_no_dox7" = "7d no dox", 
  "Astro_no_dox3_no_wo" = "3d no dox + washout", 
  "Astro_no_dox7_no_wo" = "7d no dox + washout"
)

data$ConditionClean <- as.character(convert_names[data$ConditionClean])

table(rowData(data)$EscapeAnnotation, rowData(data)$EscapeAnnotation2)

rowData(data) %>%
  data.frame() %>%
  mutate(EscapeAnnotation2 = case_when(
    EscapeAnnotation2 == "constitutive" ~ "Constitutive", 
    EscapeAnnotation2 == "facultative" ~ "Facultative", 
    EscapeAnnotation2 == "silenced / variable" ~ "NPC-specific",
    .default = ""
  )) %>%
  dplyr::filter(EscapeAnnotation2 != "") %>%
  dplyr::filter(EscapeAnnotation != EscapeAnnotation2)

```

We first quantify the number of genes which escape within and across clones and cell lines. 

```{r supplement_number_of_escapees}

data_here <- data[,data$Condition %in% c("Astro_no_dox3", "Astro_no_dox7", "Astro_no_dox3_no_wo", "Astro_no_dox7_no_wo")]

# get d-scores across genes + samples
ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))

# Genes with average ASE > 0.7 in the control sample are likely mapping artefacts: 
genes_out <- names(rowMeans(ratios, na.rm = T)[rowMeans(ratios, na.rm = T) > 0.8 & rownames(ratios) != "Xist"])
genes_out_na <- rownames( ratios[!apply(ratios, 1, function(x){!any(is.na(x))}), ] )

genes_out <- c(genes_out, genes_out_na)

ratios <- ratios[!rownames(ratios) %in% genes_out, ]

genes_show <- c("Xist", "Mecp2", "Kdm6a", "Kdm5c")
genes_of_choice <- setNames(rep("", nrow(ratios)), rownames(ratios))
genes_of_choice[genes_show] <- genes_show

# Plot d-scores for genes after ordering the genes by chromosomale coordinate
ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Sample" = data_here$SampleName) %>%
  pivot_longer(-c(Condition, Sample)) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) -> df_plot_s5b

df_plot_s5b %>%
  ggplot(aes(Sample, y = reorder(name, position), fill = value)) + geom_tile(width = 0.9) +
    theme_paper() + theme(axis.text.x=element_text(angle = 90, hjust = 1)) + 
    theme(axis.ticks.y = element_blank()) + xlab("") + ylab("Chromosome position") + 
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_y_discrete(labels = genes_of_choice, position = "right") + ylab("") + labs(fill = "d-score")
ggsave("../Figures/FiguresAstrocytes/astrocyte_heatmap_controls.pdf", width = 10, height = 10)
ggsave("./FiguresIllustrator/FigS1/heatmap_ordered.pdf", width = 10, height = 10)

# Plot d-scores for genes that escape in any sample, excluding genes with > 0.7 d-score
ratios <- ratios[!rownames(ratios) %in% genes_out, ]
ratios <- ratios[order(rowMeans(ratios), decreasing = T), ]

# Now we define the baseline escapees per sample as the genes with d-score > 0.1, this will be used throughout the analysis
escapees <- apply(ratios, 2, function(x){x > 0.1})
escapees_per_clone <- apply(escapees, 2, function(x){rownames(escapees)[x]})
escapees_per_clone <- lapply(escapees_per_clone, function(x){x[x != "Xist"]})

# How many do we get per control sample and do they overlap?
ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Sample" = data_here$SampleName) %>%
  pivot_longer(-c(Condition, Sample)) %>%
  dplyr::filter(value > 0.1) %>%
  group_by(name) %>%
  mutate(escape_in_clones = n() >= 8) -> df_plot_s5c

df_plot_s5c %>% # CAVE - here we need to adjust
  ggplot(aes(x = Sample, fill = escape_in_clones)) + geom_bar() + theme_paper() + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + xlab("") + ylab("Number of genes with d-score > 0.1") + 
    scale_fill_manual(values = c("grey", "darkred"), name = "Escape in all samples")
ggsave("../Figures/FiguresAstrocytes/astrocyte_number_of_escapees_per_clone.pdf", width = 8, height = 8)

ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Sample" = data_here$SampleName) %>%
  pivot_longer(-c(Condition, Sample)) %>%
  dplyr::filter(value > 0.1) %>%
  group_by(name) %>%
  mutate(escape_in_clones = n() >= 8) %>%
  add_column(escape_group = rowData(data[.$name, ])$EscapeAnnotation) %>%
  dplyr::filter(!is.na(escape_group)) -> df_plot_s5d

df_plot_s5d %>%
  ggplot(aes(x = Sample, fill = escape_in_clones)) + geom_bar() + theme_paper() + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + xlab("") + ylab("Number of genes with d-score > 0.1") + 
    scale_fill_manual(values = c("grey", "darkred"), name = "Escape in all clones") + facet_wrap(~escape_group)
ggsave("../Figures/FiguresAstrocytes/astrocyte_number_of_escapees_per_clone_stratified.pdf", width = 8, height = 8)

ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Sample" = data_here$SampleName) %>%
  pivot_longer(-c(Condition, Sample)) %>%
  dplyr::filter(value > 0.1) %>%
  group_by(name) %>%
  mutate(escape_in_clones = n() >= 8) %>%
  add_column(escape_group = rowData(data[.$name, ])$EscapeAnnotation) %>%
  # dplyr::filter(!is.na(escape_group)) %>%
  group_by(name, escape_group, escape_in_clones) %>%
  summarize(mean_escape = mean(value)) %>%
  dplyr::filter(escape_in_clones) %>%
  write.csv("../ProcessedData/escapees_baseline_astrocytes.csv")

```

For the main figure, we now plot escape during Xist-overexpression (Fig1 a-...)

```{r }

# Subset data on relevant samples and for the main figure, only show E6
data_here <- data[ , data$Condition %in% c("Astro_dox3", "Astro_dox7", "Astro_dox3_wo", "Astro_dox7_wo", "Astro_no_dox3", "Astro_no_dox7")]

# For Xist-expression, we need a scaling factor per sample to account for differences in sequencing depth
size_factors <- colSums(counts(data_with_autosomes)) / colSums(counts(data_with_autosomes))[[1]]

# check that Xist-overexpression works by plotting Xist counts per million (CPM) per sample -- all samples
data.frame(
  Condition = data$ConditionClean,
  Replicate = data$Replicates, 
  Expression = as.numeric(counts(data["Xist", ]) / colSums(counts(data)) * 1e6)
) %>%
  mutate(Expression = Expression / mean(Expression[Condition == "3d no dox"])) -> xist_data


xist_data %>%
  ggplot(aes(x = Condition, y = Expression)) + 
    stat_summary(stat = "mean", geom = "bar", fill = "grey", col = "black") + 
    coord_flip() + theme_paper() + 
    xlab("") + ylab("Expression (Fold change)") + geom_jitter(size = 3, width = 0.1) + 
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
#ggsave("../Figures/FiguresAstrocytes/astrocyte_xist_overexpression_supplement.pdf", width = 4, height = 4)

# check that Xist-overexpression works by plotting Xist counts per million (CPM) per sample
data.frame(
  Condition = data_here$ConditionClean,
  Replicate = data_here$Replicates, 
  Expression = as.numeric(counts(data_here["Xist", ]) /  colSums(counts(data_here)) * 1e6)
) %>%
  dplyr::filter(!Condition %in% c("3d dox + washout", "7d dox + washout")) %>%
  mutate(Condition = factor(Condition, levels = c("7d dox", "7d no dox", "3d dox", "3d no dox"))) %>%
  mutate(Expression = Expression / mean(Expression[Condition == "3d no dox"])) -> xist_plot_data

xist_plot_data -> df_plot_2c

xist_plot_data %>%
  ggplot(aes(x = Condition, y = Expression)) + 
    stat_summary(stat = "mean", geom = "bar", fill = "grey", col = "black") + 
    coord_flip() + theme_paper() + 
    xlab("") + ylab("Expression (CPM)") + geom_jitter(size = 3, width = 0.1) + 
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggsave("../Figures/FiguresAstrocytes/astrocyte_xist_overexpression.pdf", width = 4, height = 4)
# ggsave("./FiguresIllustrator/Fig1/xist_overexpression.pdf", width = 4, height = 4)

xist_plot_data %>%
  group_by(Condition) %>%
  summarize(Expression = mean(Expression))

xist_plot_data %>% 
  dplyr::filter(Condition %in% c("3d no dox", "3d dox")) %>%
  group_by(Condition) %>%
  summarize(Expression = list(Expression)) %>%
  mutate(Expression_Control = Expression[Condition == "3d no dox"]) %>%
  rowwise() %>%
  dplyr::filter(Condition != "Control") %>%
  mutate(pval = t.test(Expression_Control, Expression)$p.value) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pval, method = "BH"))

xist_plot_data %>% 
  dplyr::filter(Condition %in% c("7d no dox", "7d dox")) %>%
  group_by(Condition) %>%
  summarize(Expression = list(Expression)) %>%
  mutate(Expression_Control = Expression[Condition == "7d no dox"]) %>%
  rowwise() %>%
  dplyr::filter(Condition != "Control") %>%
  mutate(pval = t.test(Expression_Control, Expression)$p.value) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pval, method = "BH"))

### 
data_here <- data

ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))
ratios <- ratios[!rownames(ratios) %in% genes_out, ]

# For the next heatmap, we want to highlight a few genes
genes_show <- c("Xist", "Mecp2", "Kdm6a", "Kdm5c")
genes_of_choice <- setNames(rep("", nrow(ratios)), rownames(ratios))
genes_of_choice[genes_show] <- genes_show

# Get escapees that are consistent across E6 samples
consistent_escapees <- Reduce("intersect", escapees_per_clone)

ratios[consistent_escapees, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Replicate" = data_here$Replicates) %>%
  pivot_longer(-c(Condition, Replicate)) %>%
  mutate(name = factor(name, levels = unique(.$name))) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  dplyr::filter(Condition %in% c("3d dox", "3d no dox", "7d dox", "7d no dox")) -> to_plot
  
to_plot %>%
  group_by(Condition, name, position) %>%
  summarize(value = mean(value)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  pivot_wider(names_from = Condition, values_from = value) %>%
  ungroup() %>%
  arrange(as.character(name)) %>% 
  write.csv("../ProcessedData/dscores_astrocytes_treatment.csv")

# Also do this reordering
to_plot -> df_plot_2d
p1 <- to_plot %>%
  group_by(Condition, name, position) %>%
  summarize(value = mean(value)) %>%
  mutate(Condition = factor(Condition, levels = c("7d dox", "7d no dox", "3d dox", "3d no dox"))) %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    labs(fill = "d-score") + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 8)) + 
    scale_x_discrete(position = "top") + guides(fill = "none")

# Plot escapee annotation at the bottom of this:
p2 <- data.frame(Gene = consistent_escapees,
           Annotation = rowData(data)[consistent_escapees, ]$EscapeAnnotation, 
           Position = start(rowRanges(data)[consistent_escapees, ])) %>%
  ggplot(aes(x = reorder(Gene, Position), y = 1, fill = Annotation)) + theme_void() + geom_tile(col = "black") + 
    scale_fill_manual(values = escape_colors) + 
    theme(legend.position = "None")

p1 / p2 + plot_layout(heights = c(4, 0.2))
ggsave("../Figures/FiguresAstrocytes/astrocyte_heatmap_ordered_only_escapees_by_position_FIXED.pdf", width = 20, height = 6)

# Quantify for genes that escape in E6 in the Control sample
ratios[consistent_escapees, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Replicate" = data_here$Replicates) %>%
  pivot_longer(-c(Condition, Replicate)) %>%
  group_by(Replicate) %>%
  # add_column(Condition = ifelse(.$Dox == 0, "Control", paste0("Dox (", .$Dox, " days)"))) %>%
  # mutate(Condition = factor(Condition, levels = (c("Control", paste0("Dox (", c(3, 7, 14, 21), " days)"))))) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, names(escape_colors))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  group_by(Condition, name, escape_status) %>%
  summarize(value = mean(value)) %>%
  ggplot(aes(x = Condition, y = value, fill = escape_status)) + 
    geom_boxplot(col = "black", outlier.color = "grey") + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = names(escape_colors)) + 
    labs(fill = "Escape Category") + xlab("")

ratios[consistent_escapees, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Replicate" = data_here$Replicates) %>%
  pivot_longer(-c(Condition, Replicate)) %>%
  group_by(Replicate) %>%
  # add_column(Condition = ifelse(.$Dox == 0, "Control", paste0("Dox (", .$Dox, " days)"))) %>%
  # mutate(Condition = factor(Condition, levels = (c("Control", paste0("Dox (", c(3, 7, 14, 21), " days)"))))) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, names(escape_colors))) %>%
  dplyr::filter(!is.na(escape_status)) -> df_plot_s5f

df_plot_s5f %>%
  group_by(Condition, name, escape_status) %>%
  summarize(value = mean(value)) %>%
  dplyr::filter(Condition %in% c("3d no dox", "3d dox", "7d no dox", "7d dox")) %>%
  mutate(Condition = factor(Condition, levels = rev(c("3d no dox", "3d dox", "7d no dox", "7d dox")))) %>%
  mutate(escape_status = factor(escape_status, levels = rev(c("NPC-specific", "Facultative", "Constitutive")))) %>%
  ggplot(aes(x = Condition, y = value, fill = escape_status)) + 
    geom_boxplot(col = "black", outlier.color = "grey") + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    scale_fill_manual(values = escape_colors) + 
    labs(fill = "Escape Category") + xlab("") + coord_flip()
ggsave("../Figures/FiguresAstrocytes/astrocyte_escapee_silencing_boxplots.pdf", width = 9, height = 6)

rename_conditions <- list(
  "3d no dox" = "Control (3d)", 
  "3d dox" = "Dox (3d)", 
  "7d no dox" = "Control (7d)", 
  "7d dox" = "Control (7d)"
)
ratios[consistent_escapees, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Replicate" = data_here$Replicates) %>%
  pivot_longer(-c(Condition, Replicate)) %>%
  group_by(Replicate) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, names(escape_colors))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  dplyr::filter(Condition %in% c("3d dox", "3d no dox", "7d dox", "7d no dox")) %>%
  group_by(name, Condition, escape_status) %>%
  summarize(value = mean(value)) %>%
  ungroup() %>%
  group_by(Condition, escape_status) %>%
  summarize(values = list(value)) %>%
  ungroup() %>%
  tidyr::expand(nesting(Condition, escape_status, values), nesting(Condition, escape_status, values),  .name_repair = "unique") %>%
  setNames(tolower(gsub("\\.","", names(.)))) %>%
  dplyr::filter((condition1 != condition4) & (escape_status2 == escape_status5)) %>%
  rowwise() %>%
  mutate(condition1 = sort(c(condition1, condition4))[1], condition4 = sort(c(condition1, condition4))[2]) %>% dplyr::filter(condition1 != condition4) %>%
  mutate(pvals = wilcox.test(values3, values6)$p.value, paired = T) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvals, method = "BH")) %>% 
  dplyr::select(-c("values3", "values6")) %>% 
  dplyr::select(c("condition1", "condition4", "escape_status5", "pvals", "padj")) %>%
  mutate(condition1 = unlist(rename_conditions[condition1]), condition4 = unlist(rename_conditions[condition4])) -> pval_results
write.csv(pval_results, "../ProcessedData/pvals_figs5f.csv")

# ggsave("./FiguresIllustrator/Fig1/escapee_silencing_boxplots.pdf", width = 9, height = 6)

ratios[consistent_escapees, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$Condition) %>%
  add_column("Replicate" = data_here$Replicates) %>%
  pivot_longer(-c(Condition, Replicate)) %>%
  group_by(Replicate) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, names(escape_colors))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  ggplot(aes(x = Condition, y = value, fill = Replicate)) + 
    geom_boxplot(col = "black", outlier.color = "grey", position = position_dodge2(preserve = "single")) + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    labs(fill = "Escape Category") + xlab("") + scale_fill_manual(values = c("red2", "green4"))

ratios[consistent_escapees, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$Condition) %>%
  add_column("Replicate" = data_here$Replicates) %>%
  pivot_longer(-c(Condition, Replicate)) %>%
  group_by(Replicate) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, names(escape_colors))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  dplyr::filter(Condition %in% c("Astro_no_dox3", "Astro_dox3", "Astro_no_dox7", "Astro_dox7")) %>%
  mutate(Condition = factor(Condition, levels = (c("Astro_no_dox3", "Astro_dox3", "Astro_no_dox7", "Astro_dox7")))) -> df_plot_s5e

df_plot_s5e %>%
  ggplot(aes(x = Condition, y = value, fill = Replicate)) + 
    geom_boxplot(col = "black", outlier.color = "grey", position = position_dodge2(preserve = "single")) + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    labs(fill = "Escape Category") + xlab("") + scale_fill_manual(values = c("red2", "green4"))
ggsave("../Figures/FiguresAstrocytes/astrocyte_escapee_silencing_boxplots_per_clone.pdf", width = 9, height = 6)

```

Compare escape and changes in escape to the results from npcs: 

```{r asd}

colData(data) <- colData(data)[ , c("Condition", "Clone")]

data_npc <- readRDS("../ProcessedData/merged_dataset.rds")
data_npc <- data_npc[seqnames(data_npc) == "X", ]
data_npc <- data_npc[rownames(data), ]
data_npc <- data_npc[, data_npc$Clone == "E6"]
data_npc <- data_npc[, data_npc$Condition %in% c("Aux_0_Dox_0_WO_0_WOAuxNO", "Aux_0_Dox_3_WO_0_WOAuxNO", "Aux_0_Dox_3_WO_4_WOAuxNO", 
                                                 "Aux_0_Dox_7_WO_0_WOAuxNO", "Aux_0_Dox_7_WO_4_WOAuxNO")]
data_npc <- data_npc[ , data_npc$Experiment != "October2023"]

# check that escapee annotations are consistent
data.frame(
  gene = rownames(data),
  astro = rowData(data)$EscapeAnnotation, 
  npc = rowData(data_npc)$EscapeAnnotation
) %>%
  dplyr::filter(astro != npc)

mcols(data_npc) <- mcols(data)
data_npc$Condition <- list(
  "Aux_0_Dox_0_WO_0_WOAuxNO" = "Astro_no_dox7", 
  "Aux_0_Dox_3_WO_0_WOAuxNO" = "Astro_dox3", 
  "Aux_0_Dox_3_WO_4_WOAuxNO" = "Astro_dox3_wo", 
  "Aux_0_Dox_7_WO_0_WOAuxNO" = "Astro_dox7", 
  "Aux_0_Dox_7_WO_4_WOAuxNO" = "Astro_dox7_wo",
  "Aux_0_Dox_7_WO_7_WOAuxNO" = "Astro_dox7_wo",
  "Aux_0_Dox_7_WO_14_WOAuxNO" = "Astro_dox7_wo"
)[data_npc$Condition] %>% unlist()
colData(data_npc) <- colData(data_npc)[ , c("Condition", "Clone")]

data$CellType = "Astrocyte"
data_npc$CellType = "NPC"

ratios_test <- counts_inactive(data_npc[consistent_escapees, data_npc$Condition  == "Astro_no_dox7"]) / 
  (counts_inactive(data_npc[consistent_escapees, data_npc$Condition  == "Astro_no_dox7"]) + counts_active(data_npc[consistent_escapees, data_npc$Condition  == "Astro_no_dox7"]))

sort(rowMeans(ratios_test))
names(rowMeans(ratios_test)[rowMeans(ratios_test) < .1])
table(apply(ratios_test, 1, function(x){sum(x >= 0.1)}) == 3)

ratios_test_2 <- counts_inactive(data[consistent_escapees, data$Condition  %in% c("Astro_no_dox3", "Astro_no_dox7")]) / 
  (counts_inactive(data[consistent_escapees, data$Condition %in% c("Astro_no_dox3", "Astro_no_dox7")]) + counts_active(data[consistent_escapees, data$Condition %in% c("Astro_no_dox3", "Astro_no_dox7")]))

rowMeans(ratios_test_2[names(rowMeans(ratios_test)[rowMeans(ratios_test) < .1]), ])

#
read_csv("../ProcessedData/d_score_differences.csv") %>% pull(name) -> esacpe_set
# read_csv("./ProcessedData/figure1_d_score_differences_all_genes_e6.csv") %>% dplyr::filter(Control > .1) %>% pull(name) -> esacpe_set

data_merged <- cbind(data, data_npc)

ratios <- counts_inactive(data_merged) / (counts_inactive(data_merged) + counts_active(data_merged))

ratios %>%
  t() %>% data.frame() %>%
  add_column("Condition" = data_merged$Condition) %>%
  add_column("CellType" = data_merged$CellType) %>%
  pivot_longer(-c(Condition, CellType)) %>% 
  group_by(Condition, CellType, name) %>%
  summarize(mean = mean(value)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, names(escape_colors))) %>%
  dplyr::filter(Condition == "Astro_no_dox7") %>%
  pivot_wider(names_from = CellType, values_from = mean) %>%
  dplyr::filter(name != "Xist") %>%
  dplyr::filter(!is.na(escape_status)) -> plot_here

table(plot_here$Astrocyte > .1, plot_here$NPC > .1) # 1 - ((8 + 11) / (8 + 11 + 227 + 140)) > 95% the same
table(plot_here$Astrocyte > .1, plot_here$NPC > .1) %>% fisher.test()

# plot_here %>% 
#   mutate(e6_set = name %in% e6_escapees) %>%
#   dplyr::filter(Astrocyte > .1 & e6_set) %>%
#   dplyr::filter(name != "Xist") %>% nrow()

plot_here -> df_plot_2e_1
plot_here %>% {
    ggplot(data = ., aes(x = NPC, y = Astrocyte, fill = escape_status)) + 
      geom_point(pch = 21, size = 4) +
      theme_paper() + ylab("ASE (d-score)") + 
      scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = c("Silenced / Variable", "Facultative", "Constitutive")) + 
      labs(fill = "Escape Category") + xlab("NPC") + ylab("Astrocyte") + geom_abline() + 
      xlim(c(0, 1)) + ylim(c(0, 1)) + 
      # ggrepel::geom_text_repel(aes(label = name), data = . %>% dplyr::filter(abs(NPC - Astrocyte) > 0.15)) + 
      annotate(x = 0.8, y = 0.2, label = paste0("Pearson correlation:", 
                                                round(cor(.$Astrocyte, .$NPC, use = "complete.obs"), digits = 4)), geom = "text") + 
      annotate(x = 0.8, y = 0.15, label = paste0("Genes matching: 95%"), geom = "text")
  }
ggsave("../Figures/FiguresAstrocytes/astro_npc_compare_baseline_escape.pdf", width = 9, height = 6)

ratios[esacpe_set, ] %>%
  t() %>% data.frame() %>%
  add_column("Condition" = data_merged$Condition) %>%
  add_column("CellType" = data_merged$CellType) %>%
  pivot_longer(-c(Condition, CellType)) %>% 
  group_by(Condition, CellType, name) %>%
  summarize(mean = mean(value)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, names(escape_colors))) %>%
  dplyr::filter(Condition == "Astro_dox3") %>%
  pivot_wider(names_from = CellType, values_from = mean) %>%
  dplyr::filter(!is.na(escape_status)) -> df_plot_2e_2

df_plot_2e_2 %>% {
    ggplot(data = ., aes(x = NPC, y = Astrocyte, fill = escape_status)) + 
      geom_point(pch = 21, size = 4) +
      theme_paper() + ylab("ASE (d-score)") + 
      scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = c("NPC-specific", "Facultative", "Constitutive")) + 
      labs(fill = "Escape Category") + xlab("NPC") + ylab("Astrocyte") + geom_abline() + 
      xlim(c(0, .8)) + ylim(c(0, .8))
  }
ggsave("../Figures/FiguresAstrocytes/astro_npc_compare_3d_escape.pdf", width = 9, height = 6)

ratios[esacpe_set, ] %>%
  t() %>% data.frame() %>%
  add_column("Condition" = data_merged$Condition) %>%
  add_column("CellType" = data_merged$CellType) %>%
  pivot_longer(-c(Condition, CellType)) %>% 
  group_by(Condition, CellType, name) %>%
  summarize(mean = mean(value)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, names(escape_colors))) %>%
  dplyr::filter(Condition == "Astro_dox7") %>%
  pivot_wider(names_from = CellType, values_from = mean) %>%
  dplyr::filter(!is.na(escape_status)) -> df_plot_2e_3

df_plot_2e_3 %>% {
    ggplot(data = ., aes(x = NPC, y = Astrocyte, fill = escape_status)) + 
      geom_point(pch = 21, size = 4) +
      theme_paper() + ylab("ASE (d-score)") + 
      scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = c("Silenced / Variable", "Facultative", "Constitutive")) + 
      labs(fill = "Escape Category") + xlab("NPC") + ylab("Astrocyte") + geom_abline() + 
      xlim(c(0, .8)) + ylim(c(0, .8))
      # ggrepel::geom_text_repel(aes(label = name), data = . %>% dplyr::filter(abs(NPC - Astrocyte) > 0.15)) + 
  }
ggsave("../Figures/FiguresAstrocytes/astro_npc_compare_7d_escape.pdf", width = 9, height = 6)

df_plot_2e <- rbind(
  cbind(df_plot_2e_1, "Condition" = "Control"), 
  cbind(df_plot_2e_2, "Condition" = "3d dox"), 
  cbind(df_plot_2e_3, "Condition" = "7d dox")
)

ratios[esacpe_set, ] %>%
  t() %>% data.frame() %>%
  add_column("Condition" = data_merged$Condition) %>%
  add_column("CellType" = data_merged$CellType) %>%
  pivot_longer(-c(Condition, CellType)) %>% 
  group_by(Condition, CellType, name) %>%
  summarize(mean = mean(value)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, names(escape_colors))) %>%
  dplyr::filter(Condition == "Astro_dox7_wo") %>%
  pivot_wider(names_from = CellType, values_from = mean) %>%
  dplyr::filter(!is.na(escape_status)) %>% {
    ggplot(data = ., aes(x = NPC, y = Astrocyte, fill = escape_status)) + 
      geom_point(pch = 21, size = 4) +
      theme_paper() + ylab("ASE (d-score)") + 
      scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = c("Silenced / Variable", "Facultative", "Constitutive")) + 
      labs(fill = "Escape Category") + xlab("NPC") + ylab("Astrocyte") + geom_abline() + 
      xlim(c(0, .8)) + ylim(c(0, .8))
      # ggrepel::geom_text_repel(aes(label = name), data = . %>% dplyr::filter(abs(NPC - Astrocyte) > 0.15)) + 
  }
ggsave("../Figures/FiguresAstrocytes/astro_npc_compare_7d_washout.pdf", width = 9, height = 6)

# ratios[consistent_escapees, ] %>%
#   t() %>% data.frame() %>%
#   add_column("Condition" = data_merged$Condition) %>%
#   add_column("CellType" = data_merged$CellType) %>%
#   pivot_longer(-c(Condition, CellType)) %>% 
#   group_by(Condition, CellType, name) %>%
#   summarize(mean = mean(value)) %>%
#   add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
#   mutate(escape_status = factor(escape_status, c("silenced / variable", "facultative", "constitutive"))) %>%
#   dplyr::filter(!is.na(escape_status)) %>%
#   dplyr::filter(grepl("dox7", Condition)) %>%
#   group_by(CellType, name) %>%
#   mutate(mean_baseline = mean[Condition == "Astro_no_dox7"]) %>%
#   mutate(mean_corrected = mean - mean_baseline) %>%
#   dplyr::filter(Condition != "Astro_no_dox7") %>% 
#   select(-c(mean, mean_baseline)) %>%
#   pivot_wider(names_from = CellType, values_from = mean_corrected) %>%
#   ggplot(aes(x = NPC, y = Astrocyte, fill = escape_status)) + 
#     geom_point(pch = 21, size = 4) + facet_wrap(~Condition) + 
#     theme_paper() + ylab("ASE (d-score)") + 
#     scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = c("Silenced / Variable", "Facultative", "Constitutive")) + 
#     labs(fill = "Escape Category") + xlab("") + coord_fixed() + geom_abline() + 
#     xlab("")

### test for differential AI 

# Finally, we do statistical tests for each gene between timepoints and baseline, to see how many genes are significantly reduced
run_tests <- function(general_escapees){
  data_test_inactive <- counts_inactive(general_escapees)
  data_test_active <- counts_active(general_escapees)
  
  test <- data_test_inactive / (data_test_active + data_test_inactive)
  
  fitting_metadata_here <- colData(general_escapees)
  
  all_coefs <- data.frame(do.call("rbind", lapply(1:nrow(data_test_active), function(i){
    tryCatch({
      y = as.numeric(data_test_inactive[i, ]) + 1
      N = as.numeric(data_test_active[i, ] + data_test_inactive[i, ]) + 2
      fit0 <- glm(cbind(y, N-y) ~ 1, family = "binomial", data = fitting_metadata_here)
      fit1 <- glm(cbind(y, N-y) ~ Condition, family = "binomial", data = fitting_metadata_here)
      coefs <- coef(fit1)
      effect_sizes <- c("control" = exp(coefs[[1]]) / (exp(coefs[[1]]) + 1), 
                        "dox" =  exp(coefs[[1]] + coefs[[2]]) / (exp(coefs[[1]] + coefs[[2]]) + 1))
      pvalue = coefficients(summary(fit1))[2,4]
      gene = rownames(data_test_active)[[i]]
      output = data.frame("gene" = gene, "Control" = effect_sizes[[1]], "Treatment" = effect_sizes[[2]], "pvalue" = pvalue)
      return(output)
    }, error = function(cond) {
      return(data.frame("gene" = gene, NA, NA, NA))
    }, warning = function(cond){
      return(c(NA))
    })
  })))
}

general_escape_genes <- plot_here %>% 
  dplyr::filter(Astrocyte > .1) %>%
  pull(name)

general_escape_genes <- consistent_escapees

data_here$Condition <- factor(data_here$Condition, levels = c("Astro_no_dox3", "Astro_no_dox7", "Astro_dox3", "Astro_dox7", 
                                                                   "Astro_no_dox3_wo", "Astro_no_dox7_wo", "Astro_dox3_wo", "Astro_dox7_wo"))

results_3d <- run_tests(data_here[general_escape_genes, data_here$ConditionClean %in% c("3d no dox", "3d dox")])
results_7d <- run_tests(data_here[general_escape_genes, data_here$ConditionClean %in% c("7d no dox", "7d dox")])

total_results <- rbind(
  cbind(results_3d, "Timepoint" = "3d"), 
  cbind(results_7d, "Timepoint" = "7d")
) %>% mutate(EscapeAnnotation = rowData(data_here[.$gene, ])$EscapeAnnotation)

data_text <- total_results %>%
  dplyr::mutate(hit_genes = p.adjust(pvalue) < 0.05 & Treatment / Control < 0.5) %>%
  group_by(Timepoint, EscapeAnnotation, hit_genes) %>%
  summarize(n_genes = n()) %>%
  mutate(hit_genes = ifelse(hit_genes, "n_genes", "total")) %>%
  pivot_wider(names_from = hit_genes, values_from = n_genes, values_fill = 0) %>%
  mutate(total = total + n_genes) %>%
  add_column(x = 1, y = rep(c(50, 48, 46), 2))
  
data_text_2 <- total_results %>%
  dplyr::mutate(hit_genes = Treatment > .1) %>%
  group_by(Timepoint, EscapeAnnotation, hit_genes) %>%
  summarize(n_genes = n()) %>%
  mutate(hit_genes = ifelse(hit_genes, "n_genes", "total")) %>%
  pivot_wider(names_from = hit_genes, values_from = n_genes, values_fill = 0) %>%
  mutate(total = total + n_genes) %>%
  add_column(x = 1, y = rep(c(50, 48, 46), 2))

total_results %>%
  dplyr::mutate(hit_genes = p.adjust(pvalue, method = "BH") < 0.05 & Treatment / Control < 0.5) %>%
  group_by(Timepoint, hit_genes) %>%
  summarize(n_genes = n()) %>%
  mutate(hit_genes = ifelse(hit_genes, "n_genes", "total")) %>%
  pivot_wider(names_from = hit_genes, values_from = n_genes, values_fill = 0) %>%
  mutate(total = total + n_genes)

data_text_3 <- total_results %>%
  dplyr::mutate(hit_genes = p.adjust(pvalue) < 0.05 & Treatment < .1) %>%
  group_by(Timepoint, hit_genes) %>%
  summarize(n_genes = n()) %>%
  mutate(hit_genes = ifelse(hit_genes, "n_genes", "total")) %>%
  pivot_wider(names_from = hit_genes, values_from = n_genes, values_fill = 0) %>%
  mutate(total = total + n_genes)

total_results -> df_plot_s5h
  
total_results %>% 
  ggplot(aes(x = Treatment / Control)) + geom_histogram(aes(fill = p.adjust(pvalue) < 0.05), col = "black", alpha = 0.5) + 
    facet_wrap(~factor(Timepoint, c("3d", "7d")),  nrow = 1) + 
    scale_fill_manual(values = c("black", "grey")) + theme_paper() + xlab("Treatment / Control (Allelic Ratio)") + 
    ylab("Number of genes") + 
    labs(fill = "Significant reduction \n (FDR < 5%, binomial model)") + geom_vline(xintercept = .5, linetype = 'dashed') + 
    scale_color_manual(values = escape_colors) + 
    annotate(x = 1, y = 55, label = "Number of genes \n with p < 0.05 & \n > 50% reduction:", geom = "text") + 
    geom_text(data = data_text, aes(x = x, y = y, col = EscapeAnnotation, label = paste0(n_genes, " / ", total))) + 
    annotate(x = 1, y = 35, label = "Number of genes \n residual escape > .1:", geom = "text") + 
    geom_text(data = data_text_2, aes(x = x, y = y - 20, col = EscapeAnnotation, label = paste0(n_genes, " / ", total)))
ggplot2::ggsave("../Figures/FiguresAstrocytes/astrocytes_number_affected_genes.pdf", width = 16, height = 8)

```

To make sure we observe substantial reduction in gene expression, we use DESeq2 to test for differences in escapee expression between Control and 7d. 
We do this either for combined, only inactive and only active counts: 

```{r }

data_here <- data[,data$Condition %in% c("Astro_no_dox7", "Astro_dox7")]
data_here$Replicates <- str_extract(colnames(data_here), "R[0-9]")
data_here <- data_here[unique(do.call("c", escapees_per_clone)), ]
data_here <- data_here[!rownames(data_here) %in% genes_out, ]
data_here$ConditionClean <- relevel(factor(data_here$Condition), ref = "Astro_no_dox7")

data_for_sf <- data_with_autosomes[,colnames(data_here)]

size_factors_global <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts(data_for_sf), colData = DataFrame(Cond = rep(1, 4)), design = ~0))
size_factors_x <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts_active(data_here) + counts_inactive(data_here), colData = DataFrame(Cond = rep(1, 4)), design = ~0))

deseq_inactive <- DESeqDataSetFromMatrix(countData = counts_inactive(data_here), colData = colData(data_here), design = ~ Replicates + ConditionClean)
deseq_inactive$sizeFactor <- size_factors_global$sizeFactor
deseq_inactive <- DESeq(deseq_inactive)

deseq_active <- DESeqDataSetFromMatrix(countData = counts_active(data_here), colData = colData(data_here), design = ~ Replicates + ConditionClean)
deseq_active$sizeFactor <- size_factors_global$sizeFactor
deseq_active <- DESeq(deseq_active)

deseq_total <- DESeqDataSetFromMatrix(countData = counts(data_here), colData = colData(data_here), design = ~ Replicates + ConditionClean)
deseq_total$sizeFactor <- size_factors_global$sizeFactor
deseq_total <- DESeq(deseq_total)

df_inactive <- results(deseq_inactive) %>%
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Status = "Inactive")

df_active <- results(deseq_active) %>%
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Status = "Active")

df_total <- results(deseq_total) %>%
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Status=  "Total")

rbind(df_inactive, rbind(df_active, df_total)) %>%
  ggplot(aes(x = baseMean, y = log2FoldChange, col = padj < 0.1)) + geom_point() + scale_x_log10() + 
    ggrepel::geom_text_repel(aes(label = Gene)) + xlab("Mean Expression (log normalized counts)") + ylab("log2 (7d dox / Control)") + 
    theme_paper() + facet_wrap(~Status) + scale_color_manual(values = c("black", "red"))
ggsave("../Figures/FiguresAstrocytes/astrocyte_escape_deseq_analysis.pdf", width = 9, height = 6)

# Visualize also as a heatmap: 

df_combined <- rbind(df_inactive, rbind(df_active, df_total))
df_combined$genomic_position <- start(rowRanges(data[df_combined$Gene, ]))

df_combined -> df_plot_s4e

df_combined %>% 
  ggplot(aes(x = 1, y = reorder(Gene, genomic_position), fill = log2FoldChange)) + geom_tile(col = "black") + facet_wrap(~Status) + 
    theme_paper() + scale_fill_gradient2() + geom_text(aes(label = ifelse(padj < 0.1, ".", "")), size = 10, nudge_y = 0.7) + 
    theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
ggsave("../Figures/FiguresAstrocytes/astrocyte_escape_deseq_analysis_heatmap.pdf", width = 9, height = 30)

```


```{r }

data_here <- data_with_autosomes[,data_with_autosomes$Condition %in% c("Astro_no_dox3", "Astro_dox3")]
data_here <- data_here[!rownames(data_here) %in% genes_out, ]
data_here$Condition <- relevel(factor(data_here$Condition), ref = "Astro_no_dox3")

data_for_sf <- data_with_autosomes[,colnames(data_here)]

size_factors_global <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts(data_for_sf), colData = DataFrame(Cond = rep(1, 4)), design = ~0))
size_factors_x <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts_active(data_here) + counts_inactive(data_here), colData = DataFrame(Cond = rep(1, 4)), design = ~0))

deseq_total <- DESeqDataSetFromMatrix(countData = counts(data_here), colData = colData(data_here), design = ~ Condition)
deseq_total$sizeFactor <- size_factors_global$sizeFactor
deseq_total <- DESeq(deseq_total)

df_total <- results(deseq_total) %>%
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Status=  "Total")

df_total %>%
  mutate(chr = as.character(seqnames(rowRanges(data_with_autosomes)[.$Gene]))) %>%
  dplyr::filter(chr != "X") %>%
  dplyr::filter(!is.na(padj)) %>%
  ggplot(aes(x = baseMean, y = log2FoldChange, col = padj < 0.1)) + geom_point() + scale_x_log10() + 
    xlab("Mean Expression (log normalized counts)") + ylab("log2 (3d dox / Control)") + 
    theme_paper() + scale_color_manual(values = c("black", "red")) + 
    labs(color = "Significant change \n (at 10% FDR)") + 
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggsave("../Figures/FiguresAstrocytes/astrocyte_autosomal_deseq_analysis_3d.pdf", width = 9, height = 6)

saveRDS(df_total, "../ProcessedData/autosomal_de_analysis_astrocytes_3d_dox.rds")

data_here <- data_with_autosomes[,data_with_autosomes$Condition %in% c("Astro_no_dox7", "Astro_dox7")]
data_here <- data_here[!rownames(data_here) %in% genes_out, ]
data_here$Condition <- relevel(factor(data_here$Condition), ref = "Astro_no_dox7")

data_for_sf <- data_with_autosomes[,colnames(data_here)]

size_factors_global <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts(data_for_sf), colData = DataFrame(Cond = rep(1, 4)), design = ~0))
size_factors_x <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts_active(data_here) + counts_inactive(data_here), colData = DataFrame(Cond = rep(1, 4)), design = ~0))

deseq_total <- DESeqDataSetFromMatrix(countData = counts(data_here), colData = colData(data_here), design = ~ Condition)
deseq_total$sizeFactor <- size_factors_global$sizeFactor
deseq_total <- DESeq(deseq_total)

df_total <- results(deseq_total) %>%
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Status=  "Total")

df_total %>%
  mutate(chr = as.character(seqnames(rowRanges(data_with_autosomes)[.$Gene]))) %>%
  dplyr::filter(chr != "X") %>%
  dplyr::filter(!is.na(padj)) -> df_plot_s3j

df_plot_s3j %>%
  ggplot(aes(x = baseMean, y = log2FoldChange, col = padj < 0.1)) + geom_point() + scale_x_log10() + 
    xlab("Mean Expression (log normalized counts)") + ylab("log2 (7d dox / Control)") + 
    theme_paper() + scale_color_manual(values = c("black", "red")) + 
    labs(color = "Significant change \n (at 10% FDR)") + 
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggsave("../Figures/FiguresAstrocytes/astrocyte_autosomal_deseq_analysis_7d.pdf", width = 9, height = 6)

saveRDS(df_total, "../ProcessedData/autosomal_de_analysis_astrocytes_7d_dox.rds")

# ggsave("./FiguresIllustrator/FigS1/escape_deseq_analysis.pdf", width = 9, height = 6)

# results(deseq_total) %>% data.frame() %>% rownames_to_column("Gene") %>%
#   mutate(chr = as.character(seqnames(rowRanges(data_with_autosomes)[.$Gene]))) %>%
#   arrange(padj) %>% 
#   dplyr::filter(chr != "X") %>%
# write.csv("./Figures/FiguresAstrocytes/results/autosomal_degs_7d_astrocytes.csv")

results(deseq_total) %>% data.frame() %>% rownames_to_column("Gene") %>%
  mutate(chr = as.character(seqnames(rowRanges(data_with_autosomes)[.$Gene]))) %>%
  arrange(padj) %>% 
  dplyr::filter(chr != "X")

results(deseq_total) %>% data.frame() %>% rownames_to_column("Gene") %>%
  mutate(chr = as.character(seqnames(rowRanges(data_with_autosomes)[.$Gene]))) %>% 
  mutate(start = as.numeric(start(rowRanges(data_with_autosomes)[.$Gene]))) %>%
  mutate(stop = as.numeric(end(rowRanges(data_with_autosomes)[.$Gene]))) -> res_test

res_test %>% 
  dplyr::filter(!is.na(padj)) %>%
  ggplot(aes(x = start, y = -log10(padj))) + geom_point() + facet_wrap(~chr)

calculate_gene_distances <- function(data, max_dist = 1e5){
  # calculate how often genes have another gene close by (atmost max_dist away)
  data %>% 
    group_by(chr) %>%
    mutate(
      close_gene = unlist(lapply(start, function(x){
        ifelse(any(start - x < max_dist & start - x > 0), "close", "far")
      }))
    ) %>%
    group_by(close_gene, .drop = F) %>% 
    summarize(n = n()) %>%
    pivot_wider(values_from = n, names_from = close_gene) %>%
    mutate(ratio = close / (close + far))
}

lapply(1:10, function(i){
  res_test %>% 
    dplyr::filter(chr == "X") %>%
    mutate(padj = sample(padj)) %>%
    dplyr::filter(padj < 0.1) %>%
    calculate_gene_distances() %>% 
    add_column(it = i)
}) %>% do.call("rbind", .) -> x_ratio_shuff 

lapply(1:10, function(i){
  res_test %>% 
    dplyr::filter(chr != "X") %>%
    mutate(padj = sample(padj)) %>%
    dplyr::filter(padj < 0.1) %>%
    calculate_gene_distances() %>% 
    add_column(it = i)
}) %>% do.call("rbind", .) -> auto_ratio_shuff

res_test %>% 
  dplyr::filter(chr == "X") %>%
  dplyr::filter(padj < 0.1) %>%
  calculate_gene_distances() %>% 
  add_column(it = 1) -> x_ratio

res_test %>% 
  dplyr::filter(chr != "X") %>%
  dplyr::filter(padj < 0.1) %>%
  calculate_gene_distances() %>%
  add_column(it = 1) -> auto_ratio

 list(
    cbind("cond" = "x_ratio_shuff", x_ratio_shuff),
    cbind("cond" = "auto_ratio_shuff", auto_ratio_shuff),
    cbind("cond" = "x_ratio", x_ratio),
    cbind("cond" = "auto_ratio", auto_ratio)
  ) %>% do.call("rbind", .) %>%
  mutate(condition = factor(cond, levels = c("x_ratio", "x_ratio_shuff", "auto_ratio", "auto_ratio_shuff"))) -> df_plot_s3k
 
df_plot_s3k %>%
  ggplot(aes(x = condition, ratio)) +
   stat_summary(geom = "bar") +
   stat_summary(geom = "errorbar", fun.data = function(x) {mean_sdl(x, mult = 1)}) + 
   ggbeeswarm::geom_quasirandom(size = 5, pch = 21, fill = "grey") + 
   coord_flip() + 
   xlab("") + ylab("Fraction of genes with DE neighbour") + theme_paper()
ggsave("../Figures/FiguresAstrocytes/astrocyte_degs_clustering.pdf", width = 9, height = 6)

```

```{r }

# data_npcs <- readRDS("./ProcessedData/for_joint_pca.rds")
# data_here <- data_with_autosomes[,data_with_autosomes$Condition %in% c("Astro_no_dox3", "Astro_dox3", "Astro_no_dox7", "Astro_dox7")]
# 
# join.genes <- intersect(rownames(data_npcs), rownames(data_here))
# data_npcs <- data_npcs[ join.genes , ]
# data_here <- data_here[ join.genes , ]
# colData(data_npcs) <- colData(data_npcs)[ , intersect(colnames(colData(data_npcs)), colnames(colData(data_here)))]
# colData(data_here) <- colData(data_here)[ , intersect(colnames(colData(data_npcs)), colnames(colData(data_here)))]
# assays(data_npcs) <- assays(data_npcs)[1]
# assays(data_here) <- assays(data_here)[1]
# reducedDims(data_npcs) <- list()
# reducedDims(data_here) <- list()
# mcols(data_npcs) <- mcols(data_npcs)[, 1:3]
# mcols(data_here) <- mcols(data_here)[, 1:3]
# 
# data_joint <- cbind(data_here, data_npcs)
# data_joint <- computeSumFactors(data_joint)
# data_joint <- logNormCounts(data_joint)
# data_joint <- runPCA(data_joint)
# 
# rotations_here <- attr(calculatePCA(data_joint), "rotation")
# 
# data.frame(
#   PC1 = reducedDims(data_joint)[["PCA"]][, 1],
#   PC2 = reducedDims(data_joint)[["PCA"]][, 2],
#   Condition = data_joint$ConditionClean,
#   Clone = data_joint$Clone
# ) %>% ggplot(aes(x = PC1, y = PC2, col = Condition, shape = Clone)) + geom_point(size = 4)
# ggsave("./Plots/FigsX_autosomal_joint_analysis/pca_with_astrocytes.pdf")

```

# mixing analysis: 
How can the authors rule out that the observed silencing effect does not simply occur within a small fraction of these dividing cells?
We believe that the data not consistent with that interpretation - silencing only occured in a small population (<50%) but was otherwise equivalent to NPCs, we would the effect to be smaller and uniformly scaled. Instead we observe substantial silencing and full silencing of a subset of genes.
We illustrate this argument with a simple simulation: 
We pick a control and treated (7d dox) sample from the astrocyte experiment and calculate reductions in allelic ratios (silencing of escape). 
We then pick a control and treated (7d dox) sample from the NPC experiment and simulate a treated sample in which only part of the cells show silencing by computing a weighted average between the treated and an independent NPC control sample, and compute the silencing effect (difference in ARs) to the other NPC control. 

```{r asdasdasda}

ratios %>%
  t() %>% data.frame() %>%
  add_column("Condition" = data_merged$Condition) %>%
  add_column("CellType" = data_merged$CellType) %>%
  pivot_longer(-c(Condition, CellType)) %>% 
  group_by(Condition, CellType, name) %>%
  summarize(mean = mean(value)) -> plot_here

colData(data_merged)

sim_data <- data.frame(
  ratios_astro_untreat = ratios[ , data_merged$CellType == "Astrocyte" & data_merged$Condition == "Astro_no_dox7"][ , 1],
  ratios_astro_treat = ratios[ , data_merged$CellType == "Astrocyte" & data_merged$Condition == "Astro_dox7"][ , 1], 
  ratios_npc_untreat = ratios[ , data_merged$CellType == "NPC" & data_merged$Condition == "Astro_no_dox7"][ , 1], 
  ratios_npc_untreat_mix = ratios[ , data_merged$CellType == "NPC" & data_merged$Condition == "Astro_no_dox7"][ , 2], 
  ratios_npc_treat = ratios[ , data_merged$CellType == "NPC" & data_merged$Condition == "Astro_dox7"][ , 1],
  ratios_npc_treat_mix = ratios[ , data_merged$CellType == "NPC" & data_merged$Condition == "Astro_dox7"][ , 2]
) %>%
  mutate(astro_diff = ratios_astro_untreat - ratios_astro_treat) %>%
  mutate(npc_diff = ratios_npc_untreat - ratios_npc_treat) %>%
  dplyr::filter(ratios_npc_untreat > .1 & ratios_astro_untreat > .1) %>%
  dplyr::filter(ratios_npc_untreat < .8 & ratios_astro_untreat < .8)

sim_data %>% 
  ggplot(aes(x = npc_diff, y = astro_diff)) + geom_point() + geom_abline(linetype = 'dashed')

sim_data %>% 
  dplyr::select(c("ratios_npc_untreat", "ratios_npc_treat", "ratios_astro_untreat", "ratios_astro_treat")) %>%
  pivot_longer(-c()) %>%
  ggplot(aes(x = value)) + facet_wrap(~name, ncol = 1) + geom_histogram()

for (i in c(0, 0.01, 0.05, 0.1, 0.2, 0.4)){
  simulated_aes = sim_data$ratios_astro_untreat * (1 - i) + sim_data$ratios_npc_treat_mix * i
  sim_data[ , paste0("ratios_npc_sim_", i)] = simulated_aes
}

sample_order <- c("ratios_npc_untreat", 
                  "ratios_npc_treat", 
                  "ratios_astro_untreat", 
                  "ratios_astro_treat", 
                  "ratios_npc_sim_0.4", 
                  "ratios_npc_sim_0.2", 
                  "ratios_npc_sim_0.1", 
                  "ratios_npc_sim_0.05", 
                  "ratios_npc_sim_0.01", 
                  "ratios_npc_sim_0")

group_colors <- setNames(
  c("grey", "darkred", "grey", "darkred", colorRampPalette(c("red", "yellow"))(5), "grey"),
  nm = sample_order)

sim_data %>% 
  dplyr::select(sample_order) %>%
  pivot_longer(-c()) %>%
  mutate(name = factor(name, levels = sample_order)) %>%
  mutate(group = case_when(
    grepl("_0", name) ~ "Simulated", 
    grepl("_astro", name) ~ "Astrocyte (Real data)", 
    .default = "NPC (Real data)"
  )) %>%
  mutate(group = factor(group, levels = c("NPC (Real data)", "Astrocyte (Real data)", "Simulated"))) -> df_plot_s5g
  
df_plot_s5g %>%
  ggplot(aes(x = value, col = name)) + geom_density(size = 2) + facet_wrap(~group, nrow = 3) + theme_bw(base_size = 20) + 
    xlab("Allelic Ratio") + ylab("Density") + 
    scale_color_manual(values = group_colors)
ggsave("~/Desktop/PhD/Projects/XChromosome_Antonia_Final/Plots/FigAstrocytes/simulation.pdf")
```

```{r asdasd}
# combine source data

source_data_2 <- list(df_plot_s3j, df_plot_s3k,
                      df_plot_s4e,
                      df_plot_2c, df_plot_2d, df_plot_2e,
                      df_plot_s5b, df_plot_s5c, df_plot_s5d, df_plot_s5e, df_plot_s5f, df_plot_s5g, df_plot_s5h)
saveRDS(source_data_2, "../ProcessedData/source_data/source_data_2.rds")

```
