---
title: "XistPaper_Figure02"
output: html_document
author: "Jasper Panten"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Desktop/Projects/XChromosome_Antonia/")
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

This script reproduces all plots for Figure 2 and associated supplementary figures in PaperName.
We first load the necessary libraries and some helper functions: 

```{r load_libraries, message=F}

setwd("~/Desktop/Projects/XChromosome_Antonia/")

library(tidyverse)
library(ggplot2)
library(scran)
library(scater)
library(DESeq2)
library(patchwork)

source("./Scripts/auxiliary.R")

# General theme for all plots: 
theme_paper <- function(){
  theme(panel.background = element_blank(),
        panel.grid = element_line(colour= "grey92"), 
        panel.grid.minor = element_line(size = rel(0.5)),
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1), 
        text = element_text(size = 20))
}

escape_colors <- setNames(c("darkgrey", "darkgreen", "orange"), nm = c("silenced / variable", "facultative", "constitutive"))
```

Now, read in the SCE, remove non-X genes and exclude lowly expressed genes

```{r load_data}

data <- readRDS("./ProcessedData/merged_dataset_paper.rds")
data_before_filtering <-  data[seqnames(data) == "X", ]

data <- data[(rowSums(counts_inactive(data)) + rowSums(counts_active(data))) / ncol(data) > 10, ]

data_with_autosomes <- data
data <- data[seqnames(data) == "X", ]

```


```{r asdasd}

data_here <- data[,data$ConditionClean %in% c("Control", "Dox (3d)", "Dox (7d)", 
                                              "Aux (2d)", "Aux (5d)", "Aux (9d)", 
                                              "Dox (3d), Aux (5d)", "Dox (7d), Aux (9d)")]
data_here <- data_here[,!data_here$Clone == "E6"]

# Here, we use CL30 / CL31 and their subclones as replicates
data_here$Clone <- gsub("\\.[0-9]*$", "", data_here$Clone)

# check that xist-overexpression works

rename_conditions <- c(
  "NA_NA" = "Control", 
  "Aux_NA" = "- Dox / + Aux", 
  "NA_Dox" = "+ Dox / - Aux", 
  "Aux_Dox" = "+ Dox / + Aux"
)

data.frame(
  Condition = factor(data_here$ConditionClean, levels = c("Control", "Aux (2d)", "Aux (5d)", "Aux (9d)", "Dox (3d)", 
                                                          "Dox (3d), Aux (5d)", "Dox (7d)", "Dox (7d), Aux (9d)")), 
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6), 
  Dox = data_here$ndDox, 
  Aux = pmax(as.numeric(data_here$ndAux) - 2, 0)
) %>%
  add_column(Timepoint = pmax(.$Dox, .$Aux)) %>%
  add_column(Condition2 = paste0(str_extract(.$Condition, "Aux"), "_", str_extract(.$Condition, "Dox"))) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(Timepoint = paste0(Timepoint, " days")) %>%
  mutate(Condition2 = factor(rename_conditions[.$Condition2], levels = rev(rename_conditions))) %>%
  #dplyr::filter(Timepoint == "7 days" & Condition == "NA_NA") %>%
  ggplot(aes(x = Condition2, y = Expression, col = Clone)) + geom_point() + 
  geom_line(aes(group = Clone), linetype = "dashed") + 
  coord_flip() + theme_paper() + 
    xlab("") + ggtitle("Xist Expression (normalized CPM)") + 
  facet_wrap(~Timepoint, ncol = 1, scales = "free_y") + 
  theme_paper()
ggsave("./FiguresIllustrator/FigS2/SPEN_Xist_by_sample.pdf")

# Main figure plot: 
size_factors <- colSums(counts(data_with_autosomes)) / colSums(counts(data_with_autosomes))[[1]]

data.frame(
  Condition = factor(data_here$ConditionClean, levels = c("Control", "Aux (2d)", "Aux (5d)", "Aux (9d)", "Dox (3d)", 
                                                          "Dox (3d), Aux (5d)", "Dox (7d)", "Dox (7d), Aux (9d)")), 
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6), 
  Dox = data_here$ndDox, 
  Aux = pmax(as.numeric(data_here$ndAux) - 2, 0)
) %>%
  add_column(Timepoint = pmax(.$Dox, .$Aux)) %>%
  add_column(Condition2 = paste0(str_extract(.$Condition, "Aux"), "_", str_extract(.$Condition, "Dox"))) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(Timepoint = paste0(Timepoint, " days")) %>%
  mutate(Condition2 = factor(rename_conditions[.$Condition2], levels = rev(rename_conditions))) %>%
  ggplot(aes(x = Condition2, y = Expression)) + 
    stat_summary(stat = "mean", geom = "bar", fill = "grey", col = "black") + 
    coord_flip() + theme_paper() + 
    xlab("") + ylab("Expression (CPM)") + geom_jitter(size = 3, width = 0.1) + 
    scale_y_continuous(expand = c(0, 0), limits = c(0, 320000)) + 
    facet_wrap(~Timepoint, ncol = 1, scales = "free_y") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggsave("./FiguresIllustrator/Fig2/SPEN_Xist_by_condition.pdf")

# check changes in escapee expression
ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))

genes_out <- names(rowMeans(ratios, na.rm = T)[rowMeans(ratios, na.rm = T) > 0.8 & rownames(ratios) != "Xist"])
genes_out_na <- rownames( ratios[!apply(ratios, 1, function(x){!any(is.na(x))}), ] )

genes_out <- c(genes_out, genes_out_na)

ratios <- ratios[!rownames(ratios) %in% genes_out, ]

# get escapees per clone
escapees <- apply(ratios, 2, function(x){x > 0.1})
escapees_per_clone <- apply(escapees[,data_here$ConditionClean == "Control"], 2, function(x){rownames(escapees)[x]})
escapees_per_clone <- lapply(escapees_per_clone, function(x){x[!x %in% c("Xist", "Tsix")]})
escapees_per_clone <- lapply(escapees_per_clone, function(x){x[!is.na(x)]})
names(escapees_per_clone) <- c("CL30 - December2021", "CL31 - December2021", "CL31 - October2022", "CL30 - October2022")

# new plot into timepoints
ratios %>%
  t() %>% data.frame() %>%
  add_column(Dox = data_here$ndDox) %>%
  add_column(Aux = pmax(as.numeric(data_here$ndAux) - 2, 0)) %>%
  add_column(Clone = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  add_column(Condition = data_here$ConditionClean) %>%
  pivot_longer(-c(Dox, Aux, Clone, Condition)) %>%
  group_by(Clone) %>%
  dplyr::filter(name %in% escapees_per_clone[[unique(Clone)]]) %>%
  ungroup() %>%
  add_column(Timepoint = pmax(.$Dox, .$Aux)) %>%
  add_column(Condition2 = paste0(str_extract(.$Condition, "Aux"), "_", str_extract(.$Condition, "Dox"))) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(Timepoint = paste0(Timepoint, " days")) %>%
  mutate(Condition2 = factor(rename_conditions[.$Condition2], levels = rename_conditions)) %>%
  ggplot(aes(x = Condition2, y = value, fill = Clone)) + geom_boxplot(col = "grey") + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    facet_wrap(~Timepoint, scale = "free_x") + 
    xlab("") + theme_paper()
ggsave("./FiguresIllustrator/FigS2/SPEN_escape_by_sample.pdf")

joint_escapees <- Reduce("intersect", escapees_per_clone)

# Main figure plots: 
ratios[joint_escapees, ] %>%
  t() %>% data.frame() %>%
  add_column(Dox = data_here$ndDox) %>%
  add_column(Aux = pmax(as.numeric(data_here$ndAux) - 2, 0)) %>%
  add_column(Clone = data_here$Clone) %>%
  add_column(Condition = data_here$ConditionClean) %>%
  pivot_longer(-c(Dox, Aux, Clone, Condition)) %>%
  group_by(Clone) %>%
  #dplyr::filter(name %in% escapees_per_clone[[unique(Clone)]]) %>%
  ungroup() %>%
  add_column(Timepoint = pmax(.$Dox, .$Aux)) %>%
  add_column(Condition2 = paste0(str_extract(.$Condition, "Aux"), "_", str_extract(.$Condition, "Dox"))) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(Timepoint = paste0(Timepoint, " days")) %>%
  mutate(Condition2 = factor(rename_conditions[.$Condition2], levels = rename_conditions)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  mutate(escape_status = factor(escape_status, levels = c("silenced / variable", "facultative", "constitutive"))) %>%
  ggplot(aes(x = Condition2, y = value, fill = escape_status)) + geom_boxplot(col = "black") + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    facet_wrap(~Timepoint, scale = "free_x") + 
    scale_fill_manual(values = escape_colors) + 
    xlab("") + theme_paper()
ggsave("./FiguresIllustrator/Fig2/SPEN_escape_by_category_boxplot.pdf", width = 8, height = 6)

# ratios %>%
#   t() %>% data.frame() %>%
#   add_column(Dox = data_here$ndDox) %>%
#   add_column(Aux = pmax(as.numeric(data_here$ndAux) - 2, 0)) %>%
#   add_column(Clone = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
#   add_column(Condition = data_here$ConditionClean) %>%
#   pivot_longer(-c(Dox, Aux, Clone, Condition)) %>%
#   group_by(Clone) %>%
#   dplyr::filter(name %in% escapees_per_clone[[unique(Clone)]]) %>%
#   ungroup() %>%
#   add_column(Timepoint = pmax(.$Dox, .$Aux)) %>%
#   add_column(Condition2 = paste0(str_extract(.$Condition, "Aux"), "_", str_extract(.$Condition, "Dox"))) %>%
#   add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7")), .$Dox)) %>%
#   unnest(TimeSeries) %>%
#   mutate(Timepoint = paste0(Timepoint, " days")) %>%
#   mutate(Condition2 = factor(rename_conditions[.$Condition2], levels = rename_conditions)) %>%
#   add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
#   dplyr::filter(!is.na(escape_status)) %>%
#   mutate(escape_status = factor(escape_status, levels = c("silenced / variable", "facultative", "constitutive"))) %>%
#   ggplot(aes(x = Condition2, y = value, fill = escape_status)) + geom_boxplot(col = "black") + 
#     theme_paper() + ylab("ASE (d-score)") + 
#     theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
#     facet_wrap(~Clone + Timepoint, scale = "free_x") + 
#     scale_fill_manual(values = escape_colors) + 
#     xlab("") + theme_paper()
# ggsave("~/Desktop/Projects/XChromosome_Antonia/Figures_new/FigS2/SPEN_escape_by_category_boxplot_Clone.pdf")

# To structure the data, we order the genes using hierarchical clustering

cl30_escapees <- c(Reduce("intersect", escapees_per_clone), "Xist")

levels_use <- c("Control", "Dox (3d)", "Dox (3d), Aux (5d)", "Dox (7d)", "Dox (7d), Aux (9d)")
data_for_clustering <- do.call("rbind", lapply(unique(data_here$ConditionClean), function(x){
  if (sum(data_here$ConditionClean == x) == 1){
    ratios[cl30_escapees, data_here$ConditionClean == x]
  } else {
    rowMeans(ratios[cl30_escapees, data_here$ConditionClean == x])
  }
}))

clustering_order <- setNames(hclust(dist(t(data_for_clustering)))$order, rownames(ratios[cl30_escapees, ]))

ratios[cl30_escapees, ][clustering_order, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Condition, Clone)) %>%
  mutate(name = factor(name, levels = unique(.$name))) %>%
  dplyr::filter(Condition %in% levels_use) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  mutate(Condition = factor(Condition, levels = rev(levels_use))) -> to_plot

###
p1 <- to_plot %>%
  ggplot(aes(name, y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    labs(fill = "d-score") + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 8)) + 
    scale_x_discrete(position = "top") + guides(colour = "none")

# Plot escapee annotation at the bottom of this: 
p2 <- data.frame(Gene = factor(cl30_escapees, levels = levels(to_plot$name)),  Annotation = rowData(data)[cl30_escapees, ]$EscapeAnnotation) %>%
  ggplot(aes(x = Gene, y = 1, fill = Annotation)) + theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) + 
  theme(legend.position = "None")

p1 / p2 + plot_layout(heights = c(4, 0.2))
ggsave("./FiguresIllustrator/Fig2/heatmap_spen_only_escapees.pdf", width = 13, height = 3.5)

### 
p1 <- to_plot %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    labs(fill = "d-score") + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 8)) + 
    scale_x_discrete(position = "top") + guides(colour = "none")

# Plot escapee annotation at the bottom of this: 
p2 <- data.frame(Gene = factor(cl30_escapees, levels = levels(to_plot$name)),  Annotation = rowData(data)[cl30_escapees, ]$EscapeAnnotation) %>%
  ggplot(aes(x = Gene, y = 1, fill = Annotation)) + theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) + 
  theme(legend.position = "None")

p1 / p2 + plot_layout(heights = c(4, 0.2))
ggsave("./FiguresIllustrator/Fig2/heatmap_spen_only_escapees_by_position.pdf", width = 13, height = 3.5)

#### LM stuff for SPEN-experiment
general_escape_genes <- Reduce("intersect", escapees_per_clone)
general_escapees <- data_here[general_escape_genes, ]

data_test_inactive <- counts_inactive(general_escapees)
data_test_active <- counts_active(general_escapees)

fitting_metadata_here <- colData(general_escapees)

genes_out <- names(which(rowMeans(data_test_inactive / (data_test_active + data_test_inactive)) > 0.8))
data_test_inactive <- data_test_inactive[!rownames(data_test_inactive) %in% genes_out, ]
data_test_active <- data_test_active[!rownames(data_test_active) %in% genes_out, ]

all_coefs <- data.frame(do.call("rbind", lapply(1:nrow(data_test_active), function(i){
  tryCatch({
    y = as.numeric(data_test_inactive[i, ]) + 1
    N = as.numeric(data_test_active[i, ] + data_test_inactive[i, ]) + 2
    fit1 <- glm(cbind(y, N-y) ~ ndDox * ndAux, family = "binomial", data = fitting_metadata_here)
    coefs <- coef(fit1)
    coefs
    return(coefs)
  }, error = function(cond) {
    return(c(NA))
  }, warning = function(cond){
    return(c(NA))
  })
})))

rownames(all_coefs) <- rownames(data_test_active)
all_coefs <- all_coefs[!apply(all_coefs, 1, function(x){all(is.na(x))}), !apply(all_coefs, 2, function(x){all(is.na(x))})]

# save coefficients for downstream analysis
#saveRDS(all_coefs, "./ProcessedData/LM_coefficients_E6.rds")

row_annotation <- data.frame(
  Gene = rownames(all_coefs), 
  escapee_annotation = as.character(rowData(data_here)[match(rownames(all_coefs), rowData(data_here)$gene_name), 7])
) %>% column_to_rownames("Gene")

all_coefs %>%
  rownames_to_column("Gene") %>%
  pivot_longer(-c(Gene)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(x = name, y = value, col = escape_status)) + geom_boxplot() + 
    geom_jitter(position = position_jitterdodge(jitter.width = 0.1), alpha = 0.2) +
    coord_flip() + theme_paper() + xlab("Covariate") +
    theme(aspect.ratio = 2) + theme(legend.position = "top") + geom_hline(yintercept = 0, linetype = "dashed") + 
    ylab("Model coefficient") + xlab("") + 
    scale_color_manual(values = escape_colors)
ggsave("./FiguresIllustrator/FigS2/SPEN_escape_by_category_boxplot.pdf")

# We now convert the coefficients to fold changes in escape: 

df_test <- data.frame(
  Gene = rownames(all_coefs), 
  control = exp(all_coefs$X.Intercept.) / (exp(all_coefs$X.Intercept.) + 1), 
  nddox3_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox3) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox3) + 1), 
  nddox7_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox7) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox7) + 1), 
  nddox3_aux = exp(all_coefs$X.Intercept. + all_coefs$ndAux5 + all_coefs$ndDox3 + all_coefs$ndDox3.ndAux5) / 
    (exp(all_coefs$X.Intercept. + all_coefs$ndAux5 + all_coefs$ndDox3 + all_coefs$ndDox3.ndAux5) + 1), 
  nddox7_aux = exp(all_coefs$X.Intercept. + all_coefs$ndAux9 + all_coefs$ndDox7 + all_coefs$ndDox7.ndAux9) / 
    (exp(all_coefs$X.Intercept. + all_coefs$ndAux9 + all_coefs$ndDox7 + all_coefs$ndDox7.ndAux9) + 1)
)

df_test %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(control - nddox3_treat, nddox3_aux - nddox3_treat, col = escape_status)) + geom_point() + 
    geom_smooth(method = "lm", col = "black", linetype = 'dashed', size = 0.5) + 
    ggrepel::geom_text_repel(aes(label = Gene), size = 5) + 
    geom_abline(linetype = "dashed", slope = 1) + 
    theme_paper() + 
    xlab("Dox-effect \n (d-score difference Control - Dox)") + 
    ylab("SPEN-effect \n (d-score difference  Dox - Dox + SPEN)") + 
    ggtitle("3d analysis") + 
    xlim(c(-0.1, 0.5)) + ylim(c(-0.1, 0.5)) + 
    coord_fixed() + 
    scale_color_manual(values = escape_colors)
ggsave("./FiguresIllustrator/Fig2/spen_vs_dox_model_3d.pdf", width = 8, height = 6)

df_test %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(control - nddox7_treat, nddox7_aux - nddox7_treat, col = escape_status)) + geom_point() + 
    geom_smooth(method = "lm", col = "black", linetype = 'dashed', size = 0.5) + 
    ggrepel::geom_text_repel(aes(label = Gene), size = 5) + 
    geom_abline(linetype = "dashed", slope = 1) + 
    theme_paper() + 
    xlab("Dox-effect \n (d-score difference Control - Dox)") + 
    ylab("SPEN-effect \n (d-score difference  Dox - Dox + SPEN)") + 
    ggtitle("7d analysis") + 
    xlim(c(-0.1, 0.5)) + ylim(c(-0.1, 0.5)) + 
    coord_fixed() + 
    scale_color_manual(values = escape_colors)
ggsave("./FiguresIllustrator/Fig2/spen_vs_dox_model_7d.pdf", width = 8, height = 6)

# Check how the model coefficients fit with allelic ratio differences: 
# 
# inv_logit <- function(x){1 / (1 + exp(-x))}
# odds_ratio <- function(x, y){
#   log((x / (1 - x)) / (y / (1 - y)))
# }
# 
# df_test <- data.frame(
#   nddox3_control_obs = rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Control"]), 
#   nddox3_treat_obs = rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Dox (3d)"]), 
#   nddox3_control = exp(all_coefs$X.Intercept.) / (exp(all_coefs$X.Intercept.) + 1), 
#   nddox3_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox3) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox3) + 1)
# )
# 
# df_test %>%
#   ggplot(aes(x = nddox3_treat_obs, nddox3_treat)) + geom_point() + geom_abline()
# 
# all_coefs$d_score_3d <- odds_ratio(rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Dox (3d)"]), 
#   rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Control"]))
# 
# all_coefs$d_score_7d <- odds_ratio(rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Dox (7d)"]), 
#   rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Control"]))
# 
# all_coefs %>% ggplot(aes(x = d_score_3d, ndDox3)) + geom_point() + geom_abline()
# 
# all_coefs %>% ggplot(aes(x = d_score_7d, ndDox7)) + geom_point() + geom_abline()

NULL

plot_gene <- function(gene, save = F, path = NULL, suffix = "_escape_plot.pdf"){
  i = gene
  y = as.numeric(data_test_inactive[i, ])
  N = as.numeric(data_test_active[i, ] + data_test_inactive[i, ])
  fit1 <- glm(cbind(y, N-y) ~ Clone + ndDox + ndAux, family = "binomial", data = fitting_metadata_here)
  fit0 <- glm(cbind(y, N-y) ~ 1, family = "binomial", data = fitting_metadata_here)
  
  pp <- data.frame(
    y = y,
    N = N,
    Clone = paste0(data_here$Clone, " - ", data_here$Experiment),
    Cov = factor(data_here$ConditionClean, levels = c("Control", "Aux (2d)", "Aux (5d)", "Aux (9d)", "Dox (3d)", "Dox (3d), Aux (5d)", "Dox (7d)", "Dox (7d), Aux (9d)"))
  ) %>% ggplot(aes(Cov, y / N, col = Clone)) + geom_point(size = 4) + coord_flip() + ylim(0, 1) +
    ggtitle(i) +
    geom_hline(yintercept = rev_logit(coef(fit1)[[1]]), linetype = "dashed") + 
    geom_line(aes(group = Clone), linetype = "dashed") + 
    theme_paper() + xlab("") + ylab("ASE (d-score)")

  if (save){
    ggsave(paste0(path, "./", gene, suffix), pp)
  } 
  
  return(pp)
}

plot_gene("G6pdx", save = F, "~/Desktop/Projects/XChromosome_Antonia/Figures_new/Fig2/", "_escape_all_clonesAux.pdf")
plot_gene("Eif2s3x", save = F, "~/Desktop/Projects/XChromosome_Antonia/Figures_new/Fig2/", "_escape_all_clonesAux.pdf")
# plot_gene("Plxnb3", save = F, "~/Desktop/Projects/XChromosome_Antonia/Figures_new/Fig2/", "_escape_all_clonesAux.pdf")
# plot_gene("Ddx3x", save = F, "~/Desktop/Projects/XChromosome_Antonia/Figures_new/Fig2/", "_escape_all_clonesAux.pdf")
# plot_gene("Gm14634", save = F, "~/Desktop/Projects/XChromosome_Antonia/Figures_new/Fig2/", "_escape_all_clonesAux.pdf")

# Quantify silencing and lack-of-silencing effects in two timepoints: 

ratios_escapees <- ratios[rownames(general_escapees), ]

data.frame(
  Gene = rownames(ratios_escapees), 
  Control = rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
  dox_3d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d)"]), 
  dox_3d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d), Aux (5d)"]), 
  escape_category = rowData(data_here[rownames(ratios_escapees), ])$EscapeAnnotation
) %>%
  ggplot(aes(x = Control, y = dox_3d, col = escape_category)) + geom_point() + geom_abline() + ggrepel::geom_text_repel(aes(label = Gene)) + 
    scale_color_manual(values = escape_colors) + theme_paper() + geom_smooth(method = "lm", linetype = 'dashed', color = "darkblue")

data.frame(
  Gene = rownames(ratios_escapees), 
  Control = rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
  dox_3d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d)"]), 
  dox_3d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d), Aux (5d)"]), 
  escape_category = rowData(data_here[rownames(ratios_escapees), ])$EscapeAnnotation
) %>%
  ggplot(aes(x = Control, y = dox_3d_aux, col = escape_category)) + geom_point() + geom_abline() + ggrepel::geom_text_repel(aes(label = Gene)) + 
    scale_color_manual(values = escape_colors) + theme_paper() + geom_smooth(method = "lm", linetype = 'dashed', color = "darkblue")

data.frame(
  Gene = rownames(ratios_escapees), 
  Control = rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
  dox_7d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d)"]), 
  dox_7d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7dd), Aux (9d)"]), 
  escape_category = rowData(data_here[rownames(ratios_escapees), ])$EscapeAnnotation
) %>%
  ggplot(aes(x = Control, y = dox_7d, col = escape_category)) + geom_point() + geom_abline() + ggrepel::geom_text_repel(aes(label = Gene)) + 
    scale_color_manual(values = escape_colors) + theme_paper() + geom_smooth(method = "lm", linetype = 'dashed', color = "darkblue")

data.frame(
  Gene = rownames(ratios_escapees), 
  Control = rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
  dox_7d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d)"]), 
  dox_7d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d), Aux (9d)"]), 
  escape_category = rowData(data_here[rownames(ratios_escapees), ])$EscapeAnnotation
) %>%
  ggplot(aes(x = Control, y = dox_7d_aux, col = escape_category)) + geom_point() + geom_abline() + ggrepel::geom_text_repel(aes(label = Gene)) + 
    scale_color_manual(values = escape_colors) + theme_paper() + geom_smooth(method = "lm", linetype = 'dashed', color = "darkblue")

genes_highlight = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d), Aux (9d)"]) - rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"])
genes_highlight <- names(genes_highlight[genes_highlight < -0.08])

dff <- data.frame(
  Gene = rownames(ratios_escapees), 
  dox_3d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d)"]) - rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
  dox_3d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d), Aux (5d)"]) - rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
  dox_7d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d)"]) - rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
  dox_7d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d), Aux (9d)"]) - rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]) 
)

dff %>% pivot_longer(-c(Gene)) %>%
  ggplot(aes(x = name, y = value)) + geom_boxplot(width = 0.1, outlier.color = NA) + ggbeeswarm::geom_quasirandom() + theme_paper() + 
    ggrepel::geom_text_repel(aes(label = Gene), data = dff[dff$Gene %in% genes_highlight, ] %>% pivot_longer(-c(Gene)))
  
# When all replicates are here, run tests and ask if anything is significant

# compute p-values associated with silencing and non-silencing
# samples_here <- c("Control", "Dox (3d)", "Dox (7d)")
# data_test_inactive_here <- data_test_inactive[,data_here$ConditionClean %in% samples_here]
# data_test_active_here <- data_test_active[,data_here$ConditionClean %in% samples_here]
# fitting_metadata_here_here <- fitting_metadata_here[data_here$ConditionClean %in% samples_here, ]
# 
# pvals_dox <- data.frame(do.call("rbind", lapply(1:nrow(data_test_inactive_here), function(i){
#   tryCatch({
#     y = as.numeric(data_test_inactive_here[i, ]) + 1
#     N = as.numeric(data_test_active_here[i, ] + data_test_inactive_here[i, ]) + 2
#     fit0 <- glm(cbind(y, N-y) ~ 1, family = "binomial", data = fitting_metadata_here_here)
#     fit1 <- glm(cbind(y, N-y) ~ ndDox, family = "binomial", data = fitting_metadata_here_here)
#     pval <- lmtest::lrtest(fit0, fit1)$`Pr(>Chisq)`[[2]]
#     return("pval" = pval)
#   }, error = function(cond) {
#     return("pval" = NA)
#   }, warning = function(cond){
#     return("pval" = NA)
#   })
# })))
# 
# samples_here <- c("Control", "Dox (3d)", "Dox (7d)", "Aux (5d)", "Aux (9d)", "Dox (3d), Aux (5d)", "Dox (7d), Aux (9d)")
# data_test_inactive_here <- data_test_inactive[,data_here$ConditionClean %in% samples_here]
# data_test_active_here <- data_test_active[,data_here$ConditionClean %in% samples_here]
# fitting_metadata_here_here <- fitting_metadata_here[data_here$ConditionClean %in% samples_here, ]
# 
# pvals_aux <- data.frame(do.call("rbind", lapply(1:nrow(data_test_inactive_here), function(i){
#   tryCatch({
#     y = as.numeric(data_test_inactive_here[i, ]) + 1
#     N = as.numeric(data_test_active_here[i, ] + data_test_inactive_here[i, ]) + 2
#     fit0 <- glm(cbind(y, N-y) ~ ndDox + ndAux, family = "binomial", data = fitting_metadata_here_here)
#     fit1 <- glm(cbind(y, N-y) ~ ndDox * ndAux, family = "binomial", data = fitting_metadata_here_here)
#     pval <- lmtest::lrtest(fit0, fit1)$`Pr(>Chisq)`[[2]]
#     return("pval" = pval)
#   }, error = function(cond) {
#     return("pval" = NA)
#   }, warning = function(cond){
#     return("pval" = NA)
#   })
# })))


```
