---
title: "XistPaper_Figure03_singleClone"
output: html_document
author: "Jasper Panten"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/../../"))
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

This script reproduces all plots for Figure 3 and associated supplementary figures in PaperName.
We first load the necessary libraries and some helper functions: 

```{r load_libraries, message=F}

setwd(paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/../../../"))

library(tidyverse)
library(ggplot2)
library(scran)
library(scater)
library(DESeq2)
library(patchwork)

source("./xist_project/Scripts/General/auxiliary.R")
source("./xist_project/Scripts/General/ase_functions.R")

escape_colors <- setNames(c("darkgrey", "darkgreen", "orange"), nm = c("NPC-specific", "Facultative", "Constitutive"))

```

Now, read in the SCE, remove non-X genes and exclude lowly expressed genes

```{r load_data}

data <- readRDS("./ProcessedData/merged_dataset_paper.rds")
data_before_filtering <-  data[seqnames(data) == "X", ]

genes_keep <- readRDS("./ProcessedData/genes_keep.rds")
data <- data[genes_keep, ]

data_with_autosomes <- data
data <- data[seqnames(data) == "X", ]

# For this figure, we have two experiments which we need to consider seperately because the results look quite different
# data$ExperimentSet <- ifelse(data$Experiment %in% c("September2022"), "Rep1", "Rep2")

```


```{r ssss}

### This is for Figure 3, the reversibility aspect
### Now we look at Xist overexpression and how that silences escapees
samples_use <- c("Control", 
                 "Dox (3d)", "Dox (3d) - washout (4d)", 
                 
                 "Dox (7d)", 
                 "Dox (7d) - washout (4d)", 
                 "Dox (7d) - washout (7d)", 
                 "Dox (7d) - washout (14d)", 
                 
                 "Dox (14d)", 
                 "Dox (14d) - washout (7d)", 
                 "Dox (14d) - washout (14d)", 
                 
                 "Dox (21d)", 
                 "Dox (21d) - washout (7d)", 
                 "Dox (21d) - washout (14d)"
)

data_here <- data[,data$ConditionClean %in% samples_use]
data_here <- data_here[,data_here$Clone == "E6"]

ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))
ratios <- ratios[apply(ratios, 1, function(x){!any(is.na(x))}), ]

# genes_out <- names(rowMeans(ratios[,data_here$ConditionClean == "Control"], na.rm = T)[rowMeans(ratios, na.rm = T) > 0.8 & rownames(ratios) != "Xist"])
# genes_out <- names((ratios[,data_here$ConditionClean == "Control"])[(ratios) > 0.8 & rownames(ratios) != "Xist"])

# ratios <- ratios[!rownames(ratios) %in% genes_out, ]

# define escapees
escapees <- apply(ratios[,data_here$ConditionClean == "Control"], 2, function(x){x > 0.1})
# escapees <- ratios[,data_here$ConditionClean == "Control"] > 0.1
escapees_per_clone <- apply(escapees, 2, function(x){rownames(escapees)[x]})
escapees_per_clone <- lapply(escapees_per_clone, function(x){x[x != "Xist"]})
# escapees_per_clone <- rownames(data_here)[escapees]
# escapees_per_clone <- unlist(lapply(escapees_per_clone, function(x){x[x != "Xist"]}))
escapees_total <- Reduce("intersect", escapees_per_clone)
# escapees_total <- escapees_per_clone

# For Xist-expression, we need a scaling factor per sample to account for differences in seq depth
size_factors <- colSums(counts(data_here)) / colSums(counts(data_here))[[1]]

# # check that Xist-overexpression works
# data.frame(
#   Dox = data_here$ndDox,
#   Washout = data_here$ndWashout, 
#   Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
#   Expression = as.numeric(counts(data_here["Xist", ]) / size_factors)
# ) %>%
#   add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7", "14", "21")), .$Dox)) %>%
#   unnest(TimeSeries) %>%
#   mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>%
#   add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
#   mutate(Condition = factor(Condition, levels = rev(c("Control", "0 days washout", "4 days washout", "7 days washout", "14 days washout")))) %>%
#   mutate(TimeSeries = factor(TimeSeries, levels = c("3 days dox treatment", "7 days dox treatment", "14 days dox treatment", "21 days dox treatment"))) %>%
#   ggplot(aes(x = Condition, y = Expression)) + geom_boxplot() + geom_point() + coord_flip() + 
#     geom_line(aes(group = Clone, linetype = Clone, col = Clone)) + theme_paper() + 
#     xlab("") + ylab("Xist Expression (normalized CPM)") + facet_wrap(~TimeSeries, nrow = 2) + 
#     NULL
# ggsave("./Plots/FigS3/Washout_Xist_boxplots.pdf")

# Supplement
# we normalize this on the median control expression: 
xist_expression_levels <- data.frame(
  Dox = data_here$ndDox,
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6)
)

mean_baseline_xist <- xist_expression_levels %>% dplyr::filter(Dox == 0) %>% pull(Expression) %>% mean()

data.frame(
  Dox = data_here$ndDox,
  Washout = data_here$ndWashout, 
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ]) / size_factors) / mean_baseline_xist
) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7", "14", "21")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>%
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
  mutate(Condition = factor(Condition, levels = rev(c("Control", "0 days washout", "4 days washout", "7 days washout", "14 days washout")))) %>%
  mutate(TimeSeries = factor(TimeSeries, levels = c("3 days dox treatment", "7 days dox treatment", "14 days dox treatment", "21 days dox treatment"))) %>%
  ggplot(aes(x = Condition, y = Expression, col = grepl("May2025", Clone))) + 
    stat_summary(stat = "mean", geom = "bar", fill = "grey", col = "black") + 
    coord_flip() + theme_paper() + 
    xlab("") + ylab("Expression (CPM)") + geom_jitter(size = 3, width = 0.1) + 
    scale_y_continuous(expand = c(0, 0), limits = c(0, 30)) + 
    facet_wrap(~TimeSeries, ncol = 1, scales = "free_y") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggsave("./Plots/FigS3_Rep2/Washout_Xist_barplots.pdf", height = 8, width = 6)

# pvalue for xist
xist_plot_data %>% 
  group_by(Condition) %>%
  summarize(Expression = list(Expression)) %>%
  mutate(Expression_Control = Expression[Condition == "Control"]) %>%
  rowwise() %>%
  dplyr::filter(Condition != "Control") %>%
  mutate(pval = t.test(Expression_Control, Expression)$p.value) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pval, method = "BH"))

```

```{r }

genes_show <- c("Xist", "Mecp2", "Kdm6a", "Kdm5c")
genes_of_choice <- setNames(rep("", nrow(ratios)), rownames(ratios))
genes_of_choice[genes_show] <- genes_show

ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Condition, Clone, Dox)) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  mutate(Condition = factor(Condition, levels = rev(samples_use))) %>%
  mutate(TimePoint = case_when(
    grepl("Dox \\(3d\\)", Condition)  ~ "3d",
    grepl("Dox \\(7d\\)", Condition)  ~ "7d",
    grepl("Dox \\(14d\\)", Condition)  ~ "14d",
    grepl("Dox \\(21d\\)", Condition)  ~ "21d",
    grepl("Control", Condition)  ~ "0d")) %>% 
    mutate(TimePoint = factor(TimePoint, levels = paste0(c(0, 3, 7, 14, 21), "d"))) -> to_plot

### We see more reversibility in the 7d washout than 14d, which is odd
to_plot %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_y_discrete(labels = genes_of_choice, position = "right") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    scale_x_discrete(labels = genes_of_choice, position = "bottom") + 
    labs(fill = "d-score") +  facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.y = element_blank())
ggsave("./Plots/FigS3/Washout_full_heatmap.pdf")

### This is due to the fact that we have 2 replicates in the 7d condition, one of which shows much more escape than the other. 
### In the + dox condition this is not the case
to_plot %>%
  ggplot(aes(reorder(name, position), y = reorder(paste0(Condition, "_", Clone), Condition), fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_y_discrete(labels = genes_of_choice, position = "right") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    scale_x_discrete(labels = genes_of_choice, position = "bottom") +
    labs(fill = "d-score") + facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") +
    theme(strip.background = element_blank(), strip.text.y = element_blank())
ggsave("./Plots/FigS3/Washout_full_heatmap_allsamples.pdf")

### Visualize this nicely: 

# to_plot %>%
#   dplyr::filter(TimePoint %in% c("0d", "14d", "21d")) %>%
#   # mutate(Replicate = case_when(
#   #   grepl("September2022", Clone) ~ "Rep1", 
#   #   grepl("April2023|October2022", Clone) ~ "Rep2", 
#   #   .default = "Rep2"
#   # ))
#   mutate(Clone = case_when(
#     grepl("September2022", Clone) ~ gsub("September2022", "Rep1", Clone),
#     .default = gsub("April2023|October2022|October2021|March2022", "Rep2", Clone)
#   )) %>%
#   ggplot(aes(reorder(name, position), y = reorder(paste0(Condition, "_", Clone), Condition), fill = value, col = value)) + geom_tile(width = 0.9) +
#     scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
#     scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
#     scale_y_discrete(labels = genes_of_choice, position = "right") +
#     theme_paper() +
#     theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
#     xlab("") +
#     ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
#     scale_x_discrete(labels = genes_of_choice, position = "bottom") +
#     labs(fill = "Allelic Ratio (Xi / (Xi + Xa))", col = "") + guides(colour = "none") + 
#     facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") +
#     theme(strip.background = element_blank(), strip.text.y = element_blank())
# ggsave("./Plots/FigS3/Washout_full_heatmap_long_focussed.pdf")

####

to_plot %>%
  dplyr::filter(TimePoint %in% c("0d", "14d", "21d")) %>%
  ggplot(aes(reorder(name, position), y = reorder(paste0(Condition, "_", Clone), Condition), fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_y_discrete(labels = genes_of_choice, position = "right") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    scale_x_discrete(labels = genes_of_choice, position = "bottom") +
    labs(fill = "d-score") + facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") +
    theme(strip.background = element_blank(), strip.text.y = element_blank()) + 
    labs(fill = "Allelic Ratio (Xi / (Xi + Xa))", col = "") + guides(colour = "none")
ggsave("./Plots/FigS3/Washout_full_heatmap_allsamples_clean.pdf", width = 16, height = 12)

to_plot %>%
  dplyr::filter(TimePoint %in% c("0d", "14d", "21d")) %>%
  dplyr::filter(Clone != "October2021") %>%
  mutate(
    Clone = case_when(
      Clone == "E6 - September2022" ~ "Rep1", 
      .default = "Rep2"
    )
  ) %>%
  ggplot(aes(reorder(name, position), y = reorder(paste0(Condition, "_", Clone), Condition), fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_y_discrete(labels = genes_of_choice, position = "right") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    scale_x_discrete(labels = genes_of_choice, position = "bottom") +
    labs(fill = "d-score") + facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") +
    theme(strip.background = element_blank(), strip.text.y = element_blank()) + 
     labs(fill = "Allelic Ratio (Xi / (Xi + Xa))", col = "") + guides(colour = "none")
ggsave("./Plots/FigS3/Washout_full_heatmap_allsamples_cleaner.pdf", width = 16, height = 12)

# Split up by replicate
to_plot %>%
  dplyr::filter(TimePoint %in% c("0d", "14d", "21d")) %>%
  dplyr::filter(Clone != "October2021") %>%
  mutate(
    Clone = case_when(
      Clone == "E6 - September2022" ~ "Rep1", 
      .default = "Rep2"
    )
  ) %>%
  dplyr::filter(Clone == "Rep1") %>%
  ggplot(aes(reorder(name, position), y = reorder(paste0(Condition, "_", Clone), Condition), fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_y_discrete(labels = genes_of_choice, position = "right") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    scale_x_discrete(labels = genes_of_choice, position = "bottom") +
    labs(fill = "d-score") + facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") +
    theme(strip.background = element_blank(), strip.text.y = element_blank()) + 
     labs(fill = "Allelic Ratio (Xi / (Xi + Xa))", col = "") + guides(colour = "none")
ggsave("./Plots/FigS3/Washout_full_heatmap_allsamples_cleaner_rep1.pdf", width = 16, height = 12)

to_plot %>%
  dplyr::filter(TimePoint %in% c("0d", "14d", "21d")) %>%
  dplyr::filter(Clone != "October2021") %>%
  mutate(
    Clone = case_when(
      Clone == "E6 - September2022" ~ "Rep1", 
      .default = "Rep2"
    )
  ) %>%
  dplyr::filter(Clone == "Rep2") %>%
  ggplot(aes(reorder(name, position), y = reorder(paste0(Condition, "_", Clone), Condition), fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_y_discrete(labels = genes_of_choice, position = "right") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    scale_x_discrete(labels = genes_of_choice, position = "bottom") +
    labs(fill = "d-score") + facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") +
    theme(strip.background = element_blank(), strip.text.y = element_blank()) + 
     labs(fill = "Allelic Ratio (Xi / (Xi + Xa))", col = "") + guides(colour = "none")
ggsave("./Plots/FigS3/Washout_full_heatmap_allsamples_cleaner_rep2.pdf", width = 16, height = 12)

```

We structure the analysis in this figure as follows: 

NEW: 
- First focus on the high reversibility replicate:
  - Short silencing: Largely reversible. Confirm with Cl30 / 31 in supplement
  - Long silencing: Partly reversible
- Then with low reversibility:
  - We noticed that proliferation time is higher in some replicates + that escape is irreversible in these
  - Show things together here

OLD: 
- Overview silencing in a heatmap - check
- Short silencing -- largely reversible. Confirm with CL30/CL31 in the supplement - check
- Long silencing -- partly reversible - check
- Annotate genes based on short term silencing + reversibility and long term silencing + reversibility


```{r asdasdasadsadasd}

## Here, we remove the samples from the first long-term experiment

e6_escapees <- Reduce("intersect", escapees_per_clone)

ratios[e6_escapees, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Condition, Clone, Dox)) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  mutate(Condition = factor(Condition, levels = rev(samples_use))) %>%
  mutate(TimePoint = case_when(
    grepl("Dox \\(3d\\)", Condition)  ~ "3d",
    grepl("Dox \\(7d\\)", Condition)  ~ "7d",
    grepl("Dox \\(14d\\)", Condition)  ~ "14d",
    grepl("Dox \\(21d\\)", Condition)  ~ "21d",
    grepl("Control", Condition)  ~ "0d")) %>% 
    mutate(TimePoint = factor(TimePoint, levels = paste0(c(0, 3, 7, 14, 21), "d"))) %>%
    dplyr::filter(!(grepl("September2022", Clone) & grepl("14|21", Condition))) -> to_plot

p1 <- to_plot %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    labs(color = "Allelic ratio \n (Xi / (Xi + Xa))") + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 10)) + 
    scale_x_discrete(position = "top") + guides(fill = "none") + 
    facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.y = element_blank())

# Plot escapee annotation at the bottom of this: 
p2 <- data.frame(Gene = e6_escapees,
           Annotation = rowData(data)[e6_escapees, ]$EscapeAnnotation, 
           Position = start(rowRanges(data)[e6_escapees, ])) %>%
  ggplot(aes(x = reorder(Gene, Position), y = 1, fill = Annotation)) + theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) + 
  theme(legend.position = "None")

p1 / p2 + plot_layout(heights = c(4, 0.2))
# ggsave("./Plots/Fig1/heatmap_ordered_only_escapees_by_position.pdf", width = 30, height = 5)
ggsave("./Plots/Fig3/heatmap_ordered_only_escapees_by_position_reversibility.pdf", width = 20, height = 5)

### We see more reversibility in the 7d washout than 14d, which is odd
to_plot %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_y_discrete(labels = genes_of_choice, position = "right") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    scale_x_discrete(labels = genes_of_choice, position = "bottom") + 
    labs(fill = "d-score") +  facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.y = element_blank())

ratios[escapees_total, ] %>%
  t() %>% data.frame() %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Washout" = data_here$ndWashout) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Dox, Washout, Clone)) %>%
  group_by(Clone) %>% 
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7", "14", "21")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>% 
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
  mutate(Condition = gsub("(0 days washout)", "Dox", Condition)) %>%
  mutate(Condition = factor(Condition, levels = c("Control", "Dox", "4 days washout", "7 days washout", "14 days washout"))) %>%
  mutate(TimeSeries = factor(TimeSeries, 
                             levels = c("Control", "Dox", "3 days dox treatment", "7 days dox treatment", "14 days dox treatment", "21 days dox treatment"))) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>% 
  dplyr::filter(!is.na(escape_status)) %>%
  dplyr::filter(!(grepl("September2022", Clone) & Condition %in% c("14 days washout", "21 days washout"))) %>% 
  dplyr::filter() -> ratios_here

ratios_here %>%
  ggplot(aes(x = factor(Condition, levels = rev(levels(Condition))), y = value, fill = escape_status)) + geom_boxplot(col = "black") + 
    theme_paper() + ylab("Allelic Ratio (Xi / (Xi + Xa))") + xlab("") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    facet_grid(cols = vars(TimeSeries), scales = "free_x", space = "free_x") + 
    scale_fill_manual(values = escape_colors) + 
    labs(colour = "Escape status") + 
    coord_flip()
ggsave("./Plots/Fig3/Washout_escape_boxplots.pdf", width = 14, height = 6)

# ratios_here %>%
#   dplyr::filter(TimeSeries %in% c("14 days dox treatment", "21 days dox treatment")) %>%
#   ggplot(aes(x = Condition, y = value, fill = escape_status)) + geom_boxplot(col = "black") + 
#     theme_paper() + ylab("ASE (d-score)") + 
#     theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
#     facet_grid(cols = vars(TimeSeries), scales = "free_x", space = "free_x") + 
#     scale_fill_manual(values = escape_colors) + 
#     labs(colour = "Escape status")
# ggsave("./Plots/Fig3/Washout_Xist_boxplots_longwashout.pdf", width = 8, height = 6)

### single genes

# plot_single_gene <- function(gene, plot = F){
#   plot_levels <- c("Control", paste0("Dox (", c(3, 7, 14, 21), "d)"), 
#                  paste0("Dox (", c(3, 7, 14, 21), "d) - washout (4d)"), 
#                  paste0("Dox (", c(3, 7, 14, 21), "d) - washout (7d)"), 
#                  paste0("Dox (", c(3, 7, 14, 21), "d) - washout (14d)"))
# to_plot %>%
#   dplyr::filter(Clone != "E6 - October2021") %>%
#   mutate(
#     Clone = case_when(
#       Clone == "E6 - September2022" ~ "Rep1", 
#       .default = "Rep2"
#     )
#   ) %>%
#   dplyr::filter(name == gene) %>%
#   mutate(Condition = factor(Condition, levels = plot_levels)) %>%
#   ggplot(aes(x = Condition, y = value)) + geom_point(col = "black", size = 5) +
#     theme_paper() + ylab("ASE (d-score)") +
#     theme(axis.text.x=element_text(angle = 45, hjust = 1)) +
#     facet_grid(cols = vars(Clone), scales = "free", space = "free") + 
#     scale_fill_manual(values = escape_colors) +
#     labs(colour = "Escape status") + xlab("") + ggtitle(gene) + ylim(c(0, 1)) -> p1
#   if (plot){
#     ggsave(paste0("/omics/groups/OE0538/internal/users/panten/projects/AntoniaESCProject/allele-specific_expression/xist_project/Plots/Fig3/single_gene_", gene, ".pdf"), p1, 
#            height = 9, width = 9)
#   }
#   p1
# }
# 
# plot_single_gene("Huwe1", plot = T)
# plot_single_gene("Mecp2", plot = T)
# plot_single_gene("Kdm5c", plot = T)
# plot_single_gene("Ddx3x", plot = T)

#### Large heatmap showing everything

# samples_use <- c("Control", 
#                  "Dox (3d)", "Dox (3d) - washout (4d)", 
#                  "Dox (7d)", "Dox (7d) - washout (4d)", "Dox (7d) - washout (7d)", 
#                  "Dox (14d)", "Dox (14d) - washout (7d)", 
#                  "Dox (21d)", "Dox (21d) - washout (7d)", 
#                  
#                  
#                  "Dox (7d) - washout (14d)", 
#                  
#                   "Dox (14d) - washout (14d)", 
#                  
#                   "Dox (21d) - washout (14d)"
# )
# 
# data_here <- data[,data$ConditionClean %in% samples_use & !(data$Experiment %in% c("October2022", "April2023"))]
# data_here <- data_here[,data_here$Clone == "E6"]
# 
# ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))
# ratios <- ratios[apply(ratios, 1, function(x){!any(is.na(x))}), ]
# 
# e6_escapees <- c(Reduce("intersect", escapees_per_clone), "Xist")
# 
# # To structure the data, we order the genes using hierarchical clustering
# data_for_clustering <- do.call("rbind", lapply(unique(data_here$ConditionClean), function(x){
#   if (sum(data_here$ConditionClean == x) == 1){
#     ratios[e6_escapees, data_here$ConditionClean == x]
#   } else {
#     rowMeans(ratios[e6_escapees, data_here$ConditionClean == x])
#   }
# }))
# 
# clustering_order <- setNames(hclust(dist(t(data_for_clustering)))$order, rownames(ratios[e6_escapees, ]))
# 
# ratios[e6_escapees, ][clustering_order, ] %>%
#   t() %>% data.frame(check.names = F) %>%
#   add_column("Condition" = data_here$ConditionClean) %>%
#   add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
#   pivot_longer(-c(Condition, Clone)) %>%
#   mutate(name = factor(name, levels = unique(.$name))) %>%
#   add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
#   mutate(Condition = factor(Condition, levels = rev(samples_use))) -> to_plot
# 
# p1 <- to_plot %>%
#   ggplot(aes(name, y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
#     scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
#     scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
#     theme_paper() + 
#     theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
#     xlab("") + 
#     ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
#     labs(fill = "d-score") + 
#     theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 8)) + 
#     labs(fill = "Allelic ratio Xi / (Xi + Xa)") + 
#     scale_x_discrete(position = "top") + guides(fill = "none")
# 
# # Plot escapee annotation at the bottom of this: 
# p2 <- data.frame(Gene = e6_escapees,  
#            Annotation = rowData(data)[e6_escapees, ]$EscapeAnnotation, 
#            Position = start(rowRanges(data)[e6_escapees, ])) %>%
#   ggplot(aes(x = reorder(Gene, Position), y = 1, fill = Annotation)) + theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) + 
#     theme(legend.position = "None")
# 
# p1 / p2 + plot_layout(heights = c(4, 0.2))
# ggsave("./Plots/Fig3/heatmap_ordered_only_escapees.pdf", width = 25, height = 4)
# 
# # 
# ratios[e6_escapees, ] %>%
#   t() %>% data.frame(check.names = F) %>%
#   add_column("Condition" = data_here$ConditionClean) %>%
#   add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
#   pivot_longer(-c(Condition, Clone)) %>%
#   mutate(name = factor(name, levels = unique(.$name))) %>%
#   add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
#   mutate(Condition = factor(Condition, levels = rev(samples_use))) -> to_plot
# 
# p1 <- to_plot %>%
#   ggplot(aes(reorder(name, position),  y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
#     scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
#     scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
#     theme_paper() + 
#     theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
#     xlab("") + 
#     ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
#     labs(fill = "d-score") + 
#     theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 10)) + 
#     scale_x_discrete(position = "top") + guides(fill = "none")
# 
# # Plot escapee annotation at the bottom of this: 
# p2 <- data.frame(Gene = e6_escapees,  
#            Annotation = rowData(data)[e6_escapees, ]$EscapeAnnotation, 
#            Position = start(rowRanges(data)[e6_escapees, ])) %>%
#   ggplot(aes(x = reorder(Gene, Position), y = 1, fill = Annotation)) + theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) + 
#   theme(legend.position = "None")
# 
# p1 / p2 + plot_layout(heights = c(4, 0.2))
# ggsave("./Plots/Fig3/heatmap_ordered_only_escapees_genomic_order.pdf", width = 24, height = 5)


#### LM stuff for SPEN-experiment - first the early timepoint

samples_use <- c("Control", 
                 "Dox (3d)", "Dox (3d) - washout (4d)", 
                 "Dox (7d)", "Dox (7d) - washout (4d)"
)

data_here <- data[,data$ConditionClean %in% samples_use]
data_here <- data_here[,data_here$Clone == "E6"]
data_here <- data_here[, !(grepl("September2022", data_here$Experiment) & grepl("14|21", data_here$Condition))]

general_escape_genes <- Reduce("intersect", escapees_per_clone)
general_escapees <- data_here[general_escape_genes, ]

data_test_inactive <- counts_inactive(general_escapees)
data_test_active <- counts_active(general_escapees)

fitting_metadata_here <- colData(general_escapees)
fitting_metadata_here$washout <- ifelse(fitting_metadata_here$washout == "NO", "NO", paste0(fitting_metadata_here$washout, fitting_metadata_here$ndDox))

data_test_inactive <- data_test_inactive[!rownames(data_test_inactive) %in% genes_out, ]
data_test_active <- data_test_active[!rownames(data_test_active) %in% genes_out, ]

all_coefs <- data.frame(do.call("rbind", lapply(1:nrow(data_test_active), function(i){
  tryCatch({
    y = as.numeric(data_test_inactive[i, ]) + 1
    N = as.numeric(data_test_active[i, ] + data_test_inactive[i, ]) + 2
    fit1 <- glm(cbind(y, N-y) ~ ndDox + washout, family = "binomial", data = fitting_metadata_here)
    coefs <- coef(fit1)
    coefs
    return(coefs)
  }, error = function(cond) {
    return(c(NA))
  }, warning = function(cond){
    return(c(NA))
  })
})))

rownames(all_coefs) <- rownames(data_test_active)
all_coefs <- all_coefs[!apply(all_coefs, 1, function(x){all(is.na(x))}), !apply(all_coefs, 2, function(x){all(is.na(x))})]

row_annotation <- data.frame(
  Gene = rownames(all_coefs), 
  escapee_annotation = as.character(rowData(data_here)[match(rownames(all_coefs), rowData(data_here)$gene_name), 7])
) %>% column_to_rownames("Gene")

all_coefs %>%
  rownames_to_column("Gene") %>%
  pivot_longer(-c(Gene)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(x = name, y = value, col = escape_status)) + geom_boxplot() + 
    geom_jitter(position = position_jitterdodge(jitter.width = 0.1), alpha = 0.2) +
    coord_flip() + theme_paper() + xlab("Covariate") +
    theme(aspect.ratio = 2) + theme(legend.position = "top") + geom_hline(yintercept = 0, linetype = "dashed") + 
    ylab("Model coefficient") + xlab("") + 
    scale_color_manual(values = escape_colors)
ggsave("./Plots/FigS3/Washout_escape_by_category_boxplot.pdf")

# We now convert the coefficients to fold changes in escape: 

df_test <- data.frame(
  Gene = rownames(all_coefs), 
  control = exp(all_coefs$X.Intercept.) / (exp(all_coefs$X.Intercept.) + 1), 
  nddox3_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox3) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox3) + 1), 
  nddox7_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox7) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox7) + 1), 
  nddox3_washout = exp(all_coefs$X.Intercept. + all_coefs$ndDox3 + all_coefs$washoutYES3) / 
    (exp(all_coefs$X.Intercept. + all_coefs$ndDox3 + all_coefs$washoutYES3) + 1), 
  nddox7_washout = exp(all_coefs$X.Intercept. + all_coefs$ndDox7 + all_coefs$washoutYES7) / 
    (exp(all_coefs$X.Intercept. + all_coefs$ndDox7 + all_coefs$washoutYES7) + 1)
)

df_test %>%
  mutate(control_dox_difference = control - nddox3_treat) %>%
  mutate(washout_dox_difference = nddox3_washout - nddox3_treat) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference < 0, 0, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference < 0, 0, washout_dox_difference)) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference > 0.5, 0.5, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference > 0.5, 0.5, washout_dox_difference)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(control_dox_difference, washout_dox_difference, col = escape_status)) + geom_point() + 
    geom_smooth(method = "lm", col = "black", linetype = 'dashed', size = 0.5) + 
    #ggrepel::geom_text_repel(aes(label = Gene), size = 5) + 
    geom_abline(linetype = "dashed", slope = 1) + 
    theme_paper() + 
    xlab("Dox-effect \n (d-score difference Control - Dox)") + 
    ylab("Washout-effect \n (d-score difference  Dox - Dox + Washout)") + 
    ggtitle("3d analysis") + 
    # xlim(c(-0.1, 0.5)) + ylim(c(-0.1, 0.5)) + 
    coord_fixed() + 
    scale_color_manual(values = escape_colors) + labs(colour = "Escape status")
ggsave("./Plots/Fig3/washout_vs_dox_model_3d.pdf", width = 8, height = 8)

df_test %>%
  mutate(control_dox_difference = control - nddox7_treat) %>%
  mutate(washout_dox_difference = nddox7_washout - nddox7_treat) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference < 0, 0, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference < 0, 0, washout_dox_difference)) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference > 0.5, 0.5, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference > 0.5, 0.5, washout_dox_difference)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(control_dox_difference, washout_dox_difference, col = escape_status)) + geom_point() + 
    geom_smooth(method = "lm", col = "black", linetype = 'dashed', size = 0.5) + 
    #ggrepel::geom_text_repel(aes(label = Gene), size = 5) + 
    geom_abline(linetype = "dashed", slope = 1) + 
    theme_paper() + 
    xlab("Dox-effect \n (d-score difference Control - Dox)") + 
    ylab("Washout-effect \n (d-score difference  Dox - Dox + Washout)") + 
    ggtitle("7d analysis") + 
    # xlim(c(-0.1, 0.5)) + ylim(c(-0.1, 0.5)) + 
    coord_fixed() + 
    scale_color_manual(values = escape_colors) + labs(colour = "Escape status")
ggsave("./Plots/Fig3/Washout_vs_dox_model_7d.pdf", width = 8, height = 8)


### Now, we focus on the late timepoints: 

samples_use <- c("Control", 
                 "Dox (14d)", "Dox (14d) - washout (7d)", 
                 "Dox (21d)", "Dox (21d) - washout (7d)"
)

data_here <- data[,data$ConditionClean %in% samples_use]
data_here <- data_here[,data_here$Clone == "E6"]
data_here <- data_here[, !(grepl("September2022", data_here$Experiment) & grepl("14|21", data_here$Condition))]
data_here$washout <- gsub("NA", "NO", data_here$washout)

general_escape_genes <- Reduce("intersect", escapees_per_clone)
general_escapees <- data_here[general_escape_genes, ]

data_test_inactive <- counts_inactive(general_escapees)
data_test_active <- counts_active(general_escapees)

fitting_metadata_here <- colData(general_escapees)
fitting_metadata_here$washout <- ifelse(fitting_metadata_here$washout == "NO", "NO", paste0(fitting_metadata_here$washout, fitting_metadata_here$ndDox))

data_test_inactive <- data_test_inactive[!rownames(data_test_inactive) %in% genes_out, ]
data_test_active <- data_test_active[!rownames(data_test_active) %in% genes_out, ]

all_coefs <- data.frame(do.call("rbind", lapply(1:nrow(data_test_active), function(i){
  tryCatch({
    y = as.numeric(data_test_inactive[i, ]) + 1
    N = as.numeric(data_test_active[i, ] + data_test_inactive[i, ]) + 2
    fit1 <- glm(cbind(y, N-y) ~ ndDox + washout, family = "binomial", data = fitting_metadata_here)
    coefs <- coef(fit1)
    coefs
    return(coefs)
  }, error = function(cond) {
    return(c(NA))
  }, warning = function(cond){
    return(c(NA))
  })
})))

rownames(all_coefs) <- rownames(data_test_active)
all_coefs <- all_coefs[!apply(all_coefs, 1, function(x){all(is.na(x))}), !apply(all_coefs, 2, function(x){all(is.na(x))})]

row_annotation <- data.frame(
  Gene = rownames(all_coefs), 
  escapee_annotation = as.character(rowData(data_here)[match(rownames(all_coefs), rowData(data_here)$gene_name), 7])
) %>% column_to_rownames("Gene")

all_coefs %>%
  rownames_to_column("Gene") %>%
  pivot_longer(-c(Gene)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(x = name, y = value, col = escape_status)) + geom_boxplot() + 
    geom_jitter(position = position_jitterdodge(jitter.width = 0.1), alpha = 0.2) +
    coord_flip() + theme_paper() + xlab("Covariate") +
    theme(aspect.ratio = 2) + theme(legend.position = "top") + geom_hline(yintercept = 0, linetype = "dashed") + 
    ylab("Model coefficient") + xlab("") + 
    scale_color_manual(values = escape_colors)
ggsave("./Plots/FigS3/Washout_escape_by_category_boxplot_long.pdf")

# We now convert the coefficients to fold changes in escape: 

df_test <- data.frame(
  Gene = rownames(all_coefs), 
  control = exp(all_coefs$X.Intercept.) / (exp(all_coefs$X.Intercept.) + 1), 
  nddox14_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox14) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox14) + 1), 
  nddox21_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox21) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox21) + 1), 
  nddox14_washout = exp(all_coefs$X.Intercept. + all_coefs$ndDox14 + all_coefs$washoutYES14) / 
    (exp(all_coefs$X.Intercept. + all_coefs$ndDox14 + all_coefs$washoutYES14) + 1), 
  nddox21_washout = exp(all_coefs$X.Intercept. + all_coefs$ndDox21 + all_coefs$washoutYES21) / 
    (exp(all_coefs$X.Intercept. + all_coefs$ndDox21 + all_coefs$washoutYES21) + 1)
)

df_test %>%
  mutate(control_dox_difference = control - nddox14_treat) %>%
  mutate(washout_dox_difference = nddox14_washout - nddox14_treat) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference < 0, 0, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference < 0, 0, washout_dox_difference)) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference > 0.5, 0.5, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference > 0.5, 0.5, washout_dox_difference)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(control_dox_difference, washout_dox_difference, col = escape_status)) + geom_point() + 
    geom_smooth(method = "lm", col = "black", linetype = 'dashed', size = 0.5) + 
    #ggrepel::geom_text_repel(aes(label = Gene), size = 5) + 
    geom_abline(linetype = "dashed", slope = 1) + 
    theme_paper() + 
    xlab("Dox-effect \n (d-score difference Control - Dox)") + 
    ylab("Washout-effect \n (d-score difference  Dox - Dox + Washout)") + 
    ggtitle("14d analysis") + 
    #xlim(c(-0.1, 0.5)) + ylim(c(-0.1, 0.5)) + 
    coord_fixed() + 
    scale_color_manual(values = escape_colors) + labs(colour = "Escape status")
ggsave("./Plots/Fig3/washout_vs_dox_model_14d.pdf", width = 8, height = 8)

df_test %>%
  mutate(control_dox_difference = control - nddox21_treat) %>%
  mutate(washout_dox_difference = nddox21_washout - nddox21_treat) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference < 0, 0, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference < 0, 0, washout_dox_difference)) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference > 0.5, 0.5, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference > 0.5, 0.5, washout_dox_difference)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(control_dox_difference, washout_dox_difference, col = escape_status)) + geom_point() + 
    geom_smooth(method = "lm", col = "black", linetype = 'dashed', size = 0.5) + 
    #ggrepel::geom_text_repel(aes(label = Gene), size = 5) + 
    geom_abline(linetype = "dashed", slope = 1) + 
    theme_paper() + 
    xlab("Dox-effect \n (d-score difference Control - Dox)") + 
    ylab("Washout-effect \n (d-score difference  Dox - Dox + Washout)") + 
    ggtitle("21d analysis") + 
    #xlim(c(-0.1, 0.5)) + ylim(c(-0.1, 0.5)) + 
    coord_fixed() + 
    scale_color_manual(values = escape_colors) + labs(colour = "Escape status")
ggsave("./Plots/Fig3/Washout_vs_dox_model_21d.pdf", width = 8, height = 8)

# Check how the model coefficients fit with allelic ratio differences: 
# 
# inv_logit <- function(x){1 / (1 + exp(-x))}
# odds_ratio <- function(x, y){
#   log((x / (1 - x)) / (y / (1 - y)))
# }
# 
# df_test <- data.frame(
#   nddox3_control_obs = rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Control"]), 
#   nddox3_treat_obs = rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Dox (3d)"]), 
#   nddox3_control = exp(all_coefs$X.Intercept.) / (exp(all_coefs$X.Intercept.) + 1), 
#   nddox3_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox3) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox3) + 1)
# )
# 
# df_test %>%
#   ggplot(aes(x = nddox3_treat_obs, nddox3_treat)) + geom_point() + geom_abline()
# 
# all_coefs$d_score_3d <- odds_ratio(rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Dox (3d)"]), 
#   rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Control"]))
# 
# all_coefs$d_score_7d <- odds_ratio(rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Dox (7d)"]), 
#   rowMeans(ratios[rownames(all_coefs), data_here$ConditionClean == "Control"]))
# 
# all_coefs %>% ggplot(aes(x = d_score_3d, ndDox3)) + geom_point() + geom_abline()
# 
# all_coefs %>% ggplot(aes(x = d_score_7d, ndDox7)) + geom_point() + geom_abline()

# Quantify silencing and lack-of-silencing effects in two timepoints: 

# ratios_escapees <- ratios[rownames(general_escapees), ]
# 
# data.frame(
#   Gene = rownames(ratios_escapees), 
#   Control = rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
#   dox_3d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d)"]), 
#   dox_3d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d), Aux (5d)"]), 
#   escape_category = rowData(data_here[rownames(ratios_escapees), ])$EscapeAnnotation
# ) %>%
#   ggplot(aes(x = Control, y = dox_3d, col = escape_category)) + geom_point() + geom_abline() + ggrepel::geom_text_repel(aes(label = Gene)) + 
#     scale_color_manual(values = escape_colors) + theme_paper() + geom_smooth(method = "lm", linetype = 'dashed', color = "darkblue")
# 
# data.frame(
#   Gene = rownames(ratios_escapees), 
#   Control = rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
#   dox_3d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d)"]), 
#   dox_3d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d), Aux (5d)"]), 
#   escape_category = rowData(data_here[rownames(ratios_escapees), ])$EscapeAnnotation
# ) %>%
#   ggplot(aes(x = Control, y = dox_3d_aux, col = escape_category)) + geom_point() + geom_abline() + ggrepel::geom_text_repel(aes(label = Gene)) + 
#     scale_color_manual(values = escape_colors) + theme_paper() + geom_smooth(method = "lm", linetype = 'dashed', color = "darkblue")
# 
# data.frame(
#   Gene = rownames(ratios_escapees), 
#   Control = rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
#   dox_7d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d)"]), 
#   dox_7d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7dd), Aux (9d)"]), 
#   escape_category = rowData(data_here[rownames(ratios_escapees), ])$EscapeAnnotation
# ) %>%
#   ggplot(aes(x = Control, y = dox_7d, col = escape_category)) + geom_point() + geom_abline() + ggrepel::geom_text_repel(aes(label = Gene)) + 
#     scale_color_manual(values = escape_colors) + theme_paper() + geom_smooth(method = "lm", linetype = 'dashed', color = "darkblue")
# 
# data.frame(
#   Gene = rownames(ratios_escapees), 
#   Control = rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
#   dox_7d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d)"]), 
#   dox_7d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d), Aux (9d)"]), 
#   escape_category = rowData(data_here[rownames(ratios_escapees), ])$EscapeAnnotation
# ) %>%
#   ggplot(aes(x = Control, y = dox_7d_aux, col = escape_category)) + geom_point() + geom_abline() + ggrepel::geom_text_repel(aes(label = Gene)) + 
#     scale_color_manual(values = escape_colors) + theme_paper() + geom_smooth(method = "lm", linetype = 'dashed', color = "darkblue")
# 
# genes_highlight = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d), Aux (9d)"]) - rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"])
# genes_highlight <- names(genes_highlight[genes_highlight < -0.08])
# 
# dff <- data.frame(
#   Gene = rownames(ratios_escapees), 
#   dox_3d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d)"]) - rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
#   dox_3d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (3d), Aux (5d)"]) - rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
#   dox_7d = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d)"]) - rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]), 
#   dox_7d_aux = rowMeans(ratios_escapees[,data_here$ConditionClean == "Dox (7d), Aux (9d)"]) - rowMeans(ratios_escapees[,data_here$ConditionClean == "Control"]) 
# )
# 
# dff %>% pivot_longer(-c(Gene)) %>%
#   ggplot(aes(x = name, y = value)) + geom_boxplot(width = 0.1, outlier.color = NA) + ggbeeswarm::geom_quasirandom() + theme_paper() + 
#     ggrepel::geom_text_repel(aes(label = Gene), data = dff[dff$Gene %in% genes_highlight, ] %>% pivot_longer(-c(Gene)))

```

Now we save our classification based on the 21d analysis to use in further analysis: 

```{r save_classification}

saveRDS(df_test, "./ProcessedData/reversibility_21d.rds")

# df_test %>%
#   ggplot(aes(control - nddox21_treat, nddox21_washout - nddox21_treat)) + geom_point() + 
#     #geom_smooth(method = "lm", col = "black", linetype = 'dashed', size = 0.5) + 
#     # ggrepel::geom_text_repel(aes(label = Gene), size = 5) + 
#     geom_abline(linetype = "dashed", slope = 1) + 
#     geom_abline(intercept = -0.1, slope = 1) + 
#     geom_hline(intercept = -0.1, slope = 1) + 
#     theme_paper() + 
#     xlab("Dox-effect \n (d-score difference Control - Dox)") + 
#     ylab("Washout-effect \n (d-score difference  Dox - Dox + Washout)") + 
#     ggtitle("21d analysis") + 
#     xlim(c(-0.1, 0.5)) + ylim(c(-0.1, 0.5)) + 
#     coord_fixed()

```


```{r asdasdasadsadasd}

samples_use <- c("Control", 
                 "Dox (3d)", "Dox (3d) - washout (4d)", 
                 
                 "Dox (7d)", 
                 "Dox (7d) - washout (4d)", 
                 "Dox (7d) - washout (7d)", 
                 "Dox (7d) - washout (14d)", 
                 
                 "Dox (14d)", 
                 "Dox (14d) - washout (7d)", 
                 "Dox (14d) - washout (14d)", 
                 
                 "Dox (21d)", 
                 "Dox (21d) - washout (7d)", 
                 "Dox (21d) - washout (14d)"
)

## Here, we remove the samples from the first long-term experiment

data_here <- data[,colnames(ratios)]

e6_escapees <- Reduce("intersect", escapees_per_clone)

ratios[e6_escapees, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Condition, Clone, Dox)) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  mutate(Condition = factor(Condition, levels = rev(samples_use))) %>%
  mutate(TimePoint = case_when(
    grepl("Dox \\(3d\\)", Condition)  ~ "3d",
    grepl("Dox \\(7d\\)", Condition)  ~ "7d",
    grepl("Dox \\(14d\\)", Condition)  ~ "14d",
    grepl("Dox \\(21d\\)", Condition)  ~ "21d",
    grepl("Control", Condition)  ~ "0d")) %>% 
    mutate(TimePoint = factor(TimePoint, levels = paste0(c(0, 3, 7, 14, 21), "d"))) %>%
    dplyr::filter(!(!grepl("September2022", Clone) & grepl("14|21", Condition))) -> to_plot

p1 <- to_plot %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    labs(color = "Allelic ratio \n (Xi / (Xi + Xa))") + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 10)) + 
    scale_x_discrete(position = "top") + guides(fill = "none") + 
    facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.y = element_blank())

# Plot escapee annotation at the bottom of this: 
p2 <- data.frame(Gene = e6_escapees,
           Annotation = rowData(data)[e6_escapees, ]$EscapeAnnotation, 
           Position = start(rowRanges(data)[e6_escapees, ])) %>%
  ggplot(aes(x = reorder(Gene, Position), y = 1, fill = Annotation)) + theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) + 
  theme(legend.position = "None")

p1 / p2 + plot_layout(heights = c(4, 0.2))
# ggsave("./Plots/Fig1/heatmap_ordered_only_escapees_by_position.pdf", width = 30, height = 5)
ggsave("./Plots/Fig3/heatmap_ordered_only_escapees_by_position_lowescape.pdf", width = 20, height = 5)

ratios[escapees_total, ] %>%
  t() %>% data.frame() %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Washout" = data_here$ndWashout) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Dox, Washout, Clone)) %>%
  group_by(Clone) %>% 
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7", "14", "21")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>% 
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
  mutate(Condition = gsub("(0 days washout)", "Dox", Condition)) %>%
  mutate(Condition = factor(Condition, levels = c("Control", "Dox", "4 days washout", "7 days washout", "14 days washout"))) %>%
  mutate(TimeSeries = factor(TimeSeries, 
                             levels = c("Control", "Dox", "3 days dox treatment", "7 days dox treatment", "14 days dox treatment", "21 days dox treatment"))) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>% 
  dplyr::filter(!is.na(escape_status)) %>%
  dplyr::filter(!(!grepl("September2022", Clone) & Condition %in% c("14 days washout", "21 days washout"))) %>% 
  dplyr::filter() -> ratios_here

ratios_here %>%
  ggplot(aes(x = factor(Condition, levels = rev(levels(Condition))), y = value, fill = escape_status)) + geom_boxplot(col = "black") + 
    theme_paper() + ylab("Allelic Ratio (Xi / (Xi + Xa))") + xlab("") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    facet_grid(cols = vars(TimeSeries), scales = "free_x", space = "free_x") + 
    scale_fill_manual(values = escape_colors) + 
    labs(colour = "Escape status") + 
    coord_flip()
ggsave("./Plots/Fig3/Washout_escape_boxplots_lowescape.pdf", width = 14, height = 6)


#### LM stuff for SPEN-experiment - first the early timepoint

samples_use <- c("Control", 
                 "Dox (3d)", "Dox (3d) - washout (4d)", 
                 "Dox (7d)", "Dox (7d) - washout (4d)"
)

data_here <- data[,data$ConditionClean %in% samples_use]
data_here <- data_here[,data_here$Clone == "E6"]
data_here <- data_here[, !(!grepl("September2022", data_here$Experiment) & grepl("14|21", data_here$Condition))]

general_escape_genes <- Reduce("intersect", escapees_per_clone)
general_escapees <- data_here[general_escape_genes, ]

data_test_inactive <- counts_inactive(general_escapees)
data_test_active <- counts_active(general_escapees)

fitting_metadata_here <- colData(general_escapees)
fitting_metadata_here$washout <- ifelse(fitting_metadata_here$washout == "NO", "NO", paste0(fitting_metadata_here$washout, fitting_metadata_here$ndDox))

data_test_inactive <- data_test_inactive[!rownames(data_test_inactive) %in% genes_out, ]
data_test_active <- data_test_active[!rownames(data_test_active) %in% genes_out, ]

all_coefs <- data.frame(do.call("rbind", lapply(1:nrow(data_test_active), function(i){
  tryCatch({
    y = as.numeric(data_test_inactive[i, ]) + 1
    N = as.numeric(data_test_active[i, ] + data_test_inactive[i, ]) + 2
    fit1 <- glm(cbind(y, N-y) ~ ndDox + washout, family = "binomial", data = fitting_metadata_here)
    coefs <- coef(fit1)
    coefs
    return(coefs)
  }, error = function(cond) {
    return(c(NA))
  }, warning = function(cond){
    return(c(NA))
  })
})))

rownames(all_coefs) <- rownames(data_test_active)
all_coefs <- all_coefs[!apply(all_coefs, 1, function(x){all(is.na(x))}), !apply(all_coefs, 2, function(x){all(is.na(x))})]

row_annotation <- data.frame(
  Gene = rownames(all_coefs), 
  escapee_annotation = as.character(rowData(data_here)[match(rownames(all_coefs), rowData(data_here)$gene_name), 7])
) %>% column_to_rownames("Gene")

all_coefs %>%
  rownames_to_column("Gene") %>%
  pivot_longer(-c(Gene)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(x = name, y = value, col = escape_status)) + geom_boxplot() + 
    geom_jitter(position = position_jitterdodge(jitter.width = 0.1), alpha = 0.2) +
    coord_flip() + theme_paper() + xlab("Covariate") +
    theme(aspect.ratio = 2) + theme(legend.position = "top") + geom_hline(yintercept = 0, linetype = "dashed") + 
    ylab("Model coefficient") + xlab("") + 
    scale_color_manual(values = escape_colors)
ggsave("./Plots/FigS3/Washout_escape_by_category_boxplot_lowescape.pdf")

# We now convert the coefficients to fold changes in escape: 

df_test <- data.frame(
  Gene = rownames(all_coefs), 
  control = exp(all_coefs$X.Intercept.) / (exp(all_coefs$X.Intercept.) + 1), 
  nddox3_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox3) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox3) + 1), 
  nddox7_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox7) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox7) + 1), 
  nddox3_washout = exp(all_coefs$X.Intercept. + all_coefs$ndDox3 + all_coefs$washoutYES3) / 
    (exp(all_coefs$X.Intercept. + all_coefs$ndDox3 + all_coefs$washoutYES3) + 1), 
  nddox7_washout = exp(all_coefs$X.Intercept. + all_coefs$ndDox7 + all_coefs$washoutYES7) / 
    (exp(all_coefs$X.Intercept. + all_coefs$ndDox7 + all_coefs$washoutYES7) + 1)
)

df_test %>%
  mutate(control_dox_difference = control - nddox3_treat) %>%
  mutate(washout_dox_difference = nddox3_washout - nddox3_treat) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference < 0, 0, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference < 0, 0, washout_dox_difference)) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference > 0.5, 0.5, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference > 0.5, 0.5, washout_dox_difference)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(control_dox_difference, washout_dox_difference, col = escape_status)) + geom_point() + 
    geom_smooth(method = "lm", col = "black", linetype = 'dashed', size = 0.5) + 
    #ggrepel::geom_text_repel(aes(label = Gene), size = 5) + 
    geom_abline(linetype = "dashed", slope = 1) + 
    theme_paper() + 
    xlab("Dox-effect \n (d-score difference Control - Dox)") + 
    ylab("Washout-effect \n (d-score difference  Dox - Dox + Washout)") + 
    ggtitle("3d analysis") + 
    # xlim(c(-0.1, 0.5)) + ylim(c(-0.1, 0.5)) + 
    coord_fixed() + 
    scale_color_manual(values = escape_colors) + labs(colour = "Escape status")
ggsave("./Plots/Fig3/washout_vs_dox_model_3d_lowescape.pdf", width = 8, height = 8)

df_test %>%
  mutate(control_dox_difference = control - nddox7_treat) %>%
  mutate(washout_dox_difference = nddox7_washout - nddox7_treat) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference < 0, 0, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference < 0, 0, washout_dox_difference)) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference > 0.5, 0.5, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference > 0.5, 0.5, washout_dox_difference)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(control_dox_difference, washout_dox_difference, col = escape_status)) + geom_point() + 
    geom_smooth(method = "lm", col = "black", linetype = 'dashed', size = 0.5) + 
    #ggrepel::geom_text_repel(aes(label = Gene), size = 5) + 
    geom_abline(linetype = "dashed", slope = 1) + 
    theme_paper() + 
    xlab("Dox-effect \n (d-score difference Control - Dox)") + 
    ylab("Washout-effect \n (d-score difference  Dox - Dox + Washout)") + 
    ggtitle("7d analysis") + 
    # xlim(c(-0.1, 0.5)) + ylim(c(-0.1, 0.5)) + 
    coord_fixed() + 
    scale_color_manual(values = escape_colors) + labs(colour = "Escape status")
ggsave("./Plots/Fig3/Washout_vs_dox_model_7d_lowescape.pdf", width = 8, height = 8)


### Now, we focus on the late timepoints: 

samples_use <- c("Control", 
                 "Dox (14d)", "Dox (14d) - washout (7d)", 
                 "Dox (21d)", "Dox (21d) - washout (7d)"
)

data_here <- data[,data$ConditionClean %in% samples_use]
data_here <- data_here[,data_here$Clone == "E6"]
data_here <- data_here[, !(!grepl("September2022", data_here$Experiment) & grepl("14|21", data_here$Condition))]
data_here$washout <- gsub("NA", "NO", data_here$washout)

general_escape_genes <- Reduce("intersect", escapees_per_clone)
general_escapees <- data_here[general_escape_genes, ]

data_test_inactive <- counts_inactive(general_escapees)
data_test_active <- counts_active(general_escapees)

fitting_metadata_here <- colData(general_escapees)
fitting_metadata_here$washout <- ifelse(fitting_metadata_here$washout == "NO", "NO", paste0(fitting_metadata_here$washout, fitting_metadata_here$ndDox))

data_test_inactive <- data_test_inactive[!rownames(data_test_inactive) %in% genes_out, ]
data_test_active <- data_test_active[!rownames(data_test_active) %in% genes_out, ]

all_coefs <- data.frame(do.call("rbind", lapply(1:nrow(data_test_active), function(i){
  tryCatch({
    y = as.numeric(data_test_inactive[i, ]) + 1
    N = as.numeric(data_test_active[i, ] + data_test_inactive[i, ]) + 2
    fit1 <- glm(cbind(y, N-y) ~ ndDox + washout, family = "binomial", data = fitting_metadata_here)
    coefs <- coef(fit1)
    coefs
    return(coefs)
  }, error = function(cond) {
    return(c(NA))
  }, warning = function(cond){
    return(c(NA))
  })
})))

rownames(all_coefs) <- rownames(data_test_active)
all_coefs <- all_coefs[!apply(all_coefs, 1, function(x){all(is.na(x))}), !apply(all_coefs, 2, function(x){all(is.na(x))})]

row_annotation <- data.frame(
  Gene = rownames(all_coefs), 
  escapee_annotation = as.character(rowData(data_here)[match(rownames(all_coefs), rowData(data_here)$gene_name), 7])
) %>% column_to_rownames("Gene")

all_coefs %>%
  rownames_to_column("Gene") %>%
  pivot_longer(-c(Gene)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(x = name, y = value, col = escape_status)) + geom_boxplot() + 
    geom_jitter(position = position_jitterdodge(jitter.width = 0.1), alpha = 0.2) +
    coord_flip() + theme_paper() + xlab("Covariate") +
    theme(aspect.ratio = 2) + theme(legend.position = "top") + geom_hline(yintercept = 0, linetype = "dashed") + 
    ylab("Model coefficient") + xlab("") + 
    scale_color_manual(values = escape_colors)
ggsave("./Plots/FigS3/Washout_escape_by_category_boxplot_long_lowescape.pdf")

# We now convert the coefficients to fold changes in escape: 

df_test <- data.frame(
  Gene = rownames(all_coefs), 
  control = exp(all_coefs$X.Intercept.) / (exp(all_coefs$X.Intercept.) + 1), 
  nddox14_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox14) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox14) + 1), 
  nddox21_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox21) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox21) + 1), 
  nddox14_washout = exp(all_coefs$X.Intercept. + all_coefs$ndDox14 + all_coefs$washoutYES14) / 
    (exp(all_coefs$X.Intercept. + all_coefs$ndDox14 + all_coefs$washoutYES14) + 1), 
  nddox21_washout = exp(all_coefs$X.Intercept. + all_coefs$ndDox21 + all_coefs$washoutYES21) / 
    (exp(all_coefs$X.Intercept. + all_coefs$ndDox21 + all_coefs$washoutYES21) + 1)
)

df_test %>%
  mutate(control_dox_difference = control - nddox14_treat) %>%
  mutate(washout_dox_difference = nddox14_washout - nddox14_treat) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference < 0, 0, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference < 0, 0, washout_dox_difference)) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference > 0.5, 0.5, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference > 0.5, 0.5, washout_dox_difference)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(control_dox_difference, washout_dox_difference, col = escape_status)) + geom_point() + 
    geom_smooth(method = "lm", col = "black", linetype = 'dashed', size = 0.5) + 
    #ggrepel::geom_text_repel(aes(label = Gene), size = 5) + 
    geom_abline(linetype = "dashed", slope = 1) + 
    theme_paper() + 
    xlab("Dox-effect \n (d-score difference Control - Dox)") + 
    ylab("Washout-effect \n (d-score difference  Dox - Dox + Washout)") + 
    ggtitle("14d analysis") + 
    #xlim(c(-0.1, 0.5)) + ylim(c(-0.1, 0.5)) + 
    coord_fixed() + 
    scale_color_manual(values = escape_colors) + labs(colour = "Escape status")
ggsave("./Plots/Fig3/washout_vs_dox_model_14d_lowescape.pdf", width = 8, height = 8)

df_test %>%
  mutate(control_dox_difference = control - nddox21_treat) %>%
  mutate(washout_dox_difference = nddox21_washout - nddox21_treat) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference < 0, 0, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference < 0, 0, washout_dox_difference)) %>%
  mutate(control_dox_difference = ifelse(control_dox_difference > 0.5, 0.5, control_dox_difference)) %>%
  mutate(washout_dox_difference = ifelse(washout_dox_difference > 0.5, 0.5, washout_dox_difference)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$Gene, rowData(data_here)$gene_name), ]$EscapeAnnotation)) %>% 
  dplyr::filter(escape_status != "NA") %>%
  ggplot(aes(control_dox_difference, washout_dox_difference, col = escape_status)) + geom_point() + 
    geom_smooth(method = "lm", col = "black", linetype = 'dashed', size = 0.5) + 
    #ggrepel::geom_text_repel(aes(label = Gene), size = 5) + 
    geom_abline(linetype = "dashed", slope = 1) + 
    theme_paper() + 
    xlab("Dox-effect \n (d-score difference Control - Dox)") + 
    ylab("Washout-effect \n (d-score difference  Dox - Dox + Washout)") + 
    ggtitle("21d analysis") + 
    #xlim(c(-0.1, 0.5)) + ylim(c(-0.1, 0.5)) + 
    coord_fixed() + 
    scale_color_manual(values = escape_colors) + labs(colour = "Escape status")
ggsave("./Plots/Fig3/Washout_vs_dox_model_21d_lowescape.pdf", width = 8, height = 8)

```
