---
title: "XistPaper_Figure03"
output: html_document
author: "Jasper Panten"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/../../"))
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

This script reproduces all plots for Figure 3 and associated supplementary figures in PaperName.
We first load the necessary libraries and some helper functions: 

```{r load_libraries, message=F}

setwd(paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/../../../"))

library(tidyverse)
library(ggplot2)
library(scran)
library(scater)
library(DESeq2)
library(patchwork)

source("./xist_project/Scripts/General/auxiliary.R")
source("./xist_project/Scripts/General/ase_functions.R")

escape_colors <- setNames(c("darkgrey", "darkgreen", "orange"), nm = c("NPC-specific", "Facultative", "Constitutive"))

```

Now, read in the SCE, remove non-X genes and exclude lowly expressed genes

```{r load_data}

data <- readRDS("./ProcessedData/merged_dataset_paper.rds")
data_before_filtering <-  data[seqnames(data) == "X", ]

genes_keep <- readRDS("./ProcessedData/genes_keep.rds")
data <- data[genes_keep, ]

data_with_autosomes <- data
data <- data[seqnames(data) == "X", ]

data_test <- data[,data$Clone == "E6"]

table(data_test$Experiment, data_test$ConditionClean)

```

```{r ssss}

### This is for Figure 3, the reversibility aspect
### Now we look at Xist overexpression and how that silences escapees
samples_use <- c("Control", 

                 "Dox (7d)", 
                 "Dox (7d) - washout (7d)", 

                 "Dox (14d)",
                 "Dox (14d) - washout (7d)",

                 "Dox (21d)", 
                 "Dox (21d) - washout (7d)"
)

data_here <- data[,data$ConditionClean %in% samples_use]
data_here <- data_here[,data_here$Clone == "E6"]
data_here <- data_here[,!data_here$Experiment %in% c("September2022", "October2023")]

ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))
ratios <- ratios[apply(ratios, 1, function(x){!any(is.na(x))}), ]

# genes_out <- names(rowMeans(ratios[,data_here$ConditionClean == "Control"], na.rm = T)[rowMeans(ratios, na.rm = T) > 0.8 & rownames(ratios) != "Xist"])
# ratios <- ratios[!rownames(ratios) %in% genes_out, ]

# define escapees
npc_escape <- read.csv("./ProcessedData/excel_files/escape_metaanalysis_final.escape_per_clone.csv")
npc_escape$E6_merged <- ifelse(npc_escape$E6...Rep1 == "Escapee" & npc_escape$E6...Rep2 == "Escapee" & npc_escape$E6...Rep3 == "Escapee",
                               "Escapee", "Silenced")
e6_escapees <- npc_escape %>% dplyr::filter(E6_merged == "Escapee") %>% pull(symbol)
escapees <- escapees_total <- e6_escapees
# escapees <- apply(ratios[,data_here$ConditionClean == "Control"], 2, function(x){x > 0.1})
# escapees_per_clone <- apply(escapees, 2, function(x){rownames(escapees)[x]})
# escapees_per_clone <- lapply(escapees_per_clone, function(x){x[x != "Xist"]})
# escapees_total <- Reduce("intersect", escapees_per_clone)

# For Xist-expression, we need a scaling factor per sample to account for differences in seq depth
size_factors <- colSums(counts(data_here)) / colSums(counts(data_here))[[1]]

# Supplement
xist_expression_levels <- data.frame(
  Dox = data_here$ndDox,
  Washout = data_here$ndWashout, 
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6)
)

mean_baseline_xist <- xist_expression_levels %>% dplyr::filter(Dox == 0) %>% pull(Expression) %>% mean()

data.frame(
  Dox = data_here$ndDox,
  Washout = data_here$ndWashout, 
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ])) / colSums(counts(data_here)) * 1e6 / mean_baseline_xist
) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7", "14", "21")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>%
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
  mutate(Condition = factor(Condition, levels = rev(c("Control", "0 days washout", "4 days washout", "7 days washout", "14 days washout")))) -> df_plot_6b

df_plot_6b %>%
  mutate(TimeSeries = factor(TimeSeries, levels = c("3 days dox treatment", "7 days dox treatment", "14 days dox treatment", "21 days dox treatment"))) %>%
  ggplot(aes(x = Condition, y = Expression)) + 
    stat_summary(stat = "mean", geom = "bar", fill = "grey", col = "black") + 
    coord_flip() + theme_paper() + 
    xlab("") + ylab("Expression (CPM)") + geom_jitter(size = 3, width = 0.1) + 
    scale_y_continuous(expand = c(0, 0), limits = c(0, 12)) + 
    facet_wrap(~TimeSeries, nrow = 1, scales = "free_y") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggplot2::ggsave("./Plots/FigS3/Washout_Xist_barplots.pdf", height = 4, width = 35)

ratios %>%
  t() %>% data.frame() %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Washout" = data_here$ndWashout) %>%
  pivot_longer(-c(Dox, Washout)) %>%
  dplyr::filter(name %in% escapees) %>%
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0("Dox (", .$Dox, " days), Washout (", .$Washout, " days)"))) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  dplyr::select(-c(Dox, Washout)) %>%
  group_by(name, Condition) %>%
  summarize(value = mean(value), escape_status = unique(escape_status)) %>%
  ungroup() %>%
  pivot_wider(values_from = value, names_from = Condition) %>%
  mutate_at(vars(matches("Dox")), list(dif = ~ . - Control)) %>%
  mutate(start = start(rowRanges(data_here[.$name, ])), .after = name) %>%
  mutate(end = end(rowRanges(data_here[.$name, ])), .after = start) %>%
  mutate(chromosome = "chrX", .before = "start") %>%
  write.csv("./ProcessedData/excel_files/figure3_d_score_differences.csv")

```

We structure the analysis in this figure as follows: 

- Overview silencing in a heatmap - check
- Short silencing -- largely reversible. Confirm with CL30/CL31 in the supplement - check
- Long silencing -- partly reversible - check
- Annotate genes based on short term silencing + reversibility and long term silencing + reversibility


```{r }

genes_show <- c("Xist", "Mecp2", "Kdm6a", "Kdm5c")
genes_of_choice <- setNames(rep("", nrow(ratios)), rownames(ratios))
genes_of_choice[genes_show] <- genes_show

ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Replicate, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Condition, Clone, Dox)) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  mutate(Condition = factor(Condition, levels = rev(samples_use))) %>%
  mutate(TimePoint = case_when(
    grepl("Dox \\(3d\\)", Condition)  ~ "3d",
    grepl("Dox \\(7d\\)", Condition)  ~ "7d",
    grepl("Dox \\(14d\\)", Condition)  ~ "14d",
    grepl("Dox \\(21d\\)", Condition)  ~ "21d",
    grepl("Control", Condition)  ~ "0d")) %>% 
  mutate(TimePoint = factor(TimePoint, levels = paste0(c(0, 3, 7, 14, 21), "d"))) -> to_plot

to_plot %>%
  ggplot(aes(reorder(name, position), y = reorder(paste0(Condition, "_", Clone), Condition), fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_y_discrete(labels = genes_of_choice, position = "right") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    scale_x_discrete(labels = genes_of_choice, position = "bottom") +
    labs(fill = "d-score") + facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") +
    theme(strip.background = element_blank(), strip.text.y = element_blank())
ggplot2::ggsave("./Plots/FigS3/Washout_full_heatmap_allsamples.pdf")

ratios[escapees_total, ] %>%
  t() %>% data.frame() %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Washout" = data_here$ndWashout) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  dplyr::filter(Washout != 4) %>% ### not sure about this yet -- but plot doesnt look great keeping it in
  pivot_longer(-c(Dox, Washout, Clone)) %>%
  group_by(Clone) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("7", "14", "21")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>%
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
  mutate(Condition = gsub("(0 days washout)", "Dox", Condition)) %>%
  mutate(Condition = factor(Condition, levels = c("Control", "Dox", "4 days washout", "7 days washout", "14 days washout"))) %>%
  # dplyr::filter(TimeSeries %in% c("3 days dox treatment", "7 days dox treatment")) %>%
  add_column(escape_status = factor(unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation), levels = names(escape_colors))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  mutate(TimeSeries = factor(TimeSeries, levels = paste0(c(7, 14, 21), " days dox treatment"))) %>%
  mutate(Condition = factor(Condition, levels = rev(c("Control", "Dox", "4 days washout", "7 days washout", "14 days washout")))) %>%
  ggplot(aes(x = Condition, y = value, fill = escape_status)) + geom_boxplot(col = "black") + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    facet_grid(cols = vars(TimeSeries), scales = "free_x", space = "free_x") + 
    scale_fill_manual(values = escape_colors) + 
    labs(colour = "Escape status") + coord_flip()
ggplot2::ggsave("./Plots/Fig3/Washout_escape_boxplots.pdf", width = 14, height = 6)

```

```{r plot_single_genes}
### Large heatmap showing everything

samples_use <- c("Control", 
                 "Dox (3d)", 
                 #"Dox (3d) - washout (4d)", 
                 "Dox (7d)", 
                 #"Dox (7d) - washout (4d)", 
                 "Dox (7d) - washout (7d)", 
                 "Dox (14d)", "Dox (14d) - washout (7d)", 
                 "Dox (21d)", "Dox (21d) - washout (7d)", 
                 
                 "Dox (7d) - washout (14d)", 
                 
                  "Dox (14d) - washout (14d)", 
                 
                  "Dox (21d) - washout (14d)"
)

data_here <- data_here[,data_here$ConditionClean %in% samples_use]
data_here <- data_here[,data_here$Clone == "E6"]

ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))
ratios <- ratios[apply(ratios, 1, function(x){!any(is.na(x))}), ]

clustering_coordinates <- readRDS("./ProcessedData/escapee_clustering.rds")
clusterings <- clustering_coordinates %>% pull(clustering, name = gene)

ratios[e6_escapees, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Condition, Clone)) %>%
  mutate(name = factor(name, levels = unique(.$name))) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  dplyr::filter(!grepl("washout \\(4d", Condition)) %>% # not sure about this rn -- but looks weird including the 4d timepoint
  mutate(Condition = factor(Condition, levels = rev(samples_use))) -> to_plot

to_plot -> df_plot_6d

p1 <- to_plot %>%
  mutate(Timepoint = str_extract(Condition, "Control|(Dox \\([0-9]*d\\))")) %>% 
  mutate(Timepoint = factor(Timepoint, c("Control", "Dox (7d)", "Dox (14d)", "Dox (21d)"))) %>%
  ggplot(aes(reorder(name, position),  y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    labs(fill = "d-score") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 10)) +
    scale_x_discrete(position = "top") + guides(fill = "none") + 
    facet_grid(rows = vars(Timepoint), scales = "free", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.x = element_blank()) + 
    labs(col = "Allelic Ratio \n (Xi / (Xi + Xa))")

# Plot escapee annotation at the bottom of this:
p2 <- data.frame(Gene = e6_escapees,
           Annotation = rowData(data)[e6_escapees, ]$EscapeAnnotation,
           Position = start(rowRanges(data)[e6_escapees, ])) %>%
  ggplot(aes(x = reorder(Gene, Position), y = 1, fill = Annotation)) + 
    theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) +
    theme(legend.position = "None")

fill_colors_here <- c(RColorBrewer::brewer.pal(name = "Paired", n = 11), "white")

p3 <- data.frame(Gene = e6_escapees,
                 Clusterings = as.character(clusterings[e6_escapees]), 
                 Position = start(rowRanges(data)[e6_escapees, ])) %>%
  group_by(Clusterings) %>% mutate(ngenes = n()) %>%
  mutate(Clusterings = ifelse(ngenes < 3, "Singleton", Clusterings)) %>%
  mutate(Clusterings = ifelse(is.na(Clusterings), "Singleton", Clusterings)) %>%
  ggplot(aes(x = reorder(Gene, Position), y = "Local gene groups", fill = Clusterings)) + 
    theme_void() + geom_tile(col = "white") +
    theme(legend.position = "None") + theme(axis.text.y = element_text()) + 
    scale_fill_manual(values = fill_colors_here)

p1 / p2 / p3 + plot_layout(heights = c(4, 0.2, 0.2))
ggplot2::ggsave("./Plots/Fig3/heatmap_ordered_only_escapees_by_position_reversibility.pdf", width = 24, height = 5)

```

```{r scatterplots}

### Now, we focus on the late timepoints: 

samples_use <- c("Control", 
                 "Dox (7d)", "Dox (7d) - washout (7d)", 
                 "Dox (14d)", "Dox (14d) - washout (7d)", 
                 "Dox (21d)", "Dox (21d) - washout (7d)"
)

data_here <- data_here[,data_here$ConditionClean %in% samples_use]
data_here <- data_here[,data_here$Clone == "E6"]

# general_escape_genes <- Reduce("intersect", escapees_per_clone)
general_escape_genes <- e6_escapees
general_escapees <- data_here[general_escape_genes, ]

data_test_inactive <- counts_inactive(general_escapees)
data_test_active <- counts_active(general_escapees)

fitting_metadata_here <- colData(general_escapees) %>% data.frame() %>% mutate(washout = gsub("NA", "NO", washout))
fitting_metadata_here$washout <- ifelse(fitting_metadata_here$washout == "NO", "NO", paste0(fitting_metadata_here$washout, fitting_metadata_here$ndDox))

genes_out <- c()
data_test_inactive <- data_test_inactive[!rownames(data_test_inactive) %in% genes_out, ]
data_test_active <- data_test_active[!rownames(data_test_active) %in% genes_out, ]

all_coefs <- data.frame(do.call("rbind", lapply(1:nrow(data_test_active), function(i){
  tryCatch({
    y = as.numeric(data_test_inactive[i, ]) + 1
    N = as.numeric(data_test_active[i, ] + data_test_inactive[i, ]) + 2
    fit1 <- glm(cbind(y, N-y) ~ ndDox + washout, family = "binomial", data = fitting_metadata_here)
    coefs <- coef(fit1)
    coefs
    return(coefs)
  }, error = function(cond) {
    return(c(NA))
  }, warning = function(cond){
    return(c(NA))
  })
})))

rownames(all_coefs) <- rownames(data_test_active)
all_coefs <- all_coefs[!apply(all_coefs, 1, function(x){all(is.na(x))}), !apply(all_coefs, 2, function(x){all(is.na(x))})]

row_annotation <- data.frame(
  Gene = rownames(all_coefs), 
  escapee_annotation = as.character(rowData(data_here)[match(rownames(all_coefs), rowData(data_here)$gene_name), 7])
) %>% column_to_rownames("Gene")

# We now convert the coefficients to fold changes in escape: 
# 
df_test <- data.frame(
  Gene = rownames(all_coefs),
  control = exp(all_coefs$X.Intercept.) / (exp(all_coefs$X.Intercept.) + 1),
  nddox7_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox7) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox7) + 1),
  nddox14_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox14) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox14) + 1),
  nddox21_treat = exp(all_coefs$X.Intercept. + all_coefs$ndDox21) / (exp(all_coefs$X.Intercept. + all_coefs$ndDox21) + 1),
  nddox7_washout = exp(all_coefs$X.Intercept. + all_coefs$ndDox7 + all_coefs$washoutYES7) /
    (exp(all_coefs$X.Intercept. + all_coefs$ndDox7 + all_coefs$washoutYES7) + 1),
  nddox14_washout = exp(all_coefs$X.Intercept. + all_coefs$ndDox14 + all_coefs$washoutYES14) /
    (exp(all_coefs$X.Intercept. + all_coefs$ndDox14 + all_coefs$washoutYES14) + 1),
  nddox21_washout = exp(all_coefs$X.Intercept. + all_coefs$ndDox21 + all_coefs$washoutYES21) /
    (exp(all_coefs$X.Intercept. + all_coefs$ndDox21 + all_coefs$washoutYES21) + 1)
)

```

Now we save our classification based on the 21d analysis to use in further analysis: 

```{r save_classification}

saveRDS(df_test, "./ProcessedData/reversibility_21d.rds")

### look at number of genes by reversibility (control / washout)
df_test -> df_plot_6e
df_test %>% 
  dplyr::select(c("Gene", "control", "nddox7_washout", "nddox14_washout", "nddox21_washout")) %>%
  mutate_at(vars(matches("dox")), list(ratio = ~ . / control)) %>% 
  dplyr::select(c("Gene", "nddox7_washout_ratio", "nddox14_washout_ratio", "nddox21_washout_ratio")) %>%
  pivot_longer(-c("Gene")) %>% 
  mutate(name = gsub("nd", "", gsub("_washout_ratio", "", name))) %>%
  mutate(name = gsub("dox", "Dox", name)) %>%
  mutate(name = factor(name, levels = paste0("Dox", c(7, 14, 21)))) %>%
  add_column(EscapeAnnotation = rowData(data[.$Gene, ])$EscapeAnnotation) %>%
  ggplot(aes(x = name, y = value, fill = EscapeAnnotation)) + 
    geom_boxplot(outlier.color = "black") + scale_fill_manual(values = escape_colors) + 
    theme_paper(textsize = 30) + xlab("") + ylab("Reversibility (Allelic Ratio Washout / Control)")
ggplot2::ggsave("./Plots/Fig3/reversibility_fraction.pdf", width = 15, height = 10)

### pvals
df_test %>% 
  dplyr::select(c("Gene", "control", "nddox7_washout", "nddox14_washout", "nddox21_washout")) %>%
  mutate_at(vars(matches("dox")), list(ratio = ~ . / control)) %>% 
  dplyr::select(c("Gene", "nddox7_washout_ratio", "nddox14_washout_ratio", "nddox21_washout_ratio")) %>%
  pivot_longer(-c("Gene")) %>% 
  mutate(name = gsub("nd", "", gsub("_washout_ratio", "", name))) %>%
  mutate(name = gsub("dox", "Dox", name)) %>%
  mutate(name = factor(name, levels = paste0("Dox", c(7, 14, 21)))) %>%
  add_column(EscapeAnnotation = rowData(data[.$Gene, ])$EscapeAnnotation) %>%
  dplyr::rename(Condition = name, escape_status = EscapeAnnotation) %>%
  group_by(Condition, escape_status) %>%
  summarize(values = list(value)) %>%
  ungroup() %>%
  tidyr::expand(nesting(Condition, escape_status, values), nesting(Condition, escape_status, values),  .name_repair = "unique") %>%
  setNames(tolower(gsub("\\.","", names(.)))) %>%
  dplyr::filter((condition1 != condition4) & (escape_status2 == escape_status5)) %>%
  rowwise() %>%
  mutate(condition1 = sort(c(condition1, condition4))[1], condition4 = sort(c(condition1, condition4))[2]) %>% dplyr::filter(condition1 != condition4) %>%
  mutate(pvals = wilcox.test(values3, values6)$p.value, paired = T) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvals, method = "BH")) %>% 
  dplyr::select(-c("values3", "values6")) -> pval_results
write.csv(pval_results, "./ProcessedData/pvals_fig3e.csv")

df_test %>% 
  dplyr::select(c("Gene", "control", "nddox7_washout", "nddox14_washout", "nddox21_washout")) %>%
  mutate_at(vars(matches("dox")), list(ratio = ~ . / control)) %>% 
  dplyr::select(c("Gene", "nddox7_washout_ratio", "nddox14_washout_ratio", "nddox21_washout_ratio")) %>%
  pivot_longer(-c("Gene")) %>% 
  mutate(name = gsub("nd", "", gsub("_washout_ratio", "", name))) %>%
  mutate(name = gsub("dox", "Dox", name)) %>%
  mutate(name = factor(name, levels = paste0("Dox", c(7, 14, 21)))) %>%
  add_column(EscapeAnnotation = rowData(data[.$Gene, ])$EscapeAnnotation) %>%
  mutate(reversibility_category = case_when( ### think about these cutoffs
    value > .5 ~ "Reversible", 
    value < .1 ~ "Irreversible",
    .default = "Partially Reversible" 
  )) -> save_data_three_way
  
color_categories_here <- c("Irreversible" = "black", "Partially Reversible" = "grey", "Reversible" = "white")

# save_data_three_way %>%  
#   ggplot(aes(x = name, fill = reversibility_category)) + geom_bar(col = "black") + 
#     theme_bw(base_size = 30) + xlab("") + ylab("Number of genes") + 
#     scale_fill_manual(values = color_categories_here) + 
#     labs(fill = "Reversibility category")
# ggplot2::ggsave("./Plots/Fig3/reversibility_fraction_numbers_3_categories.pdf", width = 12, height = 8)

df_test %>% 
  dplyr::select(c("Gene", "nddox7_washout", "nddox14_washout", "nddox21_washout")) %>%
  pivot_longer(-c("Gene")) %>% 
  mutate(name = gsub("nd", "", gsub("_washout", "", name))) %>%
  mutate(name = gsub("dox", "Dox", name)) %>%
  mutate(name = factor(name, levels = paste0("Dox", c(7, 14, 21)))) %>%
  add_column(EscapeAnnotation = rowData(data[.$Gene, ])$EscapeAnnotation) %>%
  mutate(reversibility_category = ifelse(value > .1, "Reversible", "Irreversible")) -> save_data_tenpercent_rule

# save_data_tenpercent_rule %>%
#   ggplot(aes(x = name, fill = reversibility_category)) + geom_bar(col = "black") + 
#     theme_bw(base_size = 30) + xlab("") + ylab("Number of genes") + 
#     scale_fill_manual(values = color_categories_here) + 
#     labs(fill = "Reversibility category")
# ggplot2::ggsave("./Plots/Fig3/reversibility_fraction_numbers_3_tenpercent_rule.pdf", width = 12, height = 8)

## use both pieces of information: 
save_data_three_way %>% right_join(save_data_tenpercent_rule, by = c("Gene", "name")) %>% 
  dplyr::rename("Ratio" = "value.x", "Escape" = "value.y") %>%
  mutate(reversibility_category_joint = case_when(
    Ratio > .5 & Escape > .1 ~ "Reversible", 
    (Ratio < .5 & Ratio > .1) & Escape > .1 ~ "Partially Reversible", 
    .default = "Irreversible" 
  )) -> save_data_joint

# save_data_joint %>%
#   ggplot(aes(x = name, fill = reversibility_category_joint)) + geom_bar(col = "black") + 
#     theme_bw(base_size = 30) + xlab("") + ylab("Number of genes") + 
#     scale_fill_manual(values = rev(c("grey", "#325266", "#000922"))) + 
#     labs(fill = "Reversibility category") + theme_paper()
# ggplot2::ggsave("./Plots/Fig3/reversibility_fraction_numbers_3_tenpercent_rule.pdf", width = 12, height = 8)

save_data_joint %>%
  mutate(reversibility_category_joint = factor(reversibility_category_joint, c("Irreversible", "Partially Reversible", "Reversible"))) %>%
  mutate(name = factor(name, rev(c("Dox7", "Dox14", "Dox21")))) -> df_plot_6f

df_plot_6f %>%
  ggplot(aes(x = name, fill = reversibility_category_joint)) + geom_bar(col = "black") +
    theme_bw(base_size = 30) + xlab("") + ylab("Number of genes") +
    scale_fill_manual(values = rev(c("pink1", "orchid4", "darkolivegreen2"))) +
    labs(fill = "Reversibility category") + theme_paper() +
    facet_wrap(~EscapeAnnotation.x, scales = "free", nrow = 3) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
    coord_flip()
ggplot2::ggsave("./Plots/Fig3/reversibility_fraction_numbers_3_tenpercent_rule_split.pdf", width = 12, height = 8)

table(save_data_joint$name, save_data_joint$reversibility_category_joint)
table(save_data_joint$name, save_data_joint$reversibility_category_joint, save_data_joint$EscapeAnnotation.x)

# matters very little which definition we use --> i prefer the 3-way one
table(save_data_three_way$reversibility_category, save_data_tenpercent_rule$reversibility_category, save_data_tenpercent_rule$name)

save_data_joint %>% 
  # dplyr::filter(name == "Dox7") %>%
  group_by(name, reversibility_category_joint) %>%
  summarize(n = n())

save_data_joint %>% 
  group_by(name, reversibility_category_joint, EscapeAnnotation.x) %>%
  summarize(n = n()) %>% dplyr::filter(name == "Dox14")

```

Compare reversibility to silencing speed: 

```{r adasdassssss}

silencing_halftime_data <- readRDS("./Data/halflife_fits.rds")

# df_test %>% 
#   mutate(reversibility_index = nddox21_washout / control) %>%
#   dplyr::rename("gene" = "Gene") %>%
#   right_join(silencing_halftime_data) %>% 
#   dplyr::select(c(gene, reversibility_index, halflifes_off, category)) %>% 
#   ggplot(aes(x = halflifes_off, y = reversibility_index)) + 
#     geom_point(aes(fill = category), size = 5, pch = 21) + 
#     theme_paper(textsize = 20) + 
#     scale_fill_manual(values = escape_colors) + 
#     ylab("Fold recovery after washout") + xlab("Halflife of silencing") + 
#     ggpubr::stat_cor(size = 8) + 
#     geom_hline(yintercept = .1, linetype = "dashed", col = "darkgrey") + 
#     annotate(geom = "text", x = 8, y = .13, label = "partially reversible", color = "darkgrey", size = 6) + 
#     geom_hline(yintercept = .5, linetype = "dashed", col = "darkgrey") + 
#     annotate(geom = "text", x = 8, y = .53, label = "reversible", color = "darkgrey", size = 6) + 
#     annotate(geom = "text", x = 8, y = .03, label = "irreversible", color = "darkgrey", size = 6)
# ggplot2::ggsave("./Plots/Fig3/silencing_speed_recovery.pdf", width = 12, height = 8)

### overlay the cluster identities here

gene_groups <- readr::read_csv("./ProcessedData/multigene_grouping_data.csv") %>% 
  mutate(genes = lapply(genes, function(x){str_split(x, ",")[[1]]}))
gene_groups %>% unnest_longer(genes) %>% dplyr::select(c(clustering, genes)) -> cluster_assignments

colors_here <- c("Others" = "grey", "Region 11" =  "#6A3D9A", "Region 6" = "#33A02C", "Region 7" = "#FB9A99", 
                 "Regions 5, 8, 9" = "black", "Region 1" =  "#B2DF8A")

## "#A6CEE3" "#1F78B4" "#B2DF8A" "#33A02C" "#FB9A99" "#E31A1C" "#FDBF6F" "#FF7F00" "#CAB2D6" "#6A3D9A" "#FFFF99" "white"  

df_test %>% 
  mutate(reversibility_index = nddox21_washout / control) %>%
  dplyr::rename("gene" = "Gene") %>%
  right_join(silencing_halftime_data) %>%
  dplyr::select(c(gene, reversibility_index, halflifes_off, category)) %>% 
  dplyr::rename("genes" = "gene") %>%
  right_join(cluster_assignments, by = "genes") %>%
  mutate(region_category = case_when(
    clustering == 11 ~ "Region 11", 
    # clustering %in% c(5, 8, 9) ~ "Regions 5, 8, 9", 
    clustering == 6 ~ "Region 6", 
    clustering == 7 ~ "Region 7", 
    clustering == 1 ~ "Region 1", 
    .default = "Others"
  )) -> df_plot_6g

df_plot_6g %>%
  ggplot(aes(x = halflifes_off, y = reversibility_index)) + 
    geom_point(aes(fill = region_category), size = 5, pch = 21) + 
    theme_paper(textsize = 30) + 
    scale_fill_manual(values = colors_here) + 
    ylab("Fold recovery after washout") + xlab("Halflife of silencing") + 
    ggpubr::stat_cor(size = 8) + 
    geom_hline(yintercept = .1, linetype = "dashed", col = "darkgrey") + 
    annotate(geom = "text", x = 8, y = .13, label = "partially reversible", color = "darkgrey", size = 6) + 
    geom_hline(yintercept = .5, linetype = "dashed", col = "darkgrey") + 
    annotate(geom = "text", x = 8, y = .53, label = "reversible", color = "darkgrey", size = 6) + 
    annotate(geom = "text", x = 8, y = .03, label = "irreversible", color = "darkgrey", size = 6) + 
    labs(fill = "Region")
ggplot2::ggsave("./Plots/Fig3/silencing_speed_recovery_clusters.pdf", width = 12, height = 8)

```


### BLOCK

```{r ssss}

### This is for Figure 3, the reversibility aspect
### Now we look at Xist overexpression and how that silences escapees
samples_use <- c("Control", 

                 "Dox (7d)", 
                 "Dox (7d) - washout (7d)", 
                 "Dox (7d) - washout (14d)", 
                 "Dox (7d) - washout (21d)", 

                 "Dox (14d)",
                 "Dox (14d) - washout (7d)",
                 "Dox (14d) - washout (14d)",
                 "Dox (14d) - washout (21d)",

                 "Dox (21d)", 
                 "Dox (21d) - washout (7d)", 
                 "Dox (21d) - washout (14d)", 
                 "Dox (21d) - washout (21d)"
)

data_here <- data[,data$ConditionClean %in% samples_use]
data_here <- data_here[,data_here$Clone == "E6"]
data_here <- data_here[,!data_here$Experiment == "September2022"]
data_here <- data_here[,!data_here$Experiment %in% c("September2022", "October2023")]
data_here <- data_here[,!data_here$Replicate == "Rep2b"]

ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))
ratios <- ratios[apply(ratios, 1, function(x){!any(is.na(x))}), ]

genes_out <- names(rowMeans(ratios[,data_here$ConditionClean == "Control"], na.rm = T)[rowMeans(ratios, na.rm = T) > 0.8 & rownames(ratios) != "Xist"])

ratios <- ratios[!rownames(ratios) %in% genes_out, ]

# define escapees
npc_escape <- read.csv("./ProcessedData/excel_files/escape_metaanalysis_final.escape_per_clone.csv")
npc_escape$E6_merged <- ifelse(npc_escape$E6...Rep1 == "Escapee" & npc_escape$E6...Rep2 == "Escapee" & npc_escape$E6...Rep3 == "Escapee",
                               "Escapee", "Silenced")
e6_escapees <- npc_escape %>% dplyr::filter(E6_merged == "Escapee") %>% pull(symbol)
escapees <- escapees_total <- e6_escapees
# escapees <- apply(ratios[,data_here$ConditionClean == "Control"], 2, function(x){x > 0.1})
# escapees_per_clone <- apply(escapees, 2, function(x){rownames(escapees)[x]})
# escapees_per_clone <- lapply(escapees_per_clone, function(x){x[x != "Xist"]})
# escapees_total <- Reduce("intersect", escapees_per_clone)

# For Xist-expression, we need a scaling factor per sample to account for differences in seq depth
size_factors <- colSums(counts(data_here)) / colSums(counts(data_here))[[1]]

# Supplement
xist_expression_levels <- data.frame(
  Dox = data_here$ndDox,
  Washout = data_here$ndWashout, 
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6)
)

mean_baseline_xist <- xist_expression_levels %>% dplyr::filter(Dox == 0) %>% pull(Expression) %>% mean()

data.frame(
  Dox = data_here$ndDox,
  Washout = data_here$ndWashout, 
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ])) / colSums(counts(data_here)) * 1e6 / mean_baseline_xist
) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7", "14", "21")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>%
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
  mutate(Condition = factor(Condition, levels = rev(c("Control", "0 days washout", "4 days washout", "7 days washout", "14 days washout")))) %>%
  mutate(TimeSeries = factor(TimeSeries, levels = c("3 days dox treatment", "7 days dox treatment", "14 days dox treatment", "21 days dox treatment"))) %>%
  ggplot(aes(x = Condition, y = Expression)) + 
    stat_summary(stat = "mean", geom = "bar", fill = "grey", col = "black") + 
    coord_flip() + theme_paper() + 
    xlab("") + ylab("Expression (CPM)") + geom_jitter(size = 3, width = 0.1) + 
    scale_y_continuous(expand = c(0, 0), limits = c(0, 12)) + 
    facet_wrap(~TimeSeries, ncol = 1, scales = "free_y") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
#ggplot2::ggsave("./Plots/FigS3/Washout_Xist_barplots_long_e6.pdf", height = 8, width = 6)

```

We structure the analysis in this figure as follows: 

- Overview silencing in a heatmap - check
- Short silencing -- largely reversible. Confirm with CL30/CL31 in the supplement - check
- Long silencing -- partly reversible - check
- Annotate genes based on short term silencing + reversibility and long term silencing + reversibility


```{r }

genes_show <- c("Xist", "Mecp2", "Kdm6a", "Kdm5c")
genes_of_choice <- setNames(rep("", nrow(ratios)), rownames(ratios))
genes_of_choice[genes_show] <- genes_show

ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Replicate, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Condition, Clone, Dox)) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  mutate(Condition = factor(Condition, levels = rev(samples_use))) %>%
  mutate(TimePoint = case_when(
    grepl("Dox \\(3d\\)", Condition)  ~ "3d",
    grepl("Dox \\(7d\\)", Condition)  ~ "7d",
    grepl("Dox \\(14d\\)", Condition)  ~ "14d",
    grepl("Dox \\(21d\\)", Condition)  ~ "21d",
    grepl("Control", Condition)  ~ "0d")) %>% 
  mutate(TimePoint = factor(TimePoint, levels = paste0(c(0, 3, 7, 14, 21), "d"))) -> to_plot

to_plot %>%
  group_by(Condition, TimePoint, name, position) %>%
  summarize(value = mean(value)) %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_y_discrete(labels = genes_of_choice, position = "right") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    scale_x_discrete(labels = genes_of_choice, position = "bottom") + 
    labs(fill = "d-score") +  facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.y = element_blank())
#ggplot2::ggsave("./Plots/FigS3/Washout_full_heatmap_long_e6.pdf")

to_plot %>%
  ggplot(aes(reorder(name, position), y = reorder(paste0(Condition, "_", Clone), Condition), fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_y_discrete(labels = genes_of_choice, position = "right") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    scale_x_discrete(labels = genes_of_choice, position = "bottom") +
    labs(fill = "d-score") + facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") +
    theme(strip.background = element_blank(), strip.text.y = element_blank())
# ggplot2::ggsave("./Plots/FigS3/Washout_full_heatmap_long_e6.pdf")

# ratios[e6_escapees, ] %>%
#   t() %>% data.frame(check.names = F) %>%
#   add_column("Condition" = data_here$ConditionClean) %>%
#   add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
#   pivot_longer(-c(Condition, Clone)) %>%
#   mutate(name = factor(name, levels = unique(.$name))) %>%
#   add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
#   dplyr::filter(!grepl("washout \\(4d", Condition)) %>% # not sure about this rn -- but looks weird including the 4d timepoint
#   mutate(Condition = factor(Condition, levels = rev(samples_use))) -> to_plot

p1 <- to_plot %>%
  dplyr::filter(name %in% e6_escapees) %>%
  mutate(Timepoint = str_extract(Condition, "Control|(Dox \\([0-9]*d\\))")) %>% 
  mutate(Timepoint = factor(Timepoint, c("Control", "Dox (7d)", "Dox (14d)", "Dox (21d)"))) %>%
  group_by(Timepoint, Condition, name, position) %>%
  summarize(value = mean(value)) %>%
  ggplot(aes(reorder(name, position),  y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    labs(fill = "d-score") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 10)) +
    scale_x_discrete(position = "top") + guides(fill = "none") + 
    facet_grid(rows = vars(Timepoint), scales = "free", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.x = element_blank()) + 
    labs(col = "Allelic Ratio \n (Xi / (Xi + Xa))")

# Plot escapee annotation at the bottom of this:
p2 <- data.frame(Gene = e6_escapees,
           Annotation = rowData(data)[e6_escapees, ]$EscapeAnnotation,
           Position = start(rowRanges(data)[e6_escapees, ])) %>%
  ggplot(aes(x = reorder(Gene, Position), y = 1, fill = Annotation)) + 
    theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) +
    theme(legend.position = "None")

fill_colors_here <- c(RColorBrewer::brewer.pal(name = "Paired", n = 11), "white")

p3 <- data.frame(Gene = e6_escapees,
                 Clusterings = as.character(clusterings[e6_escapees]), 
                 Position = start(rowRanges(data)[e6_escapees, ])) %>%
  group_by(Clusterings) %>% mutate(ngenes = n()) %>%
  mutate(Clusterings = ifelse(ngenes < 3, "Singleton", Clusterings)) %>%
  mutate(Clusterings = ifelse(is.na(Clusterings), "Singleton", Clusterings)) %>%
  ggplot(aes(x = reorder(Gene, Position), y = "Local gene groups", fill = Clusterings)) + 
    theme_void() + geom_tile(col = "white") +
    theme(legend.position = "None") + theme(axis.text.y = element_text()) + 
    scale_fill_manual(values = fill_colors_here)

p1 / p2 / p3 + plot_layout(heights = c(4, 0.2, 0.2))
#ggplot2::ggsave("./Plots/Fig3/heatmap_ordered_only_escapees_by_position_reversibility_with14d.pdf", width = 24, height = 5)

ratios[escapees_total, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Washout" = data_here$ndWashout) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  dplyr::filter(Washout != 4) %>% ### not sure about this yet -- but plot doesnt look great keeping it in
  pivot_longer(-c(Dox, Washout, Clone)) %>%
  group_by(Clone) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("7", "14", "21")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>%
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
  mutate(Condition = gsub("(0 days washout)", "Dox", Condition)) %>%
  mutate(Condition = factor(Condition, levels = c("Control", "Dox", "4 days washout", "7 days washout", "14 days washout"))) %>%
  add_column(escape_status = factor(unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation), levels = names(escape_colors)))  %>%
  dplyr::filter(!is.na(escape_status)) %>%
  mutate(TimeSeries = factor(TimeSeries, levels = paste0(c(7, 14, 21), " days dox treatment"))) %>%
  mutate(Condition = factor(Condition, levels = rev(c("Control", "Dox", "4 days washout", "7 days washout", "14 days washout")))) %>%
  group_by(Dox, Washout, name, TimeSeries, Condition, escape_status) %>%
  summarize(value = mean(value)) -> data_here_test

data_here_test %>%
  ggplot(aes(x = Condition, y = value, fill = escape_status)) + geom_boxplot(col = "black") + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    scale_fill_manual(values = escape_colors) + 
    labs(colour = "Escape status") + coord_flip() + 
    ggpubr::stat_compare_means(comparisons = list(c("Control", "Dox"), c("Control", "7 days washout"), c("Control", "14 days washout"),
                                                  c("7 days washout", "14 days washout")), na.rm = T) +
    facet_grid(cols = vars(TimeSeries), scales = "free_x", space = "free_x")
#ggplot2::ggsave("./Plots/Fig3/Washout_escape_boxplots_E6_long.pdf", width = 14, height = 6)

data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  mutate(Condition = factor(Condition, levels = paste0(c(7, 14, 21), " days washout"))) %>%
  mutate(escape_status = factor(escape_status, levels = c("Constitutive", "Facultative", "NPC-specific"))) -> df_plot_s8a

df_plot_s8a %>%
  ggplot(aes(x = TimeSeries, y = washout_ratio)) + geom_boxplot(aes(fill = escape_status)) + facet_wrap(~Condition, drop = T) + 
    scale_fill_manual(values = escape_colors) + theme_paper(textsize = 25) + geom_hline(yintercept = 1, linetype = 'dashed') + xlab("") + 
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + ylab("Reversibility (Allelic Ratio Washout / Control)")
    # ggpubr::stat_compare_means(comparisons = list(c("7 days dox treatment", "14 days dox treatment"), 
    #                                               c("7 days dox treatment", "21 days dox treatment"), 
    #                                               c("14 days dox treatment", "21 days dox treatment")))
#ggplot2::ggsave("./Plots/Fig3/reversibility_fraction_long_e6.pdf", width = 14, height = 6)

data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  mutate(Condition = paste0(TimeSeries, "_", Condition)) %>%
  group_by(Condition) %>%
  summarize(values = list(washout_ratio)) %>%
  ungroup() %>%
  tidyr::expand(nesting(Condition, values), nesting(Condition, values),  .name_repair = "unique") %>%
  setNames(tolower(gsub("\\.","", names(.)))) %>%
  dplyr::filter(condition1 != condition3) %>%
  rowwise() %>%
  mutate(pvals = wilcox.test(values2, values4)$p.value, paired = T) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvals, method = "BH")) %>% 
  dplyr::select(-c("values2", "values4")) %>% dplyr::filter(!duplicated(pvals)) %>%
  write.csv("./reversibility_fraction_long_e6_pvals.csv")

data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  mutate(Condition = paste0(TimeSeries, "_", Condition)) %>%
  group_by(Condition, escape_status) %>%
  summarize(values = list(washout_ratio)) %>%
  ungroup() %>%
  tidyr::expand(nesting(Condition, escape_status, values), nesting(Condition, escape_status, values),  .name_repair = "unique") %>%
  setNames(tolower(gsub("\\.","", names(.)))) %>%
  dplyr::filter(condition1 != condition4 & escape_status2 == escape_status5) %>%
  rowwise() %>%
  mutate(pvals = wilcox.test(values3, values6)$p.value, paired = T) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvals, method = "BH")) %>%
  dplyr::select(-c("values3", "values6")) %>% dplyr::filter(!duplicated(pvals)) %>%
  write.csv("./reversibility_fraction_long_e6_pvals_escapegroups.csv")

### 
data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout")) %>%
  mutate(Condition = factor(Condition, levels = paste0(c(7, 14), " days washout"))) %>%
  mutate(escape_status = factor(escape_status, levels = c("Constitutive", "Facultative", "NPC-specific"))) %>%
  mutate(reversibility_category = case_when( ### think about these cutoffs
    washout_ratio > .5 ~ "Reversible", 
    washout_ratio < .1 ~ "Irreversible",
    .default = "Partially Reversible" 
  )) -> save_data_three_way
  
data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout")) %>%
  mutate(Condition = factor(Condition, levels = paste0(c(7, 14), " days washout"))) %>%
  mutate(escape_status = factor(escape_status, levels = c("Constitutive", "Facultative", "NPC-specific"))) %>%
  mutate(reversibility_category = ifelse(washout_ratio > .1, "Reversible", "Irreversible")) -> save_data_tenpercent_rule

## use both pieces of information: 
save_data_three_way %>% right_join(save_data_tenpercent_rule, by = c("name", "TimeSeries", "Condition")) %>% 
  dplyr::rename("Ratio" = "washout_ratio.x", "Escape" = "washout_ratio.y") %>%
  mutate(reversibility_category_joint = case_when(
    Ratio > .5 & Escape > .1 ~ "Reversible", 
    (Ratio < .5 & Ratio > .1) & Escape > .1 ~ "Partially Reversible", 
    .default = "Irreversible" 
  )) -> save_data_joint

save_data_joint %>%
  ggplot(aes(x = TimeSeries, fill = reversibility_category_joint)) + geom_bar(col = "black") +
    theme_bw(base_size = 30) + xlab("") + ylab("Number of genes") +
    scale_fill_manual(values = rev(c("pink1", "orchid4", "darkolivegreen2"))) +
    labs(fill = "Reversibility category") + theme_paper() +
    facet_wrap(~Condition + escape_status.x, scales = "free_y", nrow = 1) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
    stat_count(geom = "text", colour = "black", aes(label = ..count..), position=position_stack(vjust=0.5))
#ggplot2::ggsave("./Plots/Fig3/reversibility_genes_long_e6.pdf", width = 20, height = 12)

save_data_joint_e6 <- save_data_joint

```



SAME THINGS FOR CL30s
CL30

```{r load_data}

data <- readRDS("./ProcessedData/merged_dataset_paper.rds")
data_before_filtering <-  data[seqnames(data) == "X", ]

genes_keep <- readRDS("./ProcessedData/genes_keep.rds")
data <- data[genes_keep, ]

data_with_autosomes <- data
data <- data[seqnames(data) == "X", ]

data_test <- data[ , !data$Clone == "E6"]

table(data_test$Experiment, data_test$ConditionClean)

```

```{r ssss}

### This is for Figure 3, the reversibility aspect
### Now we look at Xist overexpression and how that silences escapees
samples_use <- c("Control", 
                 # "Dox (3d)", "Dox (3d) - washout (4d)", 
                 
                 "Dox (7d)", 
                 "Dox (7d) - washout (4d)", 
                 "Dox (7d) - washout (7d)", 

                 "Dox (14d)", 
                 "Dox (14d) - washout (4d)", 
                 "Dox (14d) - washout (7d)", 
                 "Dox (14d) - washout (14d)", 
                 "Dox (14d) - washout (21d)", 

                 "Dox (21d)", 
                 "Dox (21d) - washout (4d)",
                 "Dox (21d) - washout (7d)",
                 "Dox (21d) - washout (14d)", 
                 "Dox (21d) - washout (21d)" 
)

data_here <- data[ , data$ConditionClean %in% samples_use]
data_here <- data_here[ , grepl("CL30", data_here$Clone)]
data_here <- data_here[ , data_here$Experiment == "May2025"]
data_here <- data_here[ , !(data_here$Experiment == "May2025" & data_here$Clone == "CL30.7" & data_here$Replicate == "Rep3")]
# data_here <- data_here[,!data_here$Experiment == "September2022"]
# data_here <- data_here[,!data_here$Experiment %in% c("September2022", "October2023")]

ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))
ratios <- ratios[apply(ratios, 1, function(x){!any(is.na(x))}), ]

# define escapees
npc_escape <- read.csv("./ProcessedData/excel_files/escape_metaanalysis_final.escape_per_clone.csv")
npc_escape$E6_merged <- ifelse(npc_escape$E6...Rep1 == "Escapee" & npc_escape$E6...Rep2 == "Escapee" & npc_escape$E6...Rep3 == "Escapee",
                               "Escapee", "Silenced")
npc_escape$CL30_merged <- ifelse(npc_escape$CL30...Rep1 == "Escapee" & npc_escape$CL30...Rep2 == "Escapee" & 
                                  npc_escape$CL30...Rep4 == "Escapee" & npc_escape$CL30...Rep5 == "Escapee",
                               "Escapee", "Silenced")
npc_escape$CL31_merged <- ifelse(npc_escape$CL31...Rep1 == "Escapee" & npc_escape$CL31...Rep2 == "Escapee" & npc_escape$CL31...Rep4 == "Escapee" & 
                                  npc_escape$CL31...Rep5 == "Escapee",
                               "Escapee", "Silenced")
cl30_escapees <- npc_escape %>% dplyr::filter(CL30_merged == "Escapee") %>% pull(symbol)
cl31_escapees <- npc_escape %>% dplyr::filter(CL31_merged == "Escapee") %>% pull(symbol)
escapees <- list("CL30" = cl30_escapees, "CL31" = cl31_escapees)

# For Xist-expression, we need a scaling factor per sample to account for differences in seq depth
size_factors <- colSums(counts(data_here)) / colSums(counts(data_here))[[1]]

# Supplement
xist_expression_levels <- data.frame(
  Dox = data_here$ndDox,
  Washout = data_here$ndWashout, 
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6)
)

mean_baseline_xist <- xist_expression_levels %>% dplyr::filter(Dox == 0) %>% pull(Expression) %>% mean()

data.frame(
  Dox = data_here$ndDox,
  Washout = data_here$ndWashout, 
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ])) / colSums(counts(data_here)) * 1e6 / mean_baseline_xist
) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7", "14", "21")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>%
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
  mutate(Condition = factor(Condition, levels = 
                              rev(c("Control", "0 days washout", "4 days washout", "7 days washout", "14 days washout", "21 days washout")))) %>%
  mutate(TimeSeries = factor(TimeSeries, levels = c("3 days dox treatment", "7 days dox treatment", "14 days dox treatment", "21 days dox treatment"))) %>%
  mutate(Clone = ifelse(grepl("CL30", Clone), "CL30", "CL31")) -> df_plot_s8c

df_plot_s8c %>%
  ggplot(aes(x = Condition, y = Expression)) + 
    stat_summary(stat = "mean", geom = "bar", fill = "grey", col = "black") + 
    coord_flip() + theme_paper() + 
    xlab("") + ylab("Expression (CPM)") + geom_jitter(size = 3, width = 0.1, aes(shape = Clone)) + 
    scale_y_continuous(expand = c(0, 0), limits = c(0, 13)) + 
    facet_wrap(~TimeSeries, ncol = 1, scales = "free_y") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggplot2::ggsave("./Plots/FigS3/Washout_Xist_barplots_cl30.pdf", height = 8, width = 6)

```

```{r }

genes_show <- c("Xist", "Mecp2", "Kdm6a", "Kdm5c")
genes_of_choice <- setNames(rep("", nrow(ratios)), rownames(ratios))
genes_of_choice[genes_show] <- genes_show

ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Replicate, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Condition, Clone, Dox)) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  mutate(Condition = factor(Condition, levels = rev(samples_use))) %>%
  mutate(TimePoint = case_when(
    grepl("Dox \\(3d\\)", Condition)  ~ "3d",
    grepl("Dox \\(7d\\)", Condition)  ~ "7d",
    grepl("Dox \\(14d\\)", Condition)  ~ "14d",
    grepl("Dox \\(21d\\)", Condition)  ~ "21d",
    grepl("Control", Condition)  ~ "0d")) %>% 
  mutate(TimePoint = factor(TimePoint, levels = paste0(c(0, 3, 7, 14, 21), "d"))) -> to_plot

to_plot %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_y_discrete(labels = genes_of_choice, position = "right") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    scale_x_discrete(labels = genes_of_choice, position = "bottom") + 
    labs(fill = "d-score") +  facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.y = element_blank())
ggplot2::ggsave("./Plots/FigS3/Washout_full_heatmap_cl30.pdf")

to_plot %>%
  mutate(Replicate = str_extract(Clone, "Rep[0-9]")) %>%
  ggplot(aes(reorder(name, position), y = reorder(paste0(Condition, "_", Clone), Condition), fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_y_discrete(labels = genes_of_choice, position = "right") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    scale_x_discrete(labels = genes_of_choice, position = "bottom") +
    labs(fill = "d-score") + facet_grid(rows = vars(TimePoint, Replicate), scales = "free_y", space = "free_y") +
    theme(strip.background = element_blank(), strip.text.y = element_blank())
#ggplot2::ggsave("./Plots/FigS3/Washout_full_heatmap_allsamples_cl30.pdf")

p1 <- to_plot %>%
  dplyr::filter(name %in% cl31_escapees) %>%
  mutate(Timepoint = str_extract(Condition, "Control|(Dox \\([0-9]*d\\))")) %>% 
  mutate(Timepoint = factor(Timepoint, c("Control", "Dox (7d)", "Dox (14d)", "Dox (21d)"))) %>%
  ggplot(aes(reorder(name, position),  y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    labs(fill = "d-score") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 10)) +
    scale_x_discrete(position = "top") + guides(fill = "none") + 
    facet_grid(rows = vars(Timepoint), scales = "free", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.x = element_blank()) + 
    labs(col = "Allelic Ratio \n (Xi / (Xi + Xa))")

# Plot escapee annotation at the bottom of this:
p2 <- data.frame(Gene = cl31_escapees,
           Annotation = rowData(data)[cl31_escapees, ]$EscapeAnnotation,
           Position = start(rowRanges(data)[cl31_escapees, ])) %>%
  ggplot(aes(x = reorder(Gene, Position), y = 1, fill = Annotation)) + 
    theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) +
    theme(legend.position = "None")

fill_colors_here <- c(RColorBrewer::brewer.pal(name = "Paired", n = 11), "white")

p1 / p2 + plot_layout(heights = c(4, 0.2, 0.2))
ggplot2::ggsave("./Plots/Fig3/heatmap_ordered_only_escapees_by_position_reversibility_cl30.pdf", width = 24, height = 5)


ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Washout" = data_here$ndWashout) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  dplyr::filter(Washout != 4) %>% ### not sure about this yet -- but plot doesnt look great keeping it in
  pivot_longer(-c(Dox, Washout, Clone)) %>%
  group_by(Clone) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("7", "14", "21")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>%
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
  mutate(Condition = gsub("(0 days washout)", "Dox", Condition)) %>%
  mutate(Condition = factor(Condition, levels = c("Control", "Dox", "4 days washout", "7 days washout", "14 days washout", "21 days washout"))) %>%
  # dplyr::filter(TimeSeries %in% c("3 days dox treatment", "7 days dox treatment")) %>%
  add_column(escape_status = factor(unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation), levels = names(escape_colors))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  mutate(TimeSeries = factor(TimeSeries, levels = paste0(c(7, 14, 21), " days dox treatment"))) %>%
  mutate(Condition = factor(Condition, levels = rev(c("Control", "Dox", "4 days washout", "7 days washout", "14 days washout", "21 days washout")))) %>%
  mutate(Clone_Group = ifelse(grepl("CL30", Clone), "CL30", "CL31")) %>%
  group_by(Clone_Group) %>%
  dplyr::filter(name %in% escapees[[unique(Clone_Group)]]) %>%
  group_by(Dox, Washout, name, TimeSeries, Condition, escape_status) %>% 
  summarize(value = mean(value)) %>% ungroup() -> data_here_test

data_here_test %>%
  ggplot(aes(x = Condition, y = value, fill = escape_status)) + geom_boxplot(col = "black") + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    scale_fill_manual(values = escape_colors) + 
    labs(colour = "Escape status") + coord_flip() + 
    ggpubr::stat_compare_means(comparisons = list(c("Control", "Dox"), c("Control", "7 days washout"), c("Control", "14 days washout"), 
                                                  c("Control", "14 days washout"), c("Control", "21 days washout"),
                                                  c("7 days washout", "14 days washout"), c("7 days washout", "21 days washout"))) +
    facet_grid(cols = vars(TimeSeries), scales = "free_x", space = "free_x")
ggplot2::ggsave("./Plots/Fig3/Washout_escape_boxplots_cl30.pdf", width = 14, height = 6)

# 
data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Dox, Washout)) %>%
  group_by(name) %>% 
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  mutate(Condition = factor(Condition, levels = paste0(c(7, 14, 21), " days washout"))) %>%
  mutate(escape_status = factor(escape_status, levels = c("Constitutive", "Facultative", "NPC-specific"))) -> df_plot_s8e

df_plot_s8e %>%
  ggplot(aes(x = Condition, y = washout_ratio)) + geom_boxplot(aes(fill = escape_status)) + facet_wrap(~TimeSeries, drop = T) + 
    scale_fill_manual(values = escape_colors) + theme_paper(textsize = 25) + geom_hline(yintercept = 1, linetype = 'dashed') + xlab("") + 
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + ylab("Reversibility (Allelic Ratio Washout / Control)")
ggplot2::ggsave("./Plots/Fig3/reversibility_fraction_long_cl30.pdf", width = 14, height = 6)

condition_change <- list(
  "7 days dox treatment_7 days washout" = "Dox (7 days), washout (7 days)", 
  "14 days dox treatment_7 days washout" = "Dox (14 days), washout (7 days)", 
  "14 days dox treatment_14 days washout" = "Dox (14 days), washout (14 days)", 
  "14 days dox treatment_21 days washout" = "Dox (14 days), washout (21 days)", 
  "21 days dox treatment_7 days washout" = "Dox (21 days), washout (7 days)"
)
data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  # mutate(Condition = factor(Condition, levels = paste0(c(7, 14, 21), " days washout"))) %>%
  mutate(Condition = paste0(TimeSeries, "_", Condition)) %>%
  group_by(Condition) %>%
  summarize(values = list(washout_ratio)) %>%
  ungroup() %>%
  tidyr::expand(nesting(Condition, values), nesting(Condition, values),  .name_repair = "unique") %>%
  setNames(tolower(gsub("\\.","", names(.)))) %>%
  dplyr::filter(condition1 != condition3) %>%
  rowwise() %>%
  mutate(condition1 = sort(c(condition1, condition3))[1], condition3 = sort(c(condition1, condition3))[2]) %>% dplyr::filter(condition1 != condition3) %>%
  mutate(pvals = wilcox.test(values2, values4)$p.value, paired = T) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvals, method = "BH")) %>% 
  dplyr::select(-c("values2", "values4")) %>%
  mutate(condition1 = unlist(condition_change[condition1]), condition3 = unlist(condition_change[condition3])) %>%
  write.csv("./reversibility_fraction_long_cl30_pvals.csv")

data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  mutate(Condition = paste0(TimeSeries, "_", Condition)) %>%
  group_by(Condition, escape_status) %>%
  summarize(values = list(washout_ratio)) %>%
  ungroup() %>%
  tidyr::expand(nesting(Condition, escape_status, values), nesting(Condition, escape_status, values),  .name_repair = "unique") %>%
  setNames(tolower(gsub("\\.","", names(.)))) %>%
  dplyr::filter(condition1 != condition4 & escape_status2 == escape_status5) %>%
  rowwise() %>%
  mutate(condition1 = sort(c(condition1, condition4))[1], condition4 = sort(c(condition1, condition4))[2]) %>% dplyr::filter(condition1 != condition4) %>%
  mutate(pvals = wilcox.test(values3, values6)$p.value, paired = T) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvals, method = "BH")) %>%
  dplyr::select(-c("values3", "values6")) %>% dplyr::filter(!duplicated(pvals)) %>%
  write.csv("./reversibility_fraction_long_cl30_pvals_escapegroups.csv")

# 
### 
data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  mutate(Condition = factor(Condition, levels = paste0(c(7, 14, 21), " days washout"))) %>%
  mutate(escape_status = factor(escape_status, levels = c("Constitutive", "Facultative", "NPC-specific"))) %>%
  mutate(reversibility_category = case_when( ### think about these cutoffs
    washout_ratio > .5 ~ "Reversible", 
    washout_ratio < .1 ~ "Irreversible",
    .default = "Partially Reversible" 
  )) -> save_data_three_way
  
data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  mutate(Condition = factor(Condition, levels = paste0(c(7, 14, 21), " days washout"))) %>%
  mutate(escape_status = factor(escape_status, levels = c("Constitutive", "Facultative", "NPC-specific"))) %>%
  mutate(reversibility_category = ifelse(washout_ratio > .1, "Reversible", "Irreversible")) -> save_data_tenpercent_rule

## use both pieces of information: 
save_data_three_way %>% right_join(save_data_tenpercent_rule, by = c("name", "TimeSeries", "Condition")) %>% 
  dplyr::rename("Ratio" = "washout_ratio.x", "Escape" = "washout_ratio.y") %>%
  mutate(reversibility_category_joint = case_when(
    Ratio > .5 & Escape > .1 ~ "Reversible", 
    (Ratio < .5 & Ratio > .1) & Escape > .1 ~ "Partially Reversible", 
    .default = "Irreversible" 
  )) -> save_data_joint

save_data_joint %>%
  ggplot(aes(x = TimeSeries, fill = reversibility_category_joint)) + geom_bar(col = "black") +
    theme_bw(base_size = 30) + xlab("") + ylab("Number of genes") +
    scale_fill_manual(values = rev(c("pink1", "orchid4", "darkolivegreen2"))) +
    labs(fill = "Reversibility category") + theme_paper() +
    facet_wrap(~Condition + escape_status.x, scales = "free_y", nrow = 1) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
    stat_count(geom = "text", colour = "black", aes(label = ..count..), position=position_stack(vjust=0.5))
ggplot2::ggsave("./Plots/Fig3/reversibility_genes_long_cl30.pdf", width = 20, height = 12)

save_data_joint_cl30 <- save_data_joint

```

CL31

SAME THINGS FOR CL30s
CL31

```{r load_data}

data <- readRDS("./ProcessedData/merged_dataset_paper.rds")
data_before_filtering <-  data[seqnames(data) == "X", ]

genes_keep <- readRDS("./ProcessedData/genes_keep.rds")
data <- data[genes_keep, ]

data_with_autosomes <- data
data <- data[seqnames(data) == "X", ]

data_test <- data[ , !data$Clone == "E6"]

table(data_test$Experiment, data_test$ConditionClean)

```

```{r ssss}

### This is for Figure 3, the reversibility aspect
### Now we look at Xist overexpression and how that silences escapees
samples_use <- c("Control", 
                 # "Dox (3d)", "Dox (3d) - washout (4d)", 
                 
                 "Dox (7d)", 
                 "Dox (7d) - washout (4d)", 
                 "Dox (7d) - washout (7d)", 

                 "Dox (14d)", 
                 "Dox (14d) - washout (4d)", 
                 "Dox (14d) - washout (7d)", 
                 "Dox (14d) - washout (14d)", 
                 "Dox (14d) - washout (21d)", 

                 "Dox (21d)", 
                 "Dox (21d) - washout (4d)",
                 "Dox (21d) - washout (7d)"
)

data_here <- data[ , data$ConditionClean %in% samples_use]
data_here <- data_here[ , grepl("CL31", data_here$Clone)]
data_here <- data_here[ , data_here$Experiment == "May2025"]
data_here <- data_here[ , !(data_here$Experiment == "May2025" & data_here$Clone == "CL30.7" & data_here$Replicate == "Rep3")]
# data_here <- data_here[,!data_here$Experiment == "September2022"]
# data_here <- data_here[,!data_here$Experiment %in% c("September2022", "October2023")]

ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))
ratios <- ratios[apply(ratios, 1, function(x){!any(is.na(x))}), ]

# define escapees
npc_escape <- read.csv("./ProcessedData/excel_files/escape_metaanalysis_final.escape_per_clone.csv")
npc_escape$E6_merged <- ifelse(npc_escape$E6...Rep1 == "Escapee" & npc_escape$E6...Rep2 == "Escapee" & npc_escape$E6...Rep3 == "Escapee",
                               "Escapee", "Silenced")
npc_escape$CL30_merged <- ifelse(npc_escape$CL30...Rep1 == "Escapee" & npc_escape$CL30...Rep2 == "Escapee" & 
                                  npc_escape$CL30...Rep4 == "Escapee" & npc_escape$CL30...Rep5 == "Escapee",
                               "Escapee", "Silenced")
npc_escape$CL31_merged <- ifelse(npc_escape$CL31...Rep1 == "Escapee" & npc_escape$CL31...Rep2 == "Escapee" & npc_escape$CL31...Rep4 == "Escapee" & 
                                  npc_escape$CL31...Rep5 == "Escapee",
                               "Escapee", "Silenced")
cl30_escapees <- npc_escape %>% dplyr::filter(CL30_merged == "Escapee") %>% pull(symbol)
cl31_escapees <- npc_escape %>% dplyr::filter(CL31_merged == "Escapee") %>% pull(symbol)
escapees <- list("CL30" = cl30_escapees, "CL31" = cl31_escapees)

# For Xist-expression, we need a scaling factor per sample to account for differences in seq depth
size_factors <- colSums(counts(data_here)) / colSums(counts(data_here))[[1]]

# Supplement
xist_expression_levels <- data.frame(
  Dox = data_here$ndDox,
  Washout = data_here$ndWashout, 
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6)
)

mean_baseline_xist <- xist_expression_levels %>% dplyr::filter(Dox == 0) %>% pull(Expression) %>% mean()

data.frame(
  Dox = data_here$ndDox,
  Washout = data_here$ndWashout, 
  Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ])) / colSums(counts(data_here)) * 1e6 / mean_baseline_xist
) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("3", "7", "14", "21")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>%
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
  mutate(Condition = factor(Condition, levels = 
                              rev(c("Control", "0 days washout", "4 days washout", "7 days washout", "14 days washout", "21 days washout")))) %>%
  mutate(TimeSeries = factor(TimeSeries, levels = c("3 days dox treatment", "7 days dox treatment", "14 days dox treatment", "21 days dox treatment"))) %>%
  mutate(Clone = ifelse(grepl("CL30", Clone), "CL30", "CL31")) -> df_plot_s8d

df_plot_s8d %>%
  ggplot(aes(x = Condition, y = Expression)) + 
    stat_summary(stat = "mean", geom = "bar", fill = "grey", col = "black") + 
    coord_flip() + theme_paper() + 
    xlab("") + ylab("Expression (CPM)") + geom_jitter(size = 3, width = 0.1, aes(shape = Clone)) + 
    scale_y_continuous(expand = c(0, 0), limits = c(0, 13)) + 
    facet_wrap(~TimeSeries, ncol = 1, scales = "free_y") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggplot2::ggsave("./Plots/FigS3/Washout_Xist_barplots_cl31.pdf", height = 8, width = 6)

```

```{r }

genes_show <- c("Xist", "Mecp2", "Kdm6a", "Kdm5c")
genes_of_choice <- setNames(rep("", nrow(ratios)), rownames(ratios))
genes_of_choice[genes_show] <- genes_show

ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Replicate, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Condition, Clone, Dox)) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  mutate(Condition = factor(Condition, levels = rev(samples_use))) %>%
  mutate(TimePoint = case_when(
    grepl("Dox \\(3d\\)", Condition)  ~ "3d",
    grepl("Dox \\(7d\\)", Condition)  ~ "7d",
    grepl("Dox \\(14d\\)", Condition)  ~ "14d",
    grepl("Dox \\(21d\\)", Condition)  ~ "21d",
    grepl("Control", Condition)  ~ "0d")) %>% 
  mutate(TimePoint = factor(TimePoint, levels = paste0(c(0, 3, 7, 14, 21), "d"))) -> to_plot

to_plot %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_y_discrete(labels = genes_of_choice, position = "right") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    scale_x_discrete(labels = genes_of_choice, position = "bottom") + 
    labs(fill = "d-score") +  facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.y = element_blank())
ggplot2::ggsave("./Plots/FigS3/Washout_full_heatmap_cl31.pdf")

to_plot %>%
  mutate(Replicate = str_extract(Clone, "Rep[0-9]")) %>%
  ggplot(aes(reorder(name, position), y = reorder(paste0(Condition, "_", Clone), Condition), fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_y_discrete(labels = genes_of_choice, position = "right") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    scale_x_discrete(labels = genes_of_choice, position = "bottom") +
    labs(fill = "d-score") + facet_grid(rows = vars(TimePoint, Replicate), scales = "free_y", space = "free_y") +
    theme(strip.background = element_blank(), strip.text.y = element_blank())
# ggplot2::ggsave("./Plots/FigS3/Washout_full_heatmap_allsamples_cl31.pdf")

to_plot %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_y_discrete(labels = genes_of_choice, position = "right") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    scale_x_discrete(labels = genes_of_choice, position = "bottom") + 
    labs(fill = "d-score") +  facet_grid(rows = vars(TimePoint), scales = "free_y", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.y = element_blank())
#ggplot2::ggsave("./Plots/FigS3/Washout_full_heatmap_cl30.pdf")

to_plot %>%
  mutate(Replicate = str_extract(Clone, "Rep[0-9]")) %>%
  ggplot(aes(reorder(name, position), y = reorder(paste0(Condition, "_", Clone), Condition), fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_y_discrete(labels = genes_of_choice, position = "right") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    scale_x_discrete(labels = genes_of_choice, position = "bottom") +
    labs(fill = "d-score") + facet_grid(rows = vars(TimePoint, Replicate), scales = "free_y", space = "free_y") +
    theme(strip.background = element_blank(), strip.text.y = element_blank())
#ggplot2::ggsave("./Plots/FigS3/Washout_full_heatmap_allsamples_cl30.pdf")

p1 <- to_plot %>%
  dplyr::filter(name %in% cl31_escapees) %>%
  mutate(Timepoint = str_extract(Condition, "Control|(Dox \\([0-9]*d\\))")) %>% 
  mutate(Timepoint = factor(Timepoint, c("Control", "Dox (7d)", "Dox (14d)", "Dox (21d)"))) %>%
  ggplot(aes(reorder(name, position),  y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") +
    theme_paper() +
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) +
    xlab("") +
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) +
    labs(fill = "d-score") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 10)) +
    scale_x_discrete(position = "top") + guides(fill = "none") + 
    facet_grid(rows = vars(Timepoint), scales = "free", space = "free_y") + 
    theme(strip.background = element_blank(), strip.text.x = element_blank()) + 
    labs(col = "Allelic Ratio \n (Xi / (Xi + Xa))")

# Plot escapee annotation at the bottom of this:
p2 <- data.frame(Gene = cl31_escapees,
           Annotation = rowData(data)[cl31_escapees, ]$EscapeAnnotation,
           Position = start(rowRanges(data)[cl31_escapees, ])) %>%
  ggplot(aes(x = reorder(Gene, Position), y = 1, fill = Annotation)) + 
    theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) +
    theme(legend.position = "None")

fill_colors_here <- c(RColorBrewer::brewer.pal(name = "Paired", n = 11), "white")

p1 / p2 + plot_layout(heights = c(4, 0.2, 0.2))
ggplot2::ggsave("./Plots/Fig3/heatmap_ordered_only_escapees_by_position_reversibility_cl31.pdf", width = 24, height = 5)


ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Washout" = data_here$ndWashout) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Experiment)) %>%
  dplyr::filter(Washout != 4) %>% ### not sure about this yet -- but plot doesnt look great keeping it in
  pivot_longer(-c(Dox, Washout, Clone)) %>%
  group_by(Clone) %>%
  add_column(TimeSeries = ifelse(.$Dox == 0, list(c("7", "14", "21")), .$Dox)) %>%
  unnest(TimeSeries) %>%
  mutate(TimeSeries = paste0(unlist(TimeSeries), " days dox treatment")) %>%
  add_column(Condition = ifelse(.$Dox == 0, "Control", paste0(.$Washout, " days washout"))) %>%
  mutate(Condition = gsub("(0 days washout)", "Dox", Condition)) %>%
  mutate(Condition = factor(Condition, levels = c("Control", "Dox", "4 days washout", "7 days washout", "14 days washout", "21 days washout"))) %>%
  # dplyr::filter(TimeSeries %in% c("3 days dox treatment", "7 days dox treatment")) %>%
  add_column(escape_status = factor(unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation), levels = names(escape_colors))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  mutate(TimeSeries = factor(TimeSeries, levels = paste0(c(7, 14, 21), " days dox treatment"))) %>%
  mutate(Condition = factor(Condition, levels = rev(c("Control", "Dox", "4 days washout", "7 days washout", "14 days washout", "21 days washout")))) %>%
  mutate(Clone_Group = ifelse(grepl("CL30", Clone), "CL30", "CL31")) %>%
  group_by(Clone_Group) %>%
  dplyr::filter(name %in% escapees[[unique(Clone_Group)]]) %>%
  group_by(Dox, Washout, name, TimeSeries, Condition, escape_status) %>% 
  summarize(value = mean(value)) %>% ungroup() -> data_here_test

data_here_test %>%
  ggplot(aes(x = Condition, y = value, fill = escape_status)) + geom_boxplot(col = "black") + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    scale_fill_manual(values = escape_colors) + 
    labs(colour = "Escape status") + coord_flip() + 
    ggpubr::stat_compare_means(comparisons = list(c("Control", "Dox"), c("Control", "7 days washout"), c("Control", "14 days washout"), 
                                                  c("Control", "14 days washout"), c("Control", "21 days washout"),
                                                  c("7 days washout", "14 days washout"), c("7 days washout", "21 days washout"))) +
    facet_grid(cols = vars(TimeSeries), scales = "free_x", space = "free_x")
ggplot2::ggsave("./Plots/Fig3/Washout_escape_boxplots_cl31.pdf", width = 14, height = 6)

# 
data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Dox, Washout)) %>%
  group_by(name) %>% 
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  mutate(Condition = factor(Condition, levels = paste0(c(7, 14, 21), " days washout"))) %>%
  mutate(escape_status = factor(escape_status, levels = c("Constitutive", "Facultative", "NPC-specific"))) -> df_plot_s8f

df_plot_s8f %>%
  ggplot(aes(x = Condition, y = washout_ratio)) + geom_boxplot(aes(fill = escape_status)) + facet_wrap(~TimeSeries, drop = T) + 
    scale_fill_manual(values = escape_colors) + theme_paper(textsize = 25) + geom_hline(yintercept = 1, linetype = 'dashed') + xlab("") + 
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + ylab("Reversibility (Allelic Ratio Washout / Control)")
ggplot2::ggsave("./Plots/Fig3/reversibility_fraction_long_cl31.pdf", width = 14, height = 6)

condition_change <- list(
  "7 days dox treatment_7 days washout" = "Dox (7 days), washout (7 days)", 
  "14 days dox treatment_7 days washout" = "Dox (14 days), washout (7 days)", 
  "14 days dox treatment_14 days washout" = "Dox (14 days), washout (14 days)", 
  "14 days dox treatment_21 days washout" = "Dox (14 days), washout (21 days)", 
  "21 days dox treatment_7 days washout" = "Dox (21 days), washout (7 days)"
)
data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  # mutate(Condition = factor(Condition, levels = paste0(c(7, 14, 21), " days washout"))) %>%
  mutate(Condition = paste0(TimeSeries, "_", Condition)) %>%
  group_by(Condition) %>%
  summarize(values = list(washout_ratio)) %>%
  ungroup() %>%
  tidyr::expand(nesting(Condition, values), nesting(Condition, values),  .name_repair = "unique") %>%
  setNames(tolower(gsub("\\.","", names(.)))) %>%
  dplyr::filter(condition1 != condition3) %>%
  rowwise() %>%
  mutate(condition1 = sort(c(condition1, condition3))[1], condition3 = sort(c(condition1, condition3))[2]) %>% dplyr::filter(condition1 != condition3) %>%
  mutate(pvals = wilcox.test(values2, values4)$p.value, paired = T) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvals, method = "BH")) %>% 
  dplyr::select(-c("values2", "values4")) %>%
  mutate(condition1 = unlist(condition_change[condition1]), condition3 = unlist(condition_change[condition3])) %>%
  write.csv("./reversibility_fraction_long_cl31_pvals.csv")

data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  mutate(Condition = paste0(TimeSeries, "_", Condition)) %>%
  group_by(Condition, escape_status) %>%
  summarize(values = list(washout_ratio)) %>%
  ungroup() %>%
  tidyr::expand(nesting(Condition, escape_status, values), nesting(Condition, escape_status, values),  .name_repair = "unique") %>%
  setNames(tolower(gsub("\\.","", names(.)))) %>%
  dplyr::filter(condition1 != condition4 & escape_status2 == escape_status5) %>%
  rowwise() %>%
  mutate(pvals = wilcox.test(values3, values6)$p.value, paired = T) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvals, method = "BH")) %>%
  dplyr::select(-c("values3", "values6")) %>% dplyr::filter(!duplicated(pvals)) %>%
  write.csv("./reversibility_fraction_long_cl31_pvals_escapegroups.csv")

### 
data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  mutate(Condition = factor(Condition, levels = paste0(c(7, 14, 21), " days washout"))) %>%
  mutate(escape_status = factor(escape_status, levels = c("Constitutive", "Facultative", "NPC-specific"))) %>%
  mutate(reversibility_category = case_when( ### think about these cutoffs
    washout_ratio > .5 ~ "Reversible", 
    washout_ratio < .1 ~ "Irreversible",
    .default = "Partially Reversible" 
  )) -> save_data_three_way
  
data_here_test %>%
  ungroup() %>%
  dplyr::select(-c(Washout)) %>%
  group_by(name) %>%
  mutate(value_control = unique(value[Condition == "Control"])) %>%
  mutate(washout_ratio = value / value_control) %>%
  dplyr::filter(Condition %in% c("7 days washout", "14 days washout", "21 days washout")) %>%
  mutate(Condition = factor(Condition, levels = paste0(c(7, 14, 21), " days washout"))) %>%
  mutate(escape_status = factor(escape_status, levels = c("Constitutive", "Facultative", "NPC-specific"))) %>%
  mutate(reversibility_category = ifelse(washout_ratio > .1, "Reversible", "Irreversible")) -> save_data_tenpercent_rule

## use both pieces of information: 
save_data_three_way %>% right_join(save_data_tenpercent_rule, by = c("name", "TimeSeries", "Condition")) %>% 
  dplyr::rename("Ratio" = "washout_ratio.x", "Escape" = "washout_ratio.y") %>%
  mutate(reversibility_category_joint = case_when(
    Ratio > .5 & Escape > .1 ~ "Reversible", 
    (Ratio < .5 & Ratio > .1) & Escape > .1 ~ "Partially Reversible", 
    .default = "Irreversible" 
  )) -> save_data_joint

save_data_joint %>%
  ggplot(aes(x = TimeSeries, fill = reversibility_category_joint)) + geom_bar(col = "black") +
    theme_bw(base_size = 30) + xlab("") + ylab("Number of genes") +
    scale_fill_manual(values = rev(c("pink1", "orchid4", "darkolivegreen2"))) +
    labs(fill = "Reversibility category") + theme_paper() +
    facet_wrap(~Condition + escape_status.x, scales = "free_y", nrow = 1) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
    stat_count(geom = "text", colour = "black", aes(label = ..count..), position=position_stack(vjust=0.5))
ggplot2::ggsave("./Plots/Fig3/reversibility_genes_long_cl31.pdf", width = 20, height = 12)

save_data_joint_cl31 <- save_data_joint

```

```{r assddasd}

# compare between E6 and CL30
save_data_total <- rbind(cbind("cov" = "E6", save_data_joint_e6), 
                         rbind(cbind("cov" = "CL30", save_data_joint_cl30), cbind("cov" = "CL31", save_data_joint_cl31)))

save_data_total %>%
  dplyr::filter(cov != "CL31") %>%
  dplyr::select(name, reversibility_category.x, Condition, TimeSeries, cov) %>%
  group_by(Condition, TimeSeries) %>%
  pivot_wider(names_from = cov, values_from = reversibility_category.x) %>%
  dplyr::filter(!is.na(E6) & !is.na(CL30)) %>%
  ggplot(aes(x = E6, y = CL30)) + geom_bin_2d(col = 'black') + facet_wrap(~Condition + TimeSeries) + theme_paper() + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + scale_fill_gradient(low = "white", high = "red") + 
    stat_bin2d(geom = "text", aes(label = ..count..), binwidth = 1)
ggplot2::ggsave("./Plots/Fig3/compare_reversibility_bygene_across_clones.pdf", width = 10, height = 8)

# compare between 7d and longer washouts within clones
save_data_total -> df_plot_s8b
save_data_total %>%
  dplyr::select(name, reversibility_category.x, Condition, TimeSeries, cov) %>%
  group_by(TimeSeries, cov) %>%
  pivot_wider(names_from = Condition, values_from = reversibility_category.x) %>%
  dplyr::filter(!is.na(`7 days washout`) & !is.na(`14 days washout`)) %>%
  ggplot(aes(x = `7 days washout`, y = `14 days washout`)) + geom_bin_2d(col = 'black') + facet_wrap(~cov + TimeSeries) + theme_paper() + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + scale_fill_gradient(low = "white", high = "red") + 
    stat_bin2d(geom = "text", aes(label = ..count..), binwidth = 1)
ggplot2::ggsave("./Plots/Fig3/compare_reversibility_bygene_across_times.pdf", width = 10, height = 8)

```

```{r export_capture}

samples_use <- c("Control", 

                 "Dox (7d)", 
                 "Dox (7d) - washout (4d)", 
                 
                 "Dox (21d)", 
                 "Dox (21d) - washout (7d)"
)

data_here <- data[,data$ConditionClean %in% samples_use]
data_here <- data_here[,data_here$Clone == "E6"]
data_here <- data_here[,!data_here$Experiment %in% c("September2022", "October2023")]

table(data_here$ConditionClean)

ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))
ratios <- ratios[apply(ratios, 1, function(x){!any(is.na(x))}), ]

ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Dox" = data_here$ndDox) %>%
  add_column("Clone" = paste0(data_here$Clone, " - ", data_here$Replicate, " - ", data_here$Experiment)) %>%
  pivot_longer(-c(Condition, Clone, Dox)) %>%
  add_column(start = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  add_column(end = end(rowRanges(data_here[as.character(.$name), ]))) %>%
  mutate(Condition = factor(Condition, levels = rev(samples_use))) %>%
  mutate(TimePoint = case_when(
    grepl("Dox \\(3d\\)", Condition)  ~ "3d",
    grepl("Dox \\(7d\\)", Condition)  ~ "7d",
    grepl("Dox \\(14d\\)", Condition)  ~ "14d",
    grepl("Dox \\(21d\\)", Condition)  ~ "21d",
    grepl("Control", Condition)  ~ "0d")) %>% 
  mutate(TimePoint = factor(TimePoint, levels = paste0(c(0, 3, 7, 14, 21), "d"))) %>%
  saveRDS(., "./ProcessedData/washout_for_capture.rds")

```


```{r asdasd}

test1 <- read_csv("./reversibility_fraction_long_e6_pvals_escapegroups.csv")
test2 <- read_csv("./reversibility_fraction_long_cl30_pvals_escapegroups.csv")
test3 <- read_csv("./reversibility_fraction_long_cl31_pvals_escapegroups.csv")

```

```{r asdasd}
# combine source data

source_data_4 <- list(df_plot_6b, df_plot_6d, df_plot_6e, df_plot_6f, df_plot_6g, 
                      df_plot_s8a, df_plot_s8b, df_plot_s8c, df_plot_s8d, df_plot_s8e, df_plot_s8f)
saveRDS(source_data_4, "./ProcessedData/source_data/source_data_4.rds")

```