---
title: "cut_and_run_normalization_check_genelevel.R"
output: html_document
date: "2023-11-03"
editor_options: 
  chunk_output_type: console
---

Here I'm investigating the normalization of CnR data, based on windowed counts generated by Antonia. 

```{r asss}

library(csaw)
library(DESeq2)
library(tidyverse)

theme_paper <- function(textsize = 20){
  theme(panel.background = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1), 
        text = element_text(size = textsize, color = "black"), 
        axis.text = element_text(size = textsize, color = "black"), 
        axis.ticks = element_line(colour = 'black', size = 1), 
        axis.ticks.length = unit(.25, "cm"))
}

setwd("/omics/groups/OE0538/internal/users/panten/projects/AntoniaESCProject/allele-specific_cut_and_run/")

# test <- readRDS("./from_antonia/[Extern]_-_CaR_normalization/bg_10000_Gall_noX_E6Xist")
test <- readRDS("./from_antonia/all.genes_all.bam_all.chr_E6Xist")
test <- test[,grepl("Gall", test$bam.files)]

```

```{r assdasda}

file.names <- basename(colData(test)$bam.files)

metadata <- data.frame(
  Clone = "E6", 
  HistoneMark = str_extract(file.names, "H2AK119ubi|H3K27me3"), 
  Condition = gsub("_(H2AK119ubi|H3K27me3)", "", gsub("E6", "", gsub("_Gall_sorted.bam", "", file.names)))
) %>%
  mutate(Replicate = str_extract(Condition, "r[0-9]")) %>%
  mutate(Dox = str_extract(Condition, "(21d|7d|No)")) %>% 
  mutate(Dox = as.numeric(gsub("d", "", Dox))) %>% mutate(Dox = ifelse(is.na(Dox), 0, Dox)) %>%
  mutate(Washout = ifelse(grepl("WO", Condition), 0, 7))

### assume that sample 5/6 and 13/14 are swapped
metadata_corrected <- metadata[c(1:4, 6, 5, 7:12, 14, 13, 15:20), ]
metadata_corrected$Name <- file.names[c(1:4, 6, 5, 7:12, 14, 13, 15:20)]
# metadata_corrected <- metadata
rownames(metadata_corrected) <- paste0("Sample", 1:nrow(metadata))

colData(test) <- cbind(colData(test), metadata_corrected)

colnames(test) <- paste0("Sample", 1:nrow(metadata_corrected))

```

```{r asdass}

assays(test)[[1]] %>% 
  data.frame() %>%
  pivot_longer(-c()) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) -> long_data

long_data %>%
  ggplot(aes(x = Condition, y = value + 1, fill = Condition)) + geom_violin() + scale_y_log10() + facet_wrap(~assay, scales = "free") + 
    coord_flip() + 
    geom_text(aes(label = name, y = 10), data = head(long_data, n = 20)) + 
    theme_paper(textsize = 10) + xlab("") + ylab("Number of reads (+1)")

```

```{r normalize_ubi}

test_ubi <- test[ , test$HistoneMark == "H2AK119ubi"]

rowMeans(assays(test_ubi)[[1]]) %>% 
  data.frame() %>% ggplot(aes(x = .)) + geom_histogram() + scale_x_log10() + geom_vline(xintercept = 10) + theme_paper() + 
    xlab("Average number of reads (+1)") + ylab("Number of windows")

deseq_ubi <- DESeqDataSet(test_ubi, design = ~0)
deseq_ubi <- deseq_ubi[rowMeans(counts(deseq_ubi)) > 20, ]

counts(deseq_ubi, normalized = F) %>% 
  data.frame() %>%
  add_column(Window = paste0("Window", 1:nrow(deseq_ubi))) %>%
  pivot_longer(-c(Window)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) -> long_data_ubi

long_data_ubi %>%
  ggplot(aes(x = Condition, y = value + 1, fill = Condition)) + geom_violin() + scale_y_log10() + facet_wrap(~assay, scales = "free") + 
    coord_flip() + geom_text(aes(label = name, y = 10), data = head(long_data_ubi, n = 20)) + 
    theme_paper(textsize = 10) + xlab("") + ylab("Number of reads (+1)")

deseq_ubi <- estimateSizeFactors(deseq_ubi)

counts(deseq_ubi, normalized = T) %>% 
  data.frame() %>%
  add_column(Window = paste0("Window", 1:nrow(deseq_ubi))) %>%
  add_column(Chr = as.character(seqnames(rowRanges(deseq_ubi)))) %>%
  add_column(Pos = start(rowRanges(deseq_ubi)) + 5000) %>%
  pivot_longer(-c(Window, Chr, Pos)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) -> long_data_ubi_norm

long_data_ubi_norm %>%
  ggplot(aes(x = Condition, y = value + 1, fill = Condition)) + geom_violin() + scale_y_log10() + facet_wrap(~assay, scales = "free") + 
    coord_flip() + geom_text(aes(label = name, y = 10), data = head(long_data_ubi_norm, n = 20)) + 
    theme_paper(textsize = 10) + xlab("") + ylab("Number of reads (+1)")

library(GGally)
lowerfun <- function(data, mapping){
  ggplot(data = data, mapping = mapping)+
    geom_point(size = .1) + geom_abline()
}

long_data_ubi_norm %>% dplyr::select(c(Window, name, value)) %>% 
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  dplyr::select(-c(Window)) %>% 
  sample_frac(.01) %>% 
  GGally::ggpairs(columns = 1:10, lower = list(continuous = wrap(lowerfun))) + 
    theme_paper(textsize = 10) + scale_x_log10() + scale_y_log10()

chromosome_colors <- setNames(c(rep(c("grey", "black"), 9), "grey", "red", "blue"), nm = paste0("chr", c(1:19, "X", "Y")))

### look at individual sample pairs that seem off Sample7 / 17 (21d dox vs 17d dox)
long_data_ubi_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  dplyr::filter(name %in% c("Sample17", "Sample7")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>% 
  sample_frac(0.1) %>%
  dplyr::filter(Chr != "chrX") %>%
  ggplot(aes(x = Sample7, y = Sample17, col = Chr)) + geom_point(size = .1) + geom_abline() + scale_x_log10() + scale_y_log10() + 
    theme_paper() + scale_color_manual(values = chromosome_colors) + ylab("21d / control")
# ggsave("./plots/preprocessing/")

# plot by windows
long_data_ubi_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  dplyr::filter(name %in% c("Sample17", "Sample7")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>% 
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  sample_frac(0.1) %>%
  ggplot(aes(x = Window, log2(Sample7 / Sample17), col = Chr)) + 
    geom_point(size = .1) + theme_paper() + geom_hline(yintercept = 0, color = "cyan") + 
    scale_color_manual(values = chromosome_colors) + 
    geom_hline(yintercept = c(-1, 1), linetype = "dashed")
# this is really weird. what are these spikes??

long_data_ubi_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  # dplyr::filter(name %in% c("Sample19", "Sample15")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  # mutate(across(starts_with('Sample'), ~ . / ((Sample17 + Sample19) / 2)) ) %>%
  mutate(across(starts_with('Sample'), ~ . / (Sample17)) ) %>%
  sample_frac(0.1) %>%
  pivot_longer(-c(Window, Chr, Pos)) %>%
  mutate(name2 = colData(deseq_ubi)[.$name, ]$Condition) %>%
  ggplot(aes(x = Window, log2(value), col = Chr)) + 
    geom_point(size = .1) + theme_paper() + geom_hline(yintercept = 0, color = "cyan") + 
    scale_color_manual(values = chromosome_colors) + 
    geom_hline(yintercept = c(-1, 1), linetype = "dashed") + 
    facet_wrap(~name2)
ggsave("./plots/preprocessing/normalization_ubi.pdf", width = 20, height = 12)
ggsave("./plots/preprocessing/normalization_ubi.png", width = 20, height = 12)

### confirm with the other long sample
### look at individual sample pairs that seem off Sample5 / 17 (21d dox vs 17d dox)
long_data_ubi_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  dplyr::filter(name %in% c("Sample19", "Sample5")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>% 
  sample_frac(0.1) %>%
  ggplot(aes(x = Sample5, y = Sample19, col = Chr)) + geom_point(size = .1) + geom_abline() + scale_x_log10() + scale_y_log10() + 
    theme_paper() + scale_color_manual(values = chromosome_colors)

# plot by windows
long_data_ubi_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  dplyr::filter(name %in% c("Sample19", "Sample5")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>% 
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  sample_frac(0.1) %>%
  ggplot(aes(x = Window, log2(Sample5 / Sample19), col = Chr)) + 
    geom_point(size = .1) + theme_paper() + geom_hline(yintercept = 0, color = "cyan") + 
    scale_color_manual(values = chromosome_colors) + 
    geom_hline(yintercept = c(-1, 1), linetype = "dashed")

### custom normalization: 

library(mixtools)

median_counts <- rowMedians(counts(deseq_ubi))

median_normalized_counts <- apply(counts(deseq_ubi), 2, function(x){x / median_counts}) %>% data.frame() %>% 
  sample_frac(.01)

my_mix <- normalmixEM(median_normalized_counts[, 8], k = 3)
plot(my_mix, which = 2)

median_normalized_counts %>% pivot_longer(-c()) %>%
  ggplot(aes(x = value)) + geom_histogram(bins = 100) + facet_wrap(~name)

median_normalized_counts %>% data.frame() %>%
  add_column(Window = 1:nrow(deseq_ubi)) %>%
  sample_frac(0.01) %>%
  pivot_longer(-c(Window)) %>% 
  ggplot(aes(x = Window, y = log(value + 1))) + geom_point(size = .1) + facet_wrap(~name)

```

```{r normalize_me3}

test_met <- test[ , test$HistoneMark == "H3K27me3"]

rowMeans(assays(test_met)[[1]]) %>% 
  data.frame() %>% ggplot(aes(x = .)) + geom_histogram() + scale_x_log10() + geom_vline(xintercept = 5) + theme_paper() + 
    xlab("Average number of reads (+1)") + ylab("Number of windows")

deseq_met <- DESeqDataSet(test_met, design = ~0)
deseq_met <- deseq_met[rowMeans(counts(deseq_met)) > 5, ]

counts(deseq_met, normalized = F) %>% 
  data.frame() %>%
  add_column(Window = paste0("Window", 1:nrow(deseq_met))) %>%
  pivot_longer(-c(Window)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) -> long_data_met

long_data_met %>%
  ggplot(aes(x = Condition, y = value + 1, fill = Condition)) + geom_violin() + scale_y_log10() + facet_wrap(~assay, scales = "free") + 
    coord_flip() + geom_text(aes(label = name, y = 10), data = head(long_data_met, n = 20)) + 
    theme_paper(textsize = 10) + xlab("") + ylab("Number of reads (+1)")

deseq_met <- estimateSizeFactors(deseq_met)

counts(deseq_met, normalized = T) %>% 
  data.frame() %>%
  add_column(Window = paste0("Window", 1:nrow(deseq_met))) %>%
  add_column(Chr = as.character(seqnames(rowRanges(deseq_met)))) %>%
  add_column(Pos = start(rowRanges(deseq_met)) + 5000) %>%
  pivot_longer(-c(Window, Chr, Pos)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) -> long_data_met_norm

long_data_met_norm %>%
  ggplot(aes(x = Condition, y = value + 1, fill = Condition)) + geom_violin() + scale_y_log10() + facet_wrap(~assay, scales = "free") + 
    coord_flip() + geom_text(aes(label = name, y = 10), data = head(long_data_met_norm, n = 20)) + 
    theme_paper(textsize = 10) + xlab("") + ylab("Number of reads (+1)")

library(GGally)
lowerfun <- function(data, mapping){
  ggplot(data = data, mapping = mapping)+
    geom_point(size = .1) + geom_abline()
}

long_data_met_norm %>% dplyr::select(c(Window, name, value)) %>% 
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  dplyr::select(-c(Window)) %>% 
  sample_frac(.01) %>% 
  GGally::ggpairs(columns = 1:10, lower = list(continuous = wrap(lowerfun))) + 
    theme_paper(textsize = 10) + scale_x_log10() + scale_y_log10()

chromosome_colors <- setNames(c(rep(c("grey", "black"), 9), "grey", "red", "blue"), nm = paste0("chr", c(1:19, "X", "Y")))

### look at individual sample pairs that seem off Sample7 / 17 (21d dox vs 17d dox)
long_data_met_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  dplyr::filter(name %in% c("Sample17", "Sample7")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>% 
  sample_frac(0.1) %>%
  # dplyr::filter(Chr == "chr2") %>%
  ggplot(aes(x = Sample7, y = Sample17, col = Chr)) + geom_point(size = .1) + geom_abline() + scale_x_log10() + scale_y_log10() + 
    theme_paper() + scale_color_manual(values = chromosome_colors)

# plot by windows
long_data_met_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  dplyr::filter(name %in% c("Sample17", "Sample7")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>% 
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  sample_frac(0.1) %>%
  ggplot(aes(x = Window, log2(Sample7 / Sample17), col = Chr)) + 
    geom_point(size = .1) + theme_paper() + geom_hline(yintercept = 0, color = "cyan") + 
    scale_color_manual(values = chromosome_colors) + 
    geom_hline(yintercept = c(-1, 1), linetype = "dashed")
# this is really weird. what are these spikes??

long_data_met_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  # dplyr::filter(name %in% c("Sample19", "Sample15")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  mutate(across(starts_with('Sample'), ~ . / Sample18) ) %>%
  sample_frac(0.1) %>%
  pivot_longer(-c(Window, Chr, Pos)) %>%
  mutate(name2 = colData(deseq_met)[.$name, ]$Condition) %>%
  ggplot(aes(x = Window, log2(value), col = Chr)) + 
    geom_point(size = .1) + theme_paper() + geom_hline(yintercept = 0, color = "cyan") + 
    scale_color_manual(values = chromosome_colors) + 
    geom_hline(yintercept = c(-1, 1), linetype = "dashed") + 
    facet_wrap(~name2)
ggsave("./plots/preprocessing/normalization_me3.pdf", width = 20, height = 12)
ggsave("./plots/preprocessing/normalization_me3.png", width = 20, height = 12)

### confirm with the other long sample
### look at individual sample pairs that seem off Sample5 / 17 (21d dox vs 17d dox)
long_data_met_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  dplyr::filter(name %in% c("Sample19", "Sample5")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>% 
  sample_frac(0.1) %>%
  ggplot(aes(x = Sample5, y = Sample19, col = Chr)) + geom_point(size = .1) + geom_abline() + scale_x_log10() + scale_y_log10() + 
    theme_paper() + scale_color_manual(values = chromosome_colors)

# plot by windows
long_data_met_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  dplyr::filter(name %in% c("Sample18", "Sample20")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>% 
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  sample_frac(0.1) %>%
  ggplot(aes(x = Window, log2(Sample5 / Sample19), col = Chr)) + 
    geom_point(size = .1) + theme_paper() + geom_hline(yintercept = 0, color = "cyan") + 
    scale_color_manual(values = chromosome_colors) + 
    geom_hline(yintercept = c(-1, 1), linetype = "dashed")

### look at individual sample pairs that seem off Sample11 / 13
long_data_met_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  dplyr::filter(name %in% c("Sample11", "Sample13")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>% 
  sample_frac(0.1) %>%
  ggplot(aes(x = Sample11, y = Sample13, col = Chr)) + geom_point(size = .1) + geom_abline() + scale_x_log10() + scale_y_log10() + 
    theme_paper() + scale_color_manual(values = chromosome_colors)

# plot by windows
long_data_met_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  dplyr::filter(name %in% c("Sample11", "Sample13")) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>% 
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  sample_frac(0.1) %>%
  ggplot(aes(x = Window, log2(Sample11 / Sample13), col = Chr)) + 
    geom_point(size = .1) + theme_paper() + geom_hline(yintercept = 0, color = "cyan") + 
    scale_color_manual(values = chromosome_colors) + 
    geom_hline(yintercept = c(-1, 1), linetype = "dashed")

```