---
title: "cut_and_run_normalization_check.R"
output: html_document
date: "2023-11-03"
editor_options: 
  chunk_output_type: console
---

Here I'm investigating the normalization of CnR data, based on windowed counts generated by Antonia. 
Here I'm using R4.3.0

```{r asss}

# library(csaw)
library(tidyverse)
library(DESeq2)

theme_paper <- function(textsize = 20){
  theme(panel.background = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1), 
        text = element_text(size = textsize, color = "black"), 
        axis.text = element_text(size = textsize, color = "black"), 
        axis.ticks = element_line(colour = 'black', size = 1), 
        axis.ticks.length = unit(.25, "cm"))
}

setwd("~/Desktop/PhD/Projects/XChromosome_Antonia/")

# test <- readRDS("./from_antonia/[Extern]_-_CaR_normalization/bg_10000_Gall_noX_E6Xist")
test <- readRDS("~/Desktop/Revision Antonia Paper/bg_promoters_-500+1000_all (1).bam_allchr_E6Xist")
test <- readRDS("~/Desktop/Revision Antonia Paper/bg_GeneBodies_all (1).bam_allchr_E6Xist")
test

# reorganize into allelically resolved: 
reorganize_sce <- function(sce){
  sce.hap1 <- sce[ , grepl("C57BL-6J", sce$bam.files)]
  sce.hap2 <- sce[ , grepl("CAST-EiJ", sce$bam.files)]
  sce.total <- sce[ , grepl("Gall", sce$bam.files)]
  sce.output <- SummarizedExperiment(assays = 
                                       list("counts_reference" = assays(sce.hap1)[["counts"]], 
                                            "counts_alternative" = assays(sce.hap2)[["counts"]], 
                                            "counts_total" = assays(sce.total)[["counts"]]), 
                                     rowRanges = rowRanges(sce), 
                                     colData = colData(sce.total))
  sce.output
}

test <- reorganize_sce(test)

condition_conversion <- c(
  "NoDox" = "Control",
  "7dDox" = "7d Dox",
  "7dD7dWO" = "7d Dox (Washout)",
  "21dD" = "21d Dox",
  "21dDox" = "21d Dox",
  "21dD7dWO" = "21d Dox (Washout)"
)

# get size factors from full dataset: 
total_counts_dataset <- readRDS("./Data/cut_and_run_antonia/all.genes_all.bam_all.chr_E6Xist")

deseq_sizefactors_ubi <- readRDS("./deseq_ubi_total.rds") %>% colData() %>% data.frame() %>% pull(sizeFactor, name = Condition)
deseq_sizefactors_met <- readRDS("./deseq_met_total.rds") %>% colData() %>% data.frame() %>% pull(sizeFactor, name = Condition)

sfdf_ubi <- readRDS("./deseq_ubi_total.rds") %>% colData() %>% data.frame() %>% dplyr::select(c("bam.files", "HistoneMark", "Condition", "Replicate", "sizeFactor")) %>%
  mutate(bam.files = basename(bam.files))
sfdf_met <- readRDS("./deseq_met_total.rds") %>% colData() %>% data.frame() %>% dplyr::select(c("bam.files", "HistoneMark", "Condition", "Replicate", "sizeFactor")) %>%
  mutate(bam.files = basename(bam.files))

rbind(sfdf_ubi, sfdf_met) %>% write.csv("~/Desktop/size_factors_cnr.csv")

```

First read the data and correct some annotation errors: 

```{r assdasda}

file.names <- basename(colData(test)$bam.files)

metadata <- data.frame(
  Clone = "E6", 
  HistoneMark = str_extract(file.names, "H2AK119ubi|H3K27me3"), 
  Condition = gsub("_(H2AK119ubi|H3K27me3)", "", gsub("E6", "", gsub("_Gall_sorted.bam", "", file.names)))
) %>%
  mutate(Replicate = str_extract(Condition, "r[0-9]")) %>%
  mutate(Dox = str_extract(Condition, "(21d|7d|No)")) %>% 
  mutate(Dox = as.numeric(gsub("d", "", Dox))) %>% mutate(Dox = ifelse(is.na(Dox), 0, Dox)) %>%
  mutate(Washout = ifelse(grepl("WO", Condition), 7, 0))

### assume that sample 5/6 and 13/14 are swapped
metadata_corrected <- metadata[c(1:4, 6, 5, 7:12, 14, 13, 15:20), ]
metadata_corrected$Name <- file.names[c(1:4, 6, 5, 7:12, 14, 13, 15:20)]
# metadata_corrected <- metadata
rownames(metadata_corrected) <- paste0("Sample", 1:nrow(metadata))

colData(test) <- cbind(colData(test), metadata_corrected)

colnames(test) <- paste0("Sample", 1:nrow(metadata_corrected))

```

Here we look at X-chromosomal histone mark deposition for the inactive X

```{r }

# Subset the data on the met samples
test_ubi <- test[ , test$HistoneMark == "H2AK119ubi"]
test_ubi$size_factor <- deseq_sizefactors_ubi[test_ubi$Condition]
test_ubi <- test_ubi[ seqnames(rowRanges(test_ubi)) == "chrX" , ]

assays(test_ubi)[["counts_reference"]] %>% 
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Chr = as.character(seqnames(rowRanges(test_ubi)))) %>%
  add_column(Pos = start(rowRanges(test_ubi)) + 5000) %>%
  pivot_longer(-c(Gene, Chr, Pos)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) %>%
  mutate(XChromosome = "REF") -> long_data_ubi_ref

assays(test_ubi)[["counts_alternative"]] %>% 
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Chr = as.character(seqnames(rowRanges(test_ubi)))) %>%
  add_column(Pos = start(rowRanges(test_ubi)) + 5000) %>%
  pivot_longer(-c(Gene, Chr, Pos)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) %>%
  mutate(XChromosome = "ALT") -> long_data_ubi_alt

long_data_ubi <- rbind(long_data_ubi_ref, long_data_ubi_alt) %>%
  mutate(size_factor = deseq_sizefactors_ubi[Condition]) %>%
  mutate(value = value / size_factor)

# set chromosome colors
chromosome_colors <- setNames(c(rep(c("grey", "black"), 9), "grey", "red", "blue"), nm = paste0("chr", c(1:19, "X", "Y")))

# plot scatters for each sample against the baseline coverage (mean of the two untreated samples)
# it is clear that the X in many samples is globally changing, and there is a skew at lower coverages in most samples (which is why we avoid using 
# them for size factor calculation)
long_data_ubi %>% dplyr::select(c(Gene, Chr, Pos, name, value, XChromosome)) %>% 
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  # mutate(Window = as.numeric(gsub("Window", "", Window))) %>% 
  mutate(sample_ref = (Sample17 + Sample19) / 2) %>% 
  pivot_longer(-c(Gene, Chr, Pos, XChromosome, sample_ref)) %>%
  mutate(name2 = colData(test_ubi)[.$name, ]$Condition) %>%
  ggplot(aes(x = sample_ref + 1, value + 1, col = Chr)) + 
    geom_point(size = .1) + theme_paper() + 
    scale_color_manual(values = chromosome_colors) + 
    geom_abline() + 
    scale_x_log10() + scale_y_log10() + facet_wrap(~XChromosome)
ggsave("./Plots/allelic_analysis/ubi_delta_ctrl_treated_alt_ref.pdf", width = 10, height = 10)

long_data_ubi %>% dplyr::select(c(Gene, Chr, Pos, name, value, XChromosome)) %>%
  mutate(name2 = colData(test_ubi)[.$name, ]$Condition) %>%
  mutate(size_factor = deseq_sizefactors_ubi[name2]) %>%
  mutate(value = value / size_factor) %>%
  group_by(Gene) %>%
  mutate(total_counts = sum(value)) %>%
  dplyr::filter(total_counts > 40) %>% 
  mutate(Replicate = ifelse(grepl("r2", name2), "r2", "r3")) %>%
  ggplot(aes(x = name2, y = value + 1, fill = XChromosome)) + geom_boxplot() + scale_y_log10() + theme_paper() + coord_flip() + 
    facet_wrap(~Replicate, scales = "free", ncol = 1)
ggsave("./Plots/allelic_analysis/ubi_allelic_boxplots.pdf", width = 10, height = 10)

long_data_ubi %>% dplyr::select(c(Gene, Chr, Pos, name, value, XChromosome)) %>%
  mutate(name2 = colData(test_ubi)[.$name, ]$Condition) %>%
  mutate(size_factor = deseq_sizefactors_ubi[name2]) %>%
  group_by(Gene) %>%
  mutate(total_counts = sum(value)) %>%
  dplyr::filter(total_counts > 40) %>% 
  mutate(Replicate = ifelse(grepl("r2", name2), "r2", "r3")) %>% 
  pivot_wider(names_from = XChromosome, values_from = value) %>%
  mutate(ratio = (REF + 1) / (2 + REF + ALT)) %>%
  ggplot(aes(x = name2, y = ratio)) + geom_violin() + theme_paper() + 
    coord_flip() + stat_summary(fun = median) + 
    facet_wrap(~Replicate, scales = "free", ncol = 1)
ggsave("./Plots/allelic_analysis/ubi_allelic_ratio_violin.pdf", width = 10, height = 10)

### stratify by effects on escape: 
escape_results <- read_csv("./ProcessedData/figure3_d_score_differences.csv")

escape_results %>% 
  mutate(gene_category = case_when(
    - `Dox (21 days), Washout (0 days)_dif` < .2 & - `Dox (21 days), Washout (7 days)_dif` < .2 ~ "Unresponsive", 
    - `Dox (21 days), Washout (0 days)_dif` > .2 & - `Dox (21 days), Washout (7 days)_dif` > .2 ~ "Responsive, Irreversible",
    - `Dox (21 days), Washout (0 days)_dif` > .2 & - `Dox (21 days), Washout (7 days)_dif` < .2 ~ "Responsive, Reversible", 
    .default = "test")) %>%
  dplyr::rename(Symbol = name) -> gene_annotation

gene_annotation %>% 
  ggplot(aes(x =  - `Dox (21 days), Washout (0 days)_dif`,  - `Dox (21 days), Washout (7 days)_dif`)) + geom_point(aes(col = gene_category)) + 
    xlab("Dox effect") + ylab("Reversibility") + ggrepel::geom_text_repel(aes(label = Symbol))
ggsave("./Plots/allelic_analysis/reversibility_overview_plot.pdf", width = 10, height = 10)

all_genes <- genes(EnsDb.Mmusculus.v79)

long_data_ubi %>% dplyr::select(c(Gene, Chr, Pos, name, value, XChromosome)) %>%
  mutate(name2 = colData(test_ubi)[.$name, ]$Condition) %>%
  mutate(size_factor = deseq_sizefactors_ubi[name2]) %>%
  group_by(Gene) %>%
  mutate(total_counts = sum(value)) %>%
  dplyr::filter(total_counts > 20) %>% 
  mutate(Replicate = ifelse(grepl("r2", name2), "r2", "r3")) %>% 
  pivot_wider(names_from = XChromosome, values_from = value) %>%
  mutate(ratio = (REF + 1) / (2 + REF + ALT)) %>% 
  mutate(Symbol = all_genes[Gene, ]$gene_name) %>%
  right_join(gene_annotation[ , c("escape_status", "gene_category", "Symbol")]) -> merged_data_ubi
  
merged_data_ubi %>%
  dplyr::filter(!is.na(Replicate)) %>%
  ggplot(aes(x = name2, y = ratio)) + geom_boxplot(aes(fill = escape_status)) + theme_paper() + 
    coord_flip() + 
    facet_wrap(~Replicate, scales = "free", ncol = 1)
ggsave("./Plots/allelic_analysis/ubi_boxplot_escape_category.pdf", width = 10, height = 10)

merged_data_ubi %>%
  dplyr::filter(!is.na(Replicate)) %>%
  ggplot(aes(x = name2, y = ratio)) + geom_boxplot(aes(fill = gene_category)) + theme_paper() + 
    coord_flip() + 
    facet_wrap(~Replicate, scales = "free", ncol = 1)
ggsave("./Plots/allelic_analysis/ubi_boxplot_reversibility_category.pdf", width = 10, height = 10)

```

```{r }

# Subset the data on the met samples
test_met <- test[ , test$HistoneMark == "H3K27me3"]
test_met$size_factor <- deseq_sizefactors_met[test_met$Condition]
test_met <- test_met[ seqnames(rowRanges(test_met)) == "chrX" , ]

assays(test_met)[["counts_reference"]] %>% 
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Chr = as.character(seqnames(rowRanges(test_met)))) %>%
  add_column(Pos = start(rowRanges(test_met)) + 5000) %>%
  pivot_longer(-c(Gene, Chr, Pos)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) %>%
  mutate(XChromosome = "REF") -> long_data_met_ref

assays(test_met)[["counts_alternative"]] %>% 
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Chr = as.character(seqnames(rowRanges(test_met)))) %>%
  add_column(Pos = start(rowRanges(test_met)) + 5000) %>%
  pivot_longer(-c(Gene, Chr, Pos)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) %>%
  mutate(XChromosome = "ALT") -> long_data_met_alt

long_data_met <- rbind(long_data_met_ref, long_data_met_alt) %>%
  mutate(size_factor = deseq_sizefactors_met[Condition]) %>%
  mutate(value = value / size_factor)

# set chromosome colors
chromosome_colors <- setNames(c(rep(c("grey", "black"), 9), "grey", "red", "blue"), nm = paste0("chr", c(1:19, "X", "Y")))

# plot scatters for each sample against the baseline coverage (mean of the two untreated samples)
# it is clear that the X in many samples is globally changing, and there is a skew at lower coverages in most samples (which is why we avoid using 
# them for size factor calculation)
long_data_met %>% dplyr::select(c(Gene, Chr, Pos, name, value, XChromosome)) %>% 
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  # mutate(Window = as.numeric(gsub("Window", "", Window))) %>% 
  mutate(sample_ref = (Sample18 + Sample20) / 2) %>% 
  pivot_longer(-c(Gene, Chr, Pos, XChromosome, sample_ref)) %>%
  mutate(name2 = colData(test_met)[.$name, ]$Condition) %>%
  ggplot(aes(x = sample_ref + 1, value + 1, col = Chr)) + 
    geom_point(size = .1) + theme_paper() + 
    scale_color_manual(values = chromosome_colors) + 
    geom_abline() + 
    scale_x_log10() + scale_y_log10() + facet_wrap(~XChromosome)
ggsave("./Plots/allelic_analysis/met_delta_ctrl_treated_alt_ref.pdf", width = 10, height = 10)

long_data_met %>% dplyr::select(c(Gene, Chr, Pos, name, value, XChromosome)) %>%
  mutate(name2 = colData(test_met)[.$name, ]$Condition) %>%
  mutate(size_factor = deseq_sizefactors_met[name2]) %>%
  mutate(value = value / size_factor) %>%
  group_by(Gene) %>%
  mutate(total_counts = sum(value)) %>%
  dplyr::filter(total_counts > 40) %>% 
  mutate(Replicate = ifelse(grepl("r2", name2), "r2", "r3")) %>%
  ggplot(aes(x = name2, y = value + 1, fill = XChromosome)) + geom_boxplot() + scale_y_log10() + theme_paper() + coord_flip() + 
    facet_wrap(~Replicate, scales = "free", ncol = 1)
ggsave("./Plots/allelic_analysis/met_allelic_boxplots.pdf", width = 10, height = 10)

long_data_met %>% dplyr::select(c(Gene, Chr, Pos, name, value, XChromosome)) %>%
  mutate(name2 = colData(test_met)[.$name, ]$Condition) %>%
  mutate(size_factor = deseq_sizefactors_met[name2]) %>%
  group_by(Gene) %>%
  mutate(total_counts = sum(value)) %>%
  dplyr::filter(total_counts > 40) %>% 
  mutate(Replicate = ifelse(grepl("r2", name2), "r2", "r3")) %>% 
  pivot_wider(names_from = XChromosome, values_from = value) %>%
  mutate(ratio = (REF + 1) / (2 + REF + ALT)) %>%
  ggplot(aes(x = name2, y = ratio)) + geom_violin() + theme_paper() + 
    coord_flip() + stat_summary(fun = median) + 
    facet_wrap(~Replicate, scales = "free", ncol = 1)
ggsave("./Plots/allelic_analysis/met_allelic_ratio_violin.pdf", width = 10, height = 10)

### stratify by effects on escape: 
escape_results <- read_csv("./ProcessedData/figure3_d_score_differences.csv")

escape_results %>% 
  mutate(gene_category = case_when(
    - `Dox (21 days), Washout (0 days)_dif` < .2 & - `Dox (21 days), Washout (7 days)_dif` < .2 ~ "Unresponsive", 
    - `Dox (21 days), Washout (0 days)_dif` > .2 & - `Dox (21 days), Washout (7 days)_dif` > .2 ~ "Responsive, Irreversible",
    - `Dox (21 days), Washout (0 days)_dif` > .2 & - `Dox (21 days), Washout (7 days)_dif` < .2 ~ "Responsive, Reversible", 
    .default = "test")) %>%
  dplyr::rename(Symbol = name) -> gene_annotation

gene_annotation %>% 
  ggplot(aes(x =  - `Dox (21 days), Washout (0 days)_dif`,  - `Dox (21 days), Washout (7 days)_dif`)) + geom_point(aes(col = gene_category)) + 
    xlab("Dox effect") + ylab("Reversibility") + ggrepel::geom_text_repel(aes(label = Symbol))
ggsave("./Plots/allelic_analysis/reversibility_overview_plot.pdf", width = 10, height = 10)

all_genes <- genes(EnsDb.Mmusculus.v79)

long_data_met %>% dplyr::select(c(Gene, Chr, Pos, name, value, XChromosome)) %>%
  mutate(name2 = colData(test_met)[.$name, ]$Condition) %>%
  mutate(size_factor = deseq_sizefactors_met[name2]) %>%
  group_by(Gene) %>%
  mutate(total_counts = sum(value)) %>%
  dplyr::filter(total_counts > 20) %>% 
  mutate(Replicate = ifelse(grepl("r2", name2), "r2", "r3")) %>% 
  pivot_wider(names_from = XChromosome, values_from = value) %>%
  mutate(ratio = (REF + 1) / (2 + REF + ALT)) %>% 
  mutate(Symbol = all_genes[Gene, ]$gene_name) %>%
  right_join(gene_annotation[ , c("escape_status", "gene_category", "Symbol")]) -> merged_data_met
  
merged_data_met %>%
  dplyr::filter(!is.na(Replicate)) %>%
  ggplot(aes(x = name2, y = ratio)) + geom_boxplot(aes(fill = escape_status)) + theme_paper() + 
    coord_flip() + 
    facet_wrap(~Replicate, scales = "free", ncol = 1)
ggsave("./Plots/allelic_analysis/met_boxplot_escape_category.pdf", width = 10, height = 10)

merged_data_met %>%
  dplyr::filter(!is.na(Replicate)) %>%
  ggplot(aes(x = name2, y = ratio)) + geom_boxplot(aes(fill = gene_category)) + theme_paper() + 
    coord_flip() + 
    facet_wrap(~Replicate, scales = "free", ncol = 1)
ggsave("./Plots/allelic_analysis/met_boxplot_reversibility_category.pdf", width = 10, height = 10)

```
