---
title: "cut_and_run_normalization_check.R"
output: html_document
date: "2023-11-03"
editor_options: 
  chunk_output_type: console
---

Here I'm investigating the normalization of CnR data, based on windowed counts generated by Antonia. 
Here I'm using R4.3.0

```{r asss}

library(tidyverse)
library(DESeq2)

theme_paper <- function(textsize = 20){
  theme(panel.background = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1), 
        text = element_text(size = textsize, color = "black"), 
        axis.text = element_text(size = textsize, color = "black"), 
        axis.ticks = element_line(colour = 'black', size = 1), 
        axis.ticks.length = unit(.25, "cm"))
}

setwd("~/Desktop/PhD/Projects/XChromosome_Antonia/")

# test <- readRDS("./from_antonia/[Extern]_-_CaR_normalization/bg_10000_Gall_noX_E6Xist")
test <- readRDS("./Data/cut_and_run_antonia/bg_10000_all.bam_allchr_E6Xist")

# testy <- readRDS("./Data/cut_and_run_antonia/all.genes_all.bam_all.chr_E6Xist")
# assays(testy)[["counts"]]

condition_conversion <- c(
  "NoDox" = "Control",
  "7dDox" = "7d Dox",
  "7dD7dWO" = "7d Dox (Washout)",
  "21dD" = "21d Dox",
  "21dDox" = "21d Dox",
  "21dD7dWO" = "21d Dox (Washout)"
)

```

First read the data and correct some annotation errors: 

```{r assdasda}

file.names <- basename(colData(test)$bam.files)

metadata <- data.frame(
  Clone = "E6", 
  HistoneMark = str_extract(file.names, "H2AK119ubi|H3K27me3"), 
  Condition = gsub("_(H2AK119ubi|H3K27me3)", "", gsub("E6", "", gsub("_Gall_sorted.bam", "", file.names)))
) %>%
  mutate(Replicate = str_extract(Condition, "r[0-9]")) %>%
  mutate(Dox = str_extract(Condition, "(21d|7d|No)")) %>% 
  mutate(Dox = as.numeric(gsub("d", "", Dox))) %>% mutate(Dox = ifelse(is.na(Dox), 0, Dox)) %>%
  mutate(Washout = ifelse(grepl("WO", Condition), 7, 0))

### assume that sample 5/6 and 13/14 are swapped
metadata_corrected <- metadata[c(1:4, 6, 5, 7:12, 14, 13, 15:20), ]
metadata_corrected$Name <- file.names[c(1:4, 6, 5, 7:12, 14, 13, 15:20)]
metadata_corrected <- metadata
metadata_corrected$Name <- file.names
# metadata_corrected <- metadata
rownames(metadata_corrected) <- paste0("Sample", 1:nrow(metadata))

colData(test) <- cbind(colData(test), metadata_corrected)

colnames(test) <- paste0("Sample", 1:nrow(metadata_corrected))

```

Visualize the distributions per sample

```{r asdass}

assays(test)[[1]] %>% 
  data.frame() %>%
  pivot_longer(-c()) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) -> long_data

long_data %>%
  ggplot(aes(x = Condition, y = value + 1, fill = Condition)) + geom_violin() + scale_y_log10() + facet_wrap(~assay, scales = "free") + 
    coord_flip() + geom_text(aes(label = name, y = 10), data = head(long_data, n = 20)) + 
    theme_paper(textsize = 10) + xlab("") + ylab("Number of reads (+1)")
ggsave("./Plots/total_analysis/count_distributions_both_marks.pdf")

# there is a clear need for normalization

```

First, we focus on the H2Ak119ubi mark. 

```{r normalize_ubi}

# Subset the data on the ubi samples
test_ubi <- test[ , test$HistoneMark == "H2AK119ubi"]

# plot average coverage distribution per window - we remove windows with < 10 reads on average
rowMeans(assays(test_ubi)[[1]]) %>% 
  data.frame() %>% ggplot(aes(x = .)) + geom_histogram() + scale_x_log10() + 
    geom_vline(xintercept = 10) + geom_vline(xintercept = 100) + 
    theme_paper() + 
    xlab("Average number of reads (+1)") + ylab("Number of windows")

# make ans subset deseq dataset
deseq_ubi <- DESeqDataSet(test_ubi, design = ~0)
deseq_ubi <- deseq_ubi[rowMeans(counts(deseq_ubi)) > 10, ]

# extract and rename count data
counts(deseq_ubi, normalized = F) %>% 
  data.frame() %>%
  add_column(Window = paste0("Window", 1:nrow(deseq_ubi))) %>%
  pivot_longer(-c(Window)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) -> long_data_ubi

# plot data - clearly, there is a need to normalize for library size
long_data_ubi %>%
  ggplot(aes(x = Condition, y = value + 1, fill = Condition)) + geom_violin() + scale_y_log10() + facet_wrap(~assay, scales = "free") + 
    coord_flip() + geom_text(aes(label = name, y = 10), data = head(long_data_ubi, n = 20)) + 
    theme_paper(textsize = 10) + xlab("") + ylab("Number of reads (+1)")
ggsave("./Plots/total_analysis/count_distributions_ubi.pdf")

# compute deseq size factors, but exclude the X and lowly covered regions for this calculation (will be clear further down)
deseq_ubi_filt <- estimateSizeFactors(deseq_ubi[rowMeans(counts(deseq_ubi)) > 100 & seqnames(rowRanges(deseq_ubi)) != "chrX", ])
deseq_ubi$sizeFactor <- deseq_ubi_filt$sizeFactor

# correlation maps
norm_data_matrix_ubi <- log(counts(deseq_ubi, normalized = T) + 1)
colnames(norm_data_matrix_ubi) <- deseq_ubi$Condition
norm_data_matrix_ubi <- cor(norm_data_matrix_ubi)
pdf("./Plots/total_analysis/ubi_heatmap_all.pdf")
pheatmap::pheatmap(norm_data_matrix_ubi)
dev.off()

norm_data_matrix_ubi <- log(counts(deseq_ubi[seqnames(rowRanges(deseq_ubi)) != "chrX", ], normalized = T) + 1)
colnames(norm_data_matrix_ubi) <- deseq_ubi$Condition
norm_data_matrix_ubi <- cor(norm_data_matrix_ubi)
pdf("./Plots/total_analysis/ubi_heatmap_autosomes.pdf")
pheatmap::pheatmap(norm_data_matrix_ubi)
dev.off()

norm_data_matrix_ubi <- log(counts(deseq_ubi[seqnames(rowRanges(deseq_ubi)) == "chrX", ], normalized = T) + 1)
colnames(norm_data_matrix_ubi) <- deseq_ubi$Condition
norm_data_matrix_ubi <- cor(norm_data_matrix_ubi)
pdf("./Plots/total_analysis/ubi_heatmap_chrX.pdf")
pheatmap::pheatmap(norm_data_matrix_ubi)
dev.off()

# extract normalized data as above
counts(deseq_ubi, normalized = T) %>% 
  data.frame() %>%
  add_column(Window = paste0("Window", 1:nrow(deseq_ubi))) %>%
  add_column(Chr = as.character(seqnames(rowRanges(deseq_ubi)))) %>%
  add_column(Pos = start(rowRanges(deseq_ubi)) + 5000) %>%
  pivot_longer(-c(Window, Chr, Pos)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) -> long_data_ubi_norm

# normalization solves unequal coverage issue
long_data_ubi_norm %>%
  ggplot(aes(x = Condition, y = value + 1, fill = Condition)) + geom_violin() + scale_y_log10() + facet_wrap(~assay, scales = "free") + 
    coord_flip() + geom_text(aes(label = name, y = 10), data = head(long_data_ubi_norm, n = 20)) + 
    theme_paper(textsize = 10) + xlab("") + ylab("Number of reads (+1)")
ggsave("./Plots/total_analysis/count_distributions_ubi_normalized.pdf")

# set chromosome colors
chromosome_colors <- setNames(c(rep(c("grey", "black"), 9), "grey", "red", "blue"), nm = paste0("chr", c(1:19, "X", "Y")))

# plot scatters for each sample against the baseline coverage (mean of the two untreated samples)
# it is clear that the X in many samples is globally changing, and there is a skew at lower coverages in most samples (which is why we avoid using 
# them for size factor calculation)
long_data_ubi_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  mutate(sample_ref = (Sample17 + Sample19) / 2) %>%
  sample_frac(0.1) %>%
  pivot_longer(-c(Window, Chr, Pos, sample_ref)) %>%
  mutate(name2 = colData(deseq_ubi)[.$name, ]$Condition) %>%
  ggplot(aes(x = sample_ref + 1, value + 1, col = Chr)) + 
    geom_point(size = .1) + theme_paper() + 
    scale_color_manual(values = chromosome_colors) + 
    geom_abline() + 
    scale_x_log10() + scale_y_log10() + 
    facet_wrap(~name2)
ggsave("./Plots/total_analysis/correlation_plots_ubi_normalized.pdf")

# plot equivalently the fold change to the reference coverage
long_data_ubi_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  mutate(across(starts_with('Sample'), ~ . / ((Sample17 + Sample19) / 2)) ) %>%
  sample_frac(0.1) %>%
  pivot_longer(-c(Window, Chr, Pos)) %>%
  mutate(name2 = colData(deseq_ubi)[.$name, ]$Condition) %>%
  ggplot(aes(x = Window, log2(value), col = Chr)) + 
    geom_point(size = .1) + theme_paper() + geom_hline(yintercept = 0, color = "cyan") + 
    scale_color_manual(values = chromosome_colors) + 
    geom_hline(yintercept = c(-1, 1), linetype = "dashed") + 
    facet_wrap(~name2)
ggsave("./Plots/total_analysis/normalization_ubi.pdf", width = 20, height = 12)
ggsave("./Plots/total_analysis/normalization_ubi.png", width = 20, height = 12)

# extract fold change data as in the above plot for further analysis
long_data_ubi_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  mutate(across(starts_with('Sample'), ~ . / ((Sample17 + Sample19) / 2)) ) %>%
  sample_frac(0.1) %>%
  pivot_longer(-c(Window, Chr, Pos)) %>%
  mutate(name2 = colData(deseq_ubi)[.$name, ]$Condition) -> normalized_data_ubi

library(EnsDb.Mmusculus.v79)
genes <- genes(EnsDb.Mmusculus.v79)
genes <- genes[genes$gene_biotype %in% c("protein_coding", "lincRNA"), ]
# seqlevelsStyle(genes) <- "UCSC"

# annotate windows wrt whether they overlap any genes
normalized_data_granges <- GRanges(gsub("chr", "", normalized_data_ubi$Chr), normalized_data_ubi$Pos)
normalized_data_ubi$distance <- distanceToNearest(normalized_data_granges, genes)@elementMetadata$distance

# first investigate the autosomal signal
normalized_data_ubi %>%
  mutate(name2 = gsub("r2|r3", "", name2)) %>%
  dplyr::select(c(Pos, name2, Chr, value, distance)) %>%
  group_by(Pos, name2, Chr, distance) %>%
  summarize(value = mean(value)) %>%
  mutate(name2 = factor(condition_conversion[name2], levels = rev(c("Control", "7d Dox", "7d Dox (Washout)", "21d Dox", "21d Dox (Washout)")))) %>%
  dplyr::filter(name2 != "Control") %>% 
  dplyr::filter(Chr != "chrX") -> processed_data

# there is slightly more enrichment in regions not overlapping genes 
processed_data %>% 
  mutate(distance = log10(distance + 1)) %>%
  mutate(distance = cut(distance, breaks = c(-1, 2, 4, 10))) %>%
  mutate(value = ifelse(value > 5, 5, value)) %>%
  ggplot(aes(distance, y = value)) + geom_violin() + facet_wrap(~name2) + geom_hline(yintercept = 1) + coord_flip() + 
    stat_summary() + theme_paper()
ggsave("./Plots/total_analysis/autosomal_global_ubi.pdf", width = 20, height = 12)

# calculate number of windows with substantial gain (at least 50% more than baseline)
processed_data %>% 
  mutate(distance = log10(distance + 1)) %>%
  mutate(overlap = ifelse(distance > 0, "no overlap", "overlap")) %>%
  mutate(sig_gain = value > 1.5) %>%
  ggplot(aes(x = overlap, fill = sig_gain)) + geom_bar(position = "fill") + facet_wrap(~name2, nrow = 1) + 
    theme_paper() + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggsave("./Plots/total_analysis/autosomal_global_ubi2.pdf", width = 20, height = 12)

normalized_data_ubi %>% 
  mutate(name2 = gsub("r2|r3", "", name2)) %>%
  dplyr::select(c(Pos, name2, Chr, value)) %>%
  # group_by(Pos, name2, Chr) %>%
  # summarize(value = mean(value)) %>%
  dplyr::filter(Chr == "chrX") %>%
  mutate(name2 = factor(condition_conversion[name2], levels = rev(c("Control", "7d Dox", "7d Dox (Washout)", "21d Dox", "21d Dox (Washout)")))) %>%
  ggplot(aes(x = name2, y = value, fill = )) + geom_boxplot(fill = "grey") + 
    theme_paper() + coord_flip() + geom_hline(yintercept = 1, linetype = 'dashed') + 
    ylab("Count Fold Change (to Control)") + xlab("")
ggsave("./Plots/total_analysis/ubi_levels.pdf", width = 20, height = 12)

# check whether promoters of genes that change in expression change more
genes_here <- genes(EnsDb.Mmusculus.v79) %>% data.frame() %>% dplyr::rename(Gene = "gene_name") %>% dplyr::select(c("seqnames", "start", "end", Gene))

autosomal_des <- readRDS("~/Desktop/PhD/Projects/XChromosome_Antonia/ProcessedData/autosomal_de_analysis_7d_dox.rds") %>%
  right_join(genes_here, by = "Gene") %>% 
  mutate(seqnames = paste0("chr", seqnames)) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = T)

normalized_data_ubi_granges <- normalized_data_ubi %>% 
  mutate(seqnames = Chr, start = Pos - 250, stop = Pos + 250) %>%
  makeGRangesFromDataFrame(seqnames.field = "seqnames", start.field = "start", end.field = "stop", keep.extra.columns = T)

overlaps_here <- findOverlaps(normalized_data_ubi_granges, autosomal_des, minoverlap = 200)

normalized_data_ubi_granges$overlapping_gene <- NA
normalized_data_ubi_granges$expression_foldchange <- NA
normalized_data_ubi_granges[from(overlaps_here), ]$overlapping_gene <- autosomal_des[to(overlaps_here), ]$Gene
normalized_data_ubi_granges[from(overlaps_here), ]$expression_foldchange <- autosomal_des[to(overlaps_here), ]$log2FoldChange

normalized_data_ubi_granges %>%
  data.frame() %>% 
  ggplot(aes(x = expression_foldchange, y = value)) + facet_wrap(~name2) + geom_point(size = .1) + theme_paper()
ggsave("./Plots/total_analysis/autosomal_gains_ubi.pdf", width = 20, height = 12)

normalized_data_ubi_granges %>%
  data.frame() %>% 
  mutate(seqnames2 = seqnames != "chrX") %>%
  dplyr::filter(name2 == "21dDr3") %>% {
    ggplot(data = ., aes(x = expression_foldchange, y = value)) + geom_point(size = 0.1)  + 
    ggrepel::geom_text_repel(data = . %>% dplyr::filter(expression_foldchange < -2 | value > 2), aes(label = overlapping_gene))
  }

# normalized_data_ubi_granges %>%
#   data.frame() %>% 
#   mutate(seqnames2 = seqnames != "chrX") %>%
#   dplyr::filter(name2 == "21dDr3") %>% {
#     ggplot(data = ., aes(x = expression_foldchange, y = value)) + geom_point(size = 0.1)  + 
#     ggrepel::geom_text_repel(data = . %>% dplyr::filter(expression_foldchange < -2 | value > 2), 
#                              aes(label = seqnames))
#   }

saveRDS(deseq_ubi, "./deseq_ubi_total.rds")

```

Second, we focus on the H3K27me3 mark. 

```{r normalize_ubi}

# Subset the data on the met samples
test_met <- test[ , test$HistoneMark == "H3K27me3"]

# plot average coverage distribution per window - we remove windows with < 10 reads on average
rowMeans(assays(test_met)[[1]]) %>% 
  data.frame() %>% ggplot(aes(x = .)) + geom_histogram() + scale_x_log10() + 
    geom_vline(xintercept = 10) + geom_vline(xintercept = 100) + 
    theme_paper() + 
    xlab("Average number of reads (+1)") + ylab("Number of windows")

# make ans subset deseq dataset
deseq_met <- DESeqDataSet(test_met, design = ~0)
deseq_met <- deseq_met[rowMeans(counts(deseq_met)) > 10, ]

# extract and rename count data
counts(deseq_met, normalized = F) %>% 
  data.frame() %>%
  add_column(Window = paste0("Window", 1:nrow(deseq_met))) %>%
  pivot_longer(-c(Window)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) -> long_data_met

# plot data - clearly, there is a need to normalize for library size
long_data_met %>%
  ggplot(aes(x = Condition, y = value + 1, fill = Condition)) + geom_violin() + scale_y_log10() + facet_wrap(~assay, scales = "free") + 
    coord_flip() + geom_text(aes(label = name, y = 10), data = head(long_data_met, n = 20)) + 
    theme_paper(textsize = 10) + xlab("") + ylab("Number of reads (+1)")
ggsave("./Plots/total_analysis/count_distributions_met.pdf")

# compute deseq size factors, but exclude the X and lowly covered regions for this calculation (will be clear further down)
deseq_met_filt <- estimateSizeFactors(deseq_met[rowMeans(counts(deseq_met)) > 300 & seqnames(rowRanges(deseq_met)) != "chrX", ])
deseq_met$sizeFactor <- deseq_met_filt$sizeFactor

# correlation maps
norm_data_matrix_met <- log(counts(deseq_met, normalized = T) + 1)
colnames(norm_data_matrix_met) <- deseq_met$Condition
norm_data_matrix_met <- cor(norm_data_matrix_met)
pdf("./Plots/total_analysis/met_heatmap_all.pdf")
pheatmap::pheatmap(norm_data_matrix_met)
dev.off()

norm_data_matrix_met <- log(counts(deseq_met[seqnames(rowRanges(deseq_met)) != "chrX", ], normalized = T) + 1)
colnames(norm_data_matrix_met) <- deseq_met$Condition
norm_data_matrix_met <- cor(norm_data_matrix_met)
pdf("./Plots/total_analysis/met_heatmap_autosomes.pdf")
pheatmap::pheatmap(norm_data_matrix_met)
dev.off()

norm_data_matrix_met <- log(counts(deseq_met[seqnames(rowRanges(deseq_met)) == "chrX", ], normalized = T) + 1)
colnames(norm_data_matrix_met) <- deseq_met$Condition
norm_data_matrix_met <- cor(norm_data_matrix_met)
pdf("./Plots/total_analysis/met_heatmap_chrX.pdf")
pheatmap::pheatmap(norm_data_matrix_met)
dev.off()

# extract normalized data as above
counts(deseq_met, normalized = T) %>% 
  data.frame() %>%
  add_column(Window = paste0("Window", 1:nrow(deseq_met))) %>%
  add_column(Chr = as.character(seqnames(rowRanges(deseq_met)))) %>%
  add_column(Pos = start(rowRanges(deseq_met)) + 5000) %>%
  pivot_longer(-c(Window, Chr, Pos)) %>% 
  add_column(assay = metadata_corrected[.$name, ]$HistoneMark) %>%
  add_column(Condition = metadata_corrected[.$name, ]$Condition) -> long_data_met_norm

# normalization solves unequal coverage issue
long_data_met_norm %>%
  ggplot(aes(x = Condition, y = value + 1, fill = Condition)) + geom_violin() + scale_y_log10() + facet_wrap(~assay, scales = "free") + 
    coord_flip() + geom_text(aes(label = name, y = 10), data = head(long_data_met_norm, n = 20)) + 
    theme_paper(textsize = 10) + xlab("") + ylab("Number of reads (+1)")
ggsave("./Plots/total_analysis/count_distributions_met_normalized.pdf")

# set chromosome colors
chromosome_colors <- setNames(c(rep(c("grey", "black"), 9), "grey", "red", "blue"), nm = paste0("chr", c(1:19, "X", "Y")))

# plot scatters for each sample against the baseline coverage (mean of the two untreated samples)
# it is clear that the X in many samples is globally changing, and there is a skew at lower coverages in most samples (which is why we avoid using 
# them for size factor calculation)
long_data_met_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  mutate(sample_ref = (Sample18 + Sample20) / 2) %>%
  sample_frac(0.1) %>%
  pivot_longer(-c(Window, Chr, Pos, sample_ref)) %>%
  mutate(name2 = colData(deseq_met)[.$name, ]$Condition) %>%
  ggplot(aes(x = sample_ref + 1, value + 1, col = Chr)) + 
    geom_point(size = .1) + theme_paper() + 
    scale_color_manual(values = chromosome_colors) + 
    geom_abline() + 
    scale_x_log10() + scale_y_log10() + 
    facet_wrap(~name2)
ggsave("./Plots/total_analysis/correlation_plots_met_normalized.pdf")

# plot equivalently the fold change to the reference coverage
long_data_met_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  mutate(across(starts_with('Sample'), ~ . / ((Sample18 + Sample20) / 2)) ) %>%
  sample_frac(0.1) %>%
  pivot_longer(-c(Window, Chr, Pos)) %>%
  mutate(name2 = colData(deseq_met)[.$name, ]$Condition) %>%
  ggplot(aes(x = Window, log2(value), col = Chr)) + 
    geom_point(size = .1) + theme_paper() + geom_hline(yintercept = 0, color = "cyan") + 
    scale_color_manual(values = chromosome_colors) + 
    geom_hline(yintercept = c(-1, 1), linetype = "dashed") + 
    facet_wrap(~name2)
ggsave("./Plots/total_analysis/normalization_met.pdf", width = 20, height = 12)
ggsave("./Plots/total_analysis/normalization_met.png", width = 20, height = 12)

# extract fold change data as in the above plot for further analysis
long_data_met_norm %>% dplyr::select(c(Window, Chr, Pos, name, value)) %>% 
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  mutate(Window = as.numeric(gsub("Window", "", Window))) %>%
  mutate(across(starts_with('Sample'), ~ . / ((Sample18 + Sample20) / 2)) ) %>%
  sample_frac(0.1) %>%
  pivot_longer(-c(Window, Chr, Pos)) %>%
  mutate(name2 = colData(deseq_met)[.$name, ]$Condition) -> normalized_data_met

library(EnsDb.Mmusculus.v79)
genes <- genes(EnsDb.Mmusculus.v79)
genes <- genes[genes$gene_biotype %in% c("protein_coding", "lincRNA"), ]
# seqlevelsStyle(genes) <- "UCSC"

# annotate windows wrt whether they overlap any genes
normalized_data_granges <- GRanges(gsub("chr", "", normalized_data_met$Chr), normalized_data_met$Pos)
normalized_data_met$distance <- distanceToNearest(normalized_data_granges, genes)@elementMetadata$distance

# first investigate the autosomal signal
normalized_data_met %>%
  mutate(name2 = gsub("r2|r3", "", name2)) %>%
  dplyr::select(c(Pos, name2, Chr, value, distance)) %>%
  group_by(Pos, name2, Chr, distance) %>%
  summarize(value = mean(value)) %>%
  mutate(name2 = factor(condition_conversion[name2], levels = rev(c("Control", "7d Dox", "7d Dox (Washout)", "21d Dox", "21d Dox (Washout)")))) %>%
  dplyr::filter(name2 != "Control") %>% 
  dplyr::filter(Chr != "chrX") -> processed_data

# there is slightly more enrichment in regions not overlapping genes 
processed_data %>% 
  mutate(distance = log10(distance + 1)) %>%
  mutate(distance = cut(distance, breaks = c(-1, 2, 4, 10))) %>%
  mutate(value = ifelse(value > 5, 5, value)) %>%
  ggplot(aes(distance, y = value)) + geom_violin() + facet_wrap(~name2) + geom_hline(yintercept = 1) + coord_flip() + 
    stat_summary() + theme_paper()
ggsave("./Plots/total_analysis/autosomal_global_met.pdf", width = 20, height = 12)

# calculate number of windows with substantial gain (at least 50% more than baseline)
processed_data %>% 
  mutate(distance = log10(distance + 1)) %>%
  mutate(overlap = ifelse(distance > 0, "no overlap", "overlap")) %>%
  mutate(sig_gain = value > 1.5) %>%
  ggplot(aes(x = overlap, fill = sig_gain)) + geom_bar(position = "fill") + facet_wrap(~name2, nrow = 1) + 
    theme_paper() + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggsave("./Plots/total_analysis/autosomal_global_met2.pdf", width = 20, height = 12)

normalized_data_met %>% 
  mutate(name2 = gsub("r2|r3", "", name2)) %>%
  dplyr::select(c(Pos, name2, Chr, value)) %>%
  group_by(Pos, name2, Chr) %>%
  summarize(value = mean(value)) %>%
  dplyr::filter(Chr == "chrX") %>%
  mutate(name2 = factor(condition_conversion[name2], levels = rev(c("Control", "7d Dox", "7d Dox (Washout)", "21d Dox", "21d Dox (Washout)")))) %>%
  mutate(value = ifelse(value > 20, 20, value)) %>%
  ggplot(aes(x = name2, y = value)) + geom_boxplot(fill = "grey") + 
    theme_paper() + coord_flip() + geom_hline(yintercept = 1, linetype = 'dashed') + 
    ylab("Count Fold Change (to Control)") + xlab("")
ggsave("./Plots/total_analysis/met_levels.pdf", width = 20, height = 12)

# check whether promoters of genes that change in expression change more
genes_here <- genes(EnsDb.Mmusculus.v79) %>% data.frame() %>% dplyr::rename(Gene = "gene_name") %>% dplyr::select(c("seqnames", "start", "end", Gene))

autosomal_des <- readRDS("~/Desktop/PhD/Projects/XChromosome_Antonia/ProcessedData/autosomal_de_analysis_7d_dox.rds") %>%
  right_join(genes_here, by = "Gene") %>% 
  mutate(seqnames = paste0("chr", seqnames)) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = T)

normalized_data_met_granges <- normalized_data_met %>% 
  mutate(seqnames = Chr, start = Pos - 250, stop = Pos + 250) %>%
  makeGRangesFromDataFrame(seqnames.field = "seqnames", start.field = "start", end.field = "stop", keep.extra.columns = T)

overlaps_here <- findOverlaps(normalized_data_met_granges, autosomal_des, minoverlap = 200)

normalized_data_met_granges$overlapping_gene <- NA
normalized_data_met_granges$expression_foldchange <- NA
normalized_data_met_granges[from(overlaps_here), ]$overlapping_gene <- autosomal_des[to(overlaps_here), ]$Gene
normalized_data_met_granges[from(overlaps_here), ]$expression_foldchange <- autosomal_des[to(overlaps_here), ]$log2FoldChange

normalized_data_met_granges %>%
  data.frame() %>% 
  mutate(value = ifelse(value > 20, 20, value)) %>%
  ggplot(aes(x = expression_foldchange, y = value)) + facet_wrap(~name2) + geom_point(size = .1) + theme_paper()
ggsave("./Plots/total_analysis/autosomal_gains_met.pdf", width = 20, height = 12)

normalized_data_met_granges %>%
  data.frame() %>% 
  mutate(seqnames2 = seqnames != "chrX") %>%
  dplyr::filter(name2 == "21dDr3") %>% {
    ggplot(data = ., aes(x = expression_foldchange, y = value)) + geom_point(size = 0.1)  + 
    ggrepel::geom_text_repel(data = . %>% dplyr::filter(expression_foldchange < -2 | value > 2), aes(label = overlapping_gene))
  }

# normalized_data_met_granges %>%
#   data.frame() %>% 
#   mutate(seqnames2 = seqnames != "chrX") %>%
#   dplyr::filter(name2 == "21dDr3") %>% {
#     ggplot(data = ., aes(x = expression_foldchange, y = value)) + geom_point(size = 0.1)  + 
#     ggrepel::geom_text_repel(data = . %>% dplyr::filter(expression_foldchange < -2 | value > 2), 
#                              aes(label = seqnames))
#   }

saveRDS(deseq_met, "./deseq_met_total.rds")

```

