---
title: "XistPaper_Figure01"
output: html_document
author: "Jasper Panten"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Desktop/PhD/Projects/XChromosome_Antonia/")
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

This script reproduces all plots for Figure 1 and associated supplementary figures in PaperName. Preprocessing of raw RNA sequencing data is done in ./Folder/script, that yields allele-specific expression counts which are combined into a SingleCellExperiment (SCE) object in ./Folder/script.R. [Insert in Figure 4: CnR + methylation data preprocessing]. 
We first load the necessary libraries and some helper functions: 

```{r load_libraries}

setwd("~/Desktop/PhD/Projects/XChromosome_Antonia/")

library(tidyverse)
library(ggplot2)
library(scran)
library(scater)
library(DESeq2)
library(patchwork)

source("./Scripts/auxiliary.R")
# source("./")

# General theme for all plots: 
theme_paper <- function(){
  theme(panel.background = element_blank(),
        panel.grid = element_line(colour= "grey92"), 
        panel.grid.minor = element_line(size = rel(0.5)),
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1), 
        text = element_text(size = 20))
}

counts_active <- function(sce){assays(sce)[["counts_active"]]}
counts_inactive <- function(sce){assays(sce)[["counts_inactive"]]}

escape_colors <- setNames(c("grey", "darkgreen", "orange"), nm = c("silenced / variable", "facultative", "constitutive"))

```

Now, read in the SCE, remove non-X genes and exclude lowly expressed genes

```{r load_data}

data <- readRDS("./ProcessedData/merged_dataset_astrocytes.rds")
data_before_filtering <-  data[seqnames(data) == "X", ]

data <- data[(rowSums(counts_inactive(data)) + rowSums(counts_active(data))) / ncol(data) > 10, ]

data_with_autosomes <- data
data <- data[seqnames(data) == "X", ]
data$Replicates <- str_extract(data$SampleName, "R1|R2")

convert_names <- list(
  "Astro_dox3" = "3d dox", 
  "Astro_dox7" = "7d dox", 
  "Astro_dox3_wo" = "3d dox + washout", 
  "Astro_dox7_wo" = "7d dox + washout", 
  "Astro_no_dox3" = "3d no dox", 
  "Astro_no_dox7" = "7d no dox", 
  "Astro_no_dox3_no_wo" = "3d no dox + washout", 
  "Astro_no_dox7_no_wo" = "7d no dox + washout"
)

data$ConditionClean <- as.character(convert_names[data$ConditionClean])

```

For the supplement, generate plots showing a) how many genes we detect with sufficient allele-specific coverage and b) allelic densities of autosomal and X-linked genes. Finally, we look at escape measured by d-scores (d = inact / (inact + act) = B6 / (B6 / CAST)) in two cell lines. 

```{r supplement_read_statistics }

data_baseline_x <- data[,data$Condition %in% c("Astro_no_dox3", "Astro_no_dox7", "Astro_no_dox3_no_wo", "Astro_no_dox7_no_wo")]

sample_show = 5
colnames(data_before_filtering)[[sample_show]]

data.frame(
  counts_total = counts(data_before_filtering[,sample_show])[[1]], 
  counts_allelic = (counts_active(data_before_filtering[,sample_show]) +  counts_inactive(data_before_filtering[,sample_show]))[[1]]
) %>%
  ggplot(aes(x = counts_total + 1, y = counts_allelic + 1)) + geom_point() + scale_x_log10() + scale_y_log10() + 
  theme_paper() + geom_vline(xintercept = 10, linetype = "dashed") + xlab("Total reads mapped") + ylab("Allele-specific reads mapped") + 
  geom_abline(linetype = "dashed")

data.frame(
  chromosome = as.character(seqnames(rowRanges(data_with_autosomes))) == "X", 
  total = (counts_active(data_with_autosomes[,sample_show]) + counts_inactive(data_with_autosomes[,sample_show]))[[1]], 
  ase =  (counts_active(data_with_autosomes[,sample_show]) / (counts_active(data_with_autosomes[,sample_show]) + counts_inactive(data_with_autosomes[,sample_show])))[[1]]
) %>%
  ggplot(aes(x = ase, fill = chromosome)) + geom_density() + theme_paper() + ylab("Density") + xlab("B6 / (B6 + CAST)")
# ggsave("./FiguresIllustrator/FigS1/density_ase_c30.pdf")

# Show for one E6 example: 
sample_show = 2
colnames(data_before_filtering)[[sample_show]]

data.frame(
  counts_total = counts(data_before_filtering[,sample_show])[[1]], 
  counts_allelic = (counts_active(data_before_filtering[,sample_show]) +  counts_inactive(data_before_filtering[,sample_show]))[[1]]
) %>%
  ggplot(aes(x = counts_total + 1, y = counts_allelic + 1)) + geom_point() + scale_x_log10() + scale_y_log10() + 
  theme_paper() + geom_vline(xintercept = 10, linetype = "dashed") + xlab("Total reads mapped") + ylab("Allele-specific reads mapped") + 
  geom_abline(linetype = "dashed")
# ggsave("./FiguresIllustrator/FigS1/mapped_reads_e6.pdf")

data.frame(
  chromosome = as.character(seqnames(rowRanges(data_with_autosomes))) == "X", 
  total = (counts_active(data_with_autosomes[,sample_show]) + counts_inactive(data_with_autosomes[,sample_show]))[[1]], 
  ase =  (counts_active(data_with_autosomes[,sample_show]) / (counts_active(data_with_autosomes[,sample_show]) + counts_inactive(data_with_autosomes[,sample_show])))[[1]]
) %>%
  ggplot(aes(x = ase, fill = chromosome)) + geom_density() + theme_paper() + ylab("Density") + xlab("B6 / (B6 + CAST)")
# ggsave("./FiguresIllustrator/FigS1/density_ase_e6.pdf")

# Plot escape in individual clones
sample_nr = 1

data.frame(
  total = rowSums(counts_inactive(data_baseline_x[,sample_nr])) + rowSums(counts_active(data_baseline_x[,sample_nr])),
  ase = rowSums(counts_inactive(data_baseline_x[,sample_nr])) / (rowSums(counts_inactive(data_baseline_x[,sample_nr])) +
                                                                   rowSums(counts_active(data_baseline_x[,sample_nr])))
) %>%
  rownames_to_column("Gene") %>%
  ggplot(aes(total, ase)) + geom_point() + ggrepel::geom_text_repel(aes(label = Gene)) +
  theme_paper() + scale_x_log10() + geom_hline(yintercept = 0.1, linetype = "dashed") + ggtitle("Rep1 no dox d3") +
  xlab("Total expression (allelic reads)") + ylab("ASE (d-score)")
ggsave("../Figures/FiguresAstrocytes/no_dox_d3_escape_ma.pdf")

# ggsave("./FiguresIllustrator/FigS1/cl30_escape.pdf", width = 12, height = 8)

sample_nr = 1

data.frame(
  total = rowSums(counts_inactive(data_baseline_x[,sample_nr])) + rowSums(counts_active(data_baseline_x[,sample_nr])),
  ase = rowSums(counts_inactive(data_baseline_x[,sample_nr])) / (rowSums(counts_inactive(data_baseline_x[,sample_nr])) +
                                                                   rowSums(counts_active(data_baseline_x[,sample_nr])))
) %>%
  rownames_to_column("Gene") %>%
  ggplot(aes(total, ase)) + geom_point() + ggrepel::geom_text_repel(aes(label = Gene)) +
  theme_paper() + scale_x_log10() + geom_hline(yintercept = 0.1, linetype = "dashed") + ggtitle("Clone E6") +
  xlab("Total expression (allelic reads)") + ylab("ASE (d-score)")
# ggsave("./FiguresIllustrator/FigS1/e6_escape.pdf", width = 12, height = 8)

```

We now quantify the number of genes which escape within and across clones and cell lines. 

```{r supplement_number_of_escapees}

data_here <- data[,data$Condition %in% c("Astro_no_dox3", "Astro_no_dox7", "Astro_no_dox3_no_wo", "Astro_no_dox7_no_wo")]

# get d-scores across genes + samples
ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))

# Genes with average ASE > 0.7 in the control sample are likely mapping artefacts: 
genes_out <- names(rowMeans(ratios, na.rm = T)[rowMeans(ratios, na.rm = T) > 0.8 & rownames(ratios) != "Xist"])
genes_out_na <- rownames( ratios[!apply(ratios, 1, function(x){!any(is.na(x))}), ] )

genes_out <- c(genes_out, genes_out_na)

ratios <- ratios[!rownames(ratios) %in% genes_out, ]

# colnames(ratios) <- c("CL30", "CL31", "E6 (rep1)", "E6 (rep2)", "E6 (rep3)", "CL31.16", "CL30.7")

genes_show <- c("Xist", "Mecp2", "Kdm6a", "Kdm5c")
genes_of_choice <- setNames(rep("", nrow(ratios)), rownames(ratios))
genes_of_choice[genes_show] <- genes_show

# Plot d-scores for genes after ordering the genes by chromosomale coordinate
ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Sample" = data_here$SampleName) %>%
  pivot_longer(-c(Condition, Sample)) %>%
  # dplyr::filter(Condition == "Control") %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  ggplot(aes(Sample, y = reorder(name, position), fill = value)) + geom_tile(width = 0.9) +
    theme_paper() + theme(axis.text.x=element_text(angle = 90, hjust = 1)) + 
    theme(axis.ticks.y = element_blank()) + xlab("") + ylab("Chromosome position") + 
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_y_discrete(labels = genes_of_choice, position = "right") + ylab("") + labs(fill = "d-score")
# ggsave("../Figures/FiguresAstrocytes/astrocyte_heatmap_controls.pdf", width = 10, height = 10)
# ggsave("./FiguresIllustrator/FigS1/heatmap_ordered.pdf", width = 10, height = 10)

ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Sample" = data_here$SampleName) %>%
  pivot_longer(-c(Condition, Sample)) %>%
  # dplyr::filter(Condition == "Control") %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  ggplot(aes(Sample, y = reorder(name, position), fill = value)) + geom_tile(width = 0.9) +
    theme_paper() + theme(axis.text.x=element_text(angle = 90, hjust = 1)) + 
    theme(axis.ticks.y = element_blank()) + xlab("") + ylab("Chromosome position") + 
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_y_discrete(labels = genes_of_choice, position = "left") + ylab("") + labs(fill = "d-score") + coord_flip()
ggsave("../Figures/FiguresAstrocytes/astrocyte_heatmap_controls.pdf", width = 10, height = 10)
# ggsave("./FiguresIllustrator/FigS1/heatmap_ordered_streched.pdf", width = 30, height = 10)

# Plot d-scores for genes that escape in any sample, excluding genes with > 0.7 d-score
ratios <- ratios[!rownames(ratios) %in% genes_out, ]
ratios <- ratios[order(rowMeans(ratios), decreasing = T), ]

# Now we define the baseline escapees per sample as the genes with d-score > 0.1, this will be used throughout the analysis
escapees <- apply(ratios, 2, function(x){x > 0.1})
escapees_per_clone <- apply(escapees, 2, function(x){rownames(escapees)[x]})
escapees_per_clone <- lapply(escapees_per_clone, function(x){x[x != "Xist"]})
#names(escapees_per_clone) <- c("CL30 - December2021", "CL31 - December2021", "E6 - October2021", "E6 - March2022", "E6 - September2022", 
#                               "CL30 - October2022", "CL31 - October2022")

# How many do we get per clone and do they overlap?
ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Sample" = data_here$SampleName) %>%
  pivot_longer(-c(Condition, Sample)) %>%
  # dplyr::filter(Condition == "Control") %>%
  dplyr::filter(value > 0.1) %>%
  group_by(name) %>%
  mutate(escape_in_clones = n() >= 8) %>% # CAVE - here we need to adjust
  ggplot(aes(x = Sample, fill = escape_in_clones)) + geom_bar() + theme_paper() + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + xlab("") + ylab("Number of genes with d-score > 0.1") + 
    scale_fill_manual(values = c("grey", "darkred"), name = "Escape in all samples")
ggsave("../Figures/FiguresAstrocytes/astrocyte_number_of_escapees_per_clone.pdf", width = 8, height = 8)

ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Sample" = data_here$SampleName) %>%
  pivot_longer(-c(Condition, Sample)) %>%
  # dplyr::filter(Condition == "Control") %>%
  dplyr::filter(value > 0.1) %>%
  group_by(name) %>%
  mutate(escape_in_clones = n() >= 8) %>%
  add_column(escape_group = rowData(data[.$name, ])$EscapeAnnotation) %>%
  dplyr::filter(!is.na(escape_group)) %>%
  ggplot(aes(x = Sample, fill = escape_in_clones)) + geom_bar() + theme_paper() + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + xlab("") + ylab("Number of genes with d-score > 0.1") + 
    scale_fill_manual(values = c("grey", "darkred"), name = "Escape in all clones") + facet_wrap(~escape_group)
ggsave("../Figures/FiguresAstrocytes/astrocyte_number_of_escapees_per_clone_stratified.pdf", width = 8, height = 8)

ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Sample" = data_here$SampleName) %>%
  pivot_longer(-c(Condition, Sample)) %>%
  # dplyr::filter(Condition == "Control") %>%
  dplyr::filter(value > 0.1) %>%
  group_by(name) %>%
  mutate(escape_in_clones = n() >= 8) %>%
  add_column(escape_group = rowData(data[.$name, ])$EscapeAnnotation) %>%
  dplyr::filter(!is.na(escape_group)) %>%
  group_by(name, escape_group, escape_in_clones) %>%
  summarize(mean_escape = mean(value)) %>%
  rename(escape_in_clones = "escape_in_all_samples") %>% 
  write.csv("~/Desktop/escapees_baseline_astrocytes.csv")

```

For the main figure, we now plot escape during Xist-overexpression (Fig1 a-...)

```{r }

# Subset data on relevant samples and for the main figure, only show E6
data_here <- data[ , data$Condition %in% c("Astro_dox3", "Astro_dox7", "Astro_dox3_wo", "Astro_dox7_wo", "Astro_no_dox3", "Astro_no_dox7")]

# data_here <- data_here[,data_here$Clone == "E6"]

# For Xist-expression, we need a scaling factor per sample to account for differences in sequencing depth
size_factors <- colSums(counts(data_with_autosomes)) / colSums(counts(data_with_autosomes))[[1]]

# check that Xist-overexpression works by plotting Xist counts per million (CPM) per sample -- all samples
data.frame(
  Condition = data$ConditionClean,
  Replicate = data$Replicates, 
  # Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data["Xist", ]) / colSums(counts(data_here)) * 1e6)
) %>%
  # add_column(Condition = ifelse(.$Dox == 0, "Control", paste0("Dox (", .$Dox, " days)"))) %>%
  # mutate(Condition = factor(Condition, levels = rev(c("Control", paste0("Dox (", c(3, 7, 14, 21), " days)"))))) %>%
  mutate(Expression = Expression / mean(Expression[Condition == "3d no dox"])) %>%
  ggplot(aes(x = Condition, y = Expression)) + 
    stat_summary(stat = "mean", geom = "bar", fill = "grey", col = "black") + 
    coord_flip() + theme_paper() + 
    xlab("") + ylab("Expression (CPM)") + geom_jitter(size = 3, width = 0.1) + 
    # scale_y_continuous(expand = c(0, 0), limits = c(0, 320000)) + 
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggsave("../Figures/FiguresAstrocytes/astrocyte_xist_overexpression_supplement.pdf", width = 4, height = 4)

# check that Xist-overexpression works by plotting Xist counts per million (CPM) per sample
data.frame(
  Condition = data_here$ConditionClean,
  Replicate = data_here$Replicates, 
  # Clone = paste0(data_here$Clone, " - ", data_here$Experiment), 
  Expression = as.numeric(counts(data_here["Xist", ]) / colSums(counts(data_here)) * 1e6)
) %>%
  # add_column(Condition = ifelse(.$Dox == 0, "Control", paste0("Dox (", .$Dox, " days)"))) %>%
  # mutate(Condition = factor(Condition, levels = rev(c("Control", paste0("Dox (", c(3, 7, 14, 21), " days)"))))) %>%
  mutate(Condition = factor(Condition, levels = c("3d no dox", "3d dox", "3d dox + washout", "7d no dox", "7d dox", "7d dox + washout"))) %>%
  mutate(Expression = Expression / mean(Expression[Condition == "3d no dox"])) %>%
  ggplot(aes(x = Condition, y = Expression)) + 
    stat_summary(stat = "mean", geom = "bar", fill = "grey", col = "black") + 
    coord_flip() + theme_paper() + 
    xlab("") + ylab("Expression (CPM)") + geom_jitter(size = 3, width = 0.1) + 
    # scale_y_continuous(expand = c(0, 0), limits = c(0, 320000)) + 
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggsave("../Figures/FiguresAstrocytes/astrocyte_xist_overexpression.pdf", width = 4, height = 4)
# ggsave("./FiguresIllustrator/Fig1/xist_overexpression.pdf", width = 4, height = 4)


### 
data_here <- data

ratios <- counts_inactive(data_here) / (counts_inactive(data_here) + counts_active(data_here))
ratios <- ratios[!rownames(ratios) %in% genes_out, ]

# For the next heatmap, we want to highlight a few genes
genes_show <- c("Xist", "Mecp2", "Kdm6a", "Kdm5c")
genes_of_choice <- setNames(rep("", nrow(ratios)), rownames(ratios))
genes_of_choice[genes_show] <- genes_show

# Heatmap showing d-scores of all genes across Xist-OE
ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Replicate" = data_here$Replicates) %>%
  pivot_longer(-c(Condition, Replicate)) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  # mutate(Condition = factor(Condition, levels = rev(c("Control", "Dox (3d)", "Dox (7d)", "Dox (14d)", "Dox (21d)")))) %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_y_discrete(labels = genes_of_choice, position = "right") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    scale_x_discrete(labels = genes_of_choice, position = "bottom") + 
    labs(fill = "d-score") + guides(fill = "none")
ggsave("../Figures/FiguresAstrocytes/astrocyte_heatmap_ordered.pdf", width = 6, height = 4)
# ggsave("./FiguresIllustrator/Fig1/heatmap_ordered.pdf", width = 6, height = 4)

# Get escapees that are consistent across E6 samples
consistent_escapees <- Reduce("intersect", escapees_per_clone)

# For the supplement, also plot stratified by replicates: 
ratios %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Replicate" = data_here$Replicates) %>%
  pivot_longer(-c(Condition, Replicate)) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) %>%
  # mutate(Condition = factor(Condition, levels = rev(c("Control", "Dox (3d)", "Dox (7d)", "Dox (14d)", "Dox (21d)")))) %>%
  ggplot(aes(reorder(name, position), y = interaction(Replicate, Condition, sep = " -- "), group = Replicate, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_y_discrete(labels = genes_of_choice, position = "right") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    scale_x_discrete(labels = genes_of_choice, position = "bottom") + 
    labs(fill = "d-score")
ggsave("../Figures/FiguresAstrocytes/astrocyte_heatmap_ordered_by_replicates.pdf", width = 10, height = 10)
# ggsave("./FiguresIllustrator/FigS1/heatmap_ordered_by_clones.pdf", width = 10, height = 10)

# Now, we show a heatmap with only the escapees

# To structure the data, we order the genes using hierarchical clustering
data_for_clustering <- do.call("rbind", lapply(unique(data_here$ConditionClean), function(x){
  if (sum(data_here$ConditionClean == x) == 1){
    ratios[consistent_escapees, data_here$ConditionClean == x]
  } else {
    rowMeans(ratios[consistent_escapees, data_here$ConditionClean == x])
  }
}))

clustering_order <- setNames(hclust(dist(t(data_for_clustering)))$order, rownames(ratios[consistent_escapees, ]))

ratios[consistent_escapees, ][clustering_order, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Replicate" = data_here$Replicates) %>%
  pivot_longer(-c(Condition, Replicate)) %>%
  mutate(name = factor(name, levels = unique(.$name))) %>%
  add_column(position = start(rowRanges(data_here[as.character(.$name), ]))) -> to_plot
  # mutate(Condition = factor(Condition, levels = rev(c("Control", "Dox (3d)", "Dox (7d)", "Dox (14d)", "Dox (21d)")))) 

p1 <- to_plot %>%
  ggplot(aes(name, y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    labs(fill = "d-score") + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 8)) + 
    scale_x_discrete(position = "top") + guides(fill = "none")

# Plot escapee annotation at the bottom of this: 
p2 <- data.frame(Gene = factor(consistent_escapees, levels = levels(to_plot$name)),  Annotation = rowData(data)[consistent_escapees, ]$EscapeAnnotation) %>%
  ggplot(aes(x = Gene, y = 1, fill = Annotation)) + theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) + 
  theme(legend.position = "None")

p1 / p2 + plot_layout(heights = c(4, 0.2))
# ggsave("./FiguresIllustrator/Fig1/heatmap_ordered_only_escapees.pdf", width = 20, height = 4)

# Also do this reordering
p1 <- to_plot %>%
  ggplot(aes(reorder(name, position), y = Condition, fill = value, col = value)) + geom_tile(width = 0.9) +
    scale_fill_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    scale_colour_gradient2(midpoint = 0.5, mid = "#117733", low = "#BEBEBE", high = "black", limits = c(0, 1), na.value = "white") + 
    theme_paper() + 
    theme(axis.ticks.x = element_blank(), panel.grid.major = element_blank()) + 
    xlab("") + 
    ylab("") + theme(panel.border = element_blank(), axis.line=element_blank()) + 
    labs(fill = "d-score") + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0, size = 8)) + 
    scale_x_discrete(position = "top") + guides(fill = "none")

# Plot escapee annotation at the bottom of this: 
p2 <- data.frame(Gene = factor(consistent_escapees, levels = levels(to_plot$name)),  Annotation = rowData(data)[consistent_escapees, ]$EscapeAnnotation) %>%
  ggplot(aes(x = Gene, y = 1, fill = Annotation)) + theme_void() + geom_tile(col = "black") + scale_fill_manual(values = escape_colors) + 
  theme(legend.position = "None")

p1 / p2 + plot_layout(heights = c(4, 0.2))
ggsave("../Figures/FiguresAstrocytes/astrocyte_heatmap_ordered_only_escapees_by_position.pdf", width = 20, height = 6)

# ggsave("./FiguresIllustrator/Fig1/heatmap_ordered_only_escapees_by_position.pdf", width = 20, height = 4)

# Quantify for genes that escape in E6 in the Control sample
ratios[consistent_escapees, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$ConditionClean) %>%
  add_column("Replicate" = data_here$Replicates) %>%
  pivot_longer(-c(Condition, Replicate)) %>%
  group_by(Replicate) %>%
  # add_column(Condition = ifelse(.$Dox == 0, "Control", paste0("Dox (", .$Dox, " days)"))) %>%
  # mutate(Condition = factor(Condition, levels = (c("Control", paste0("Dox (", c(3, 7, 14, 21), " days)"))))) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, c("silenced / variable", "facultative", "constitutive"))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  ggplot(aes(x = Condition, y = value, fill = escape_status)) + 
    geom_boxplot(col = "black", outlier.color = "grey") + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = c("Silenced / Variable", "Facultative", "Constitutive")) + 
    labs(fill = "Escape Category") + xlab("")
ggsave("../Figures/FiguresAstrocytes/astrocyte_escapee_silencing_boxplots.pdf", width = 9, height = 6)

# ggsave("./FiguresIllustrator/Fig1/escapee_silencing_boxplots.pdf", width = 9, height = 6)

ratios[consistent_escapees, ] %>%
  t() %>% data.frame(check.names = F) %>%
  add_column("Condition" = data_here$Condition) %>%
  add_column("Replicate" = data_here$Replicates) %>%
  pivot_longer(-c(Condition, Replicate)) %>%
  group_by(Replicate) %>%
  # add_column(Condition = ifelse(.$Dox == 0, "Control", paste0("Dox (", .$Dox, " days)"))) %>%
  # mutate(Condition = factor(Condition, levels = (c("Control", paste0("Dox (", c(3, 7, 14, 21), " days)"))))) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, c("silenced / variable", "facultative", "constitutive"))) %>%
  dplyr::filter(!is.na(escape_status)) %>%
  ggplot(aes(x = Condition, y = value, fill = Replicate)) + 
    geom_boxplot(col = "black", outlier.color = "grey", position = position_dodge2(preserve = "single")) + 
    theme_paper() + ylab("ASE (d-score)") + 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    labs(fill = "Escape Category") + xlab("")
ggsave("../Figures/FiguresAstrocytes/astrocyte_escapee_silencing_boxplots_per_clone.pdf", width = 9, height = 6)

# ggsave("./FiguresIllustrator/FigS1/escapee_silencing_boxplots_per_clone.pdf", width = 9, height = 6)

```

Compare escape and changes in escape to the results from npcs: 

```{r asd}

colData(data) <- colData(data)[ , c("Condition", "Clone")]

data_npc <- readRDS("./ProcessedData/merged_dataset.rds")
data_npc <- data_npc[seqnames(data_npc) == "X", ]
data_npc <- data_npc[rownames(data), ]
data_npc <- data_npc[, data_npc$Clone == "E6"]
data_npc <- data_npc[, data_npc$Condition %in% c("Aux_0_Dox_0_WO_0_WOAuxNO", "Aux_0_Dox_3_WO_0_WOAuxNO", "Aux_0_Dox_3_WO_4_WOAuxNO", 
                                                 "Aux_0_Dox_7_WO_0_WOAuxNO", "Aux_0_Dox_7_WO_4_WOAuxNO")]

mcols(data_npc) <- mcols(data)
data_npc$Condition <- list(
  "Aux_0_Dox_0_WO_0_WOAuxNO" = "Astro_no_dox7", 
  "Aux_0_Dox_3_WO_0_WOAuxNO" = "Astro_dox3", 
  "Aux_0_Dox_3_WO_4_WOAuxNO" = "Astro_dox3_wo", 
  "Aux_0_Dox_7_WO_0_WOAuxNO" = "Astro_dox7", 
  "Aux_0_Dox_7_WO_4_WOAuxNO" = "Astro_dox7_wo",
  "Aux_0_Dox_7_WO_7_WOAuxNO" = "Astro_dox7_wo",
  "Aux_0_Dox_7_WO_14_WOAuxNO" = "Astro_dox7_wo"
)[data_npc$Condition] %>% unlist()
colData(data_npc) <- colData(data_npc)[ , c("Condition", "Clone")]

data$CellType = "Astrocyte"
data_npc$CellType = "NPC"

#
read_csv("./ProcessedData/d_score_differences.csv") %>% pull(name) -> esacpe_set

data_merged <- cbind(data, data_npc)

ratios <- counts_inactive(data_merged) / (counts_inactive(data_merged) + counts_active(data_merged))

ratios %>%
  t() %>% data.frame() %>%
  add_column("Condition" = data_merged$Condition) %>%
  add_column("CellType" = data_merged$CellType) %>%
  pivot_longer(-c(Condition, CellType)) %>% 
  group_by(Condition, CellType, name) %>%
  summarize(mean = mean(value)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, c("silenced / variable", "facultative", "constitutive"))) %>%
  dplyr::filter(Condition == "Astro_no_dox7") %>%
  pivot_wider(names_from = CellType, values_from = mean) %>%
  dplyr::filter(name != "Xist") %>%
  dplyr::filter(!is.na(escape_status)) -> plot_here

table(plot_here$Astrocyte > .1, plot_here$NPC > .1) # 1 - ((9 + 12) / (9 + 12 + 226 + 139)) > 94% the same
table(plot_here$Astrocyte > .1, plot_here$NPC > .1) %>% fisher.test()

plot_here %>% {
    ggplot(data = ., aes(x = NPC, y = Astrocyte, fill = escape_status)) + 
      geom_point(pch = 21, size = 4) +
      theme_paper() + ylab("ASE (d-score)") + 
      scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = c("Silenced / Variable", "Facultative", "Constitutive")) + 
      labs(fill = "Escape Category") + xlab("NPC") + ylab("Astrocyte") + geom_abline() + 
      xlim(c(0, 1)) + ylim(c(0, 1)) + 
      # ggrepel::geom_text_repel(aes(label = name), data = . %>% dplyr::filter(abs(NPC - Astrocyte) > 0.15)) + 
      annotate(x = 0.8, y = 0.2, label = paste0("Pearson correlation:", round(cor(.$Astrocyte, .$NPC, use = "complete.obs"), digits = 4)), geom = "text") + 
      annotate(x = 0.8, y = 0.15, label = paste0("Genes matching: 94%"), geom = "text")
  }
ggsave("../Figures/FiguresAstrocytes/astro_npc_compare_baseline_escape.pdf", width = 9, height = 6)

ratios[esacpe_set, ] %>%
  t() %>% data.frame() %>%
  add_column("Condition" = data_merged$Condition) %>%
  add_column("CellType" = data_merged$CellType) %>%
  pivot_longer(-c(Condition, CellType)) %>% 
  group_by(Condition, CellType, name) %>%
  summarize(mean = mean(value)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, c("silenced / variable", "facultative", "constitutive"))) %>%
  dplyr::filter(Condition == "Astro_dox3") %>%
  pivot_wider(names_from = CellType, values_from = mean) %>%
  dplyr::filter(!is.na(escape_status)) %>% {
    ggplot(data = ., aes(x = NPC, y = Astrocyte, fill = escape_status)) + 
      geom_point(pch = 21, size = 4) +
      theme_paper() + ylab("ASE (d-score)") + 
      scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = c("Silenced / Variable", "Facultative", "Constitutive")) + 
      labs(fill = "Escape Category") + xlab("NPC") + ylab("Astrocyte") + geom_abline() + 
      xlim(c(0, .8)) + ylim(c(0, .8))
      # ggrepel::geom_text_repel(aes(label = name), data = . %>% dplyr::filter(abs(NPC - Astrocyte) > 0.15)) + 
      #annotate(x = 0.8, y = 0.2, label = paste0("Pearson correlation: ", round(cor(.$Astrocyte, .$NPC, use = "complete.obs"), digits = 4)), geom = "text")
  }
ggsave("../Figures/FiguresAstrocytes/astro_npc_compare_3d_escape.pdf", width = 9, height = 6)

ratios[esacpe_set, ] %>%
  t() %>% data.frame() %>%
  add_column("Condition" = data_merged$Condition) %>%
  add_column("CellType" = data_merged$CellType) %>%
  pivot_longer(-c(Condition, CellType)) %>% 
  group_by(Condition, CellType, name) %>%
  summarize(mean = mean(value)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, c("silenced / variable", "facultative", "constitutive"))) %>%
  dplyr::filter(Condition == "Astro_dox7") %>%
  pivot_wider(names_from = CellType, values_from = mean) %>%
  dplyr::filter(!is.na(escape_status)) %>% {
    ggplot(data = ., aes(x = NPC, y = Astrocyte, fill = escape_status)) + 
      geom_point(pch = 21, size = 4) +
      theme_paper() + ylab("ASE (d-score)") + 
      scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = c("Silenced / Variable", "Facultative", "Constitutive")) + 
      labs(fill = "Escape Category") + xlab("NPC") + ylab("Astrocyte") + geom_abline() + 
      xlim(c(0, .8)) + ylim(c(0, .8))
      # ggrepel::geom_text_repel(aes(label = name), data = . %>% dplyr::filter(abs(NPC - Astrocyte) > 0.15)) + 
  }
ggsave("../Figures/FiguresAstrocytes/astro_npc_compare_7d_escape.pdf", width = 9, height = 6)

ratios[esacpe_set, ] %>%
  t() %>% data.frame() %>%
  add_column("Condition" = data_merged$Condition) %>%
  add_column("CellType" = data_merged$CellType) %>%
  pivot_longer(-c(Condition, CellType)) %>% 
  group_by(Condition, CellType, name) %>%
  summarize(mean = mean(value)) %>%
  add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
  mutate(escape_status = factor(escape_status, c("silenced / variable", "facultative", "constitutive"))) %>%
  dplyr::filter(Condition == "Astro_dox7_wo") %>%
  pivot_wider(names_from = CellType, values_from = mean) %>%
  dplyr::filter(!is.na(escape_status)) %>% {
    ggplot(data = ., aes(x = NPC, y = Astrocyte, fill = escape_status)) + 
      geom_point(pch = 21, size = 4) +
      theme_paper() + ylab("ASE (d-score)") + 
      scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = c("Silenced / Variable", "Facultative", "Constitutive")) + 
      labs(fill = "Escape Category") + xlab("NPC") + ylab("Astrocyte") + geom_abline() + 
      xlim(c(0, .8)) + ylim(c(0, .8))
      # ggrepel::geom_text_repel(aes(label = name), data = . %>% dplyr::filter(abs(NPC - Astrocyte) > 0.15)) + 
  }
ggsave("../Figures/FiguresAstrocytes/astro_npc_compare_7d_washout.pdf", width = 9, height = 6)

# ratios[consistent_escapees, ] %>%
#   t() %>% data.frame() %>%
#   add_column("Condition" = data_merged$Condition) %>%
#   add_column("CellType" = data_merged$CellType) %>%
#   pivot_longer(-c(Condition, CellType)) %>% 
#   group_by(Condition, CellType, name) %>%
#   summarize(mean = mean(value)) %>%
#   add_column(escape_status = unlist(rowData(data_here)[match(.$name, rownames(data_here)), ]$EscapeAnnotation)) %>%
#   mutate(escape_status = factor(escape_status, c("silenced / variable", "facultative", "constitutive"))) %>%
#   dplyr::filter(!is.na(escape_status)) %>%
#   dplyr::filter(grepl("dox7", Condition)) %>%
#   group_by(CellType, name) %>%
#   mutate(mean_baseline = mean[Condition == "Astro_no_dox7"]) %>%
#   mutate(mean_corrected = mean - mean_baseline) %>%
#   dplyr::filter(Condition != "Astro_no_dox7") %>% 
#   select(-c(mean, mean_baseline)) %>%
#   pivot_wider(names_from = CellType, values_from = mean_corrected) %>%
#   ggplot(aes(x = NPC, y = Astrocyte, fill = escape_status)) + 
#     geom_point(pch = 21, size = 4) + facet_wrap(~Condition) + 
#     theme_paper() + ylab("ASE (d-score)") + 
#     scale_fill_manual(values = c("grey", "darkgreen", "orange"), labels = c("Silenced / Variable", "Facultative", "Constitutive")) + 
#     labs(fill = "Escape Category") + xlab("") + coord_fixed() + geom_abline() + 
#     xlab("")

```

To make sure we observe substantial reduction in gene expression, we use DESeq2 to test for differences in escapee expression between Control and 7d. 
We do this either for combined, only inactive and only active counts: 

```{r }

data_here <- data[,data$Condition %in% c("Astro_no_dox7", "Astro_dox7")]
data_here$Replicates <- str_extract(colnames(data_here), "R[0-9]")
# data_here <- data_here[,data_here$Clone == "E6"]
data_here <- data_here[unique(do.call("c", escapees_per_clone)), ]
data_here <- data_here[!rownames(data_here) %in% genes_out, ]
data_here$ConditionClean <- relevel(factor(data_here$Condition), ref = "Astro_no_dox7")

data_for_sf <- data_with_autosomes[,colnames(data_here)]

size_factors_global <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts(data_for_sf), colData = DataFrame(Cond = rep(1, 4)), design = ~0))
size_factors_x <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts_active(data_here) + counts_inactive(data_here), colData = DataFrame(Cond = rep(1, 4)), design = ~0))

deseq_inactive <- DESeqDataSetFromMatrix(countData = counts_inactive(data_here), colData = colData(data_here), design = ~ Replicates + Condition)
deseq_inactive$sizeFactor <- size_factors_global$sizeFactor
deseq_inactive <- DESeq(deseq_inactive)

deseq_active <- DESeqDataSetFromMatrix(countData = counts_active(data_here), colData = colData(data_here), design = ~ Replicates + Condition)
deseq_active$sizeFactor <- size_factors_global$sizeFactor
deseq_active <- DESeq(deseq_active)

deseq_total <- DESeqDataSetFromMatrix(countData = counts(data_here), colData = colData(data_here), design = ~ Replicates + ConditionClean)
deseq_total$sizeFactor <- size_factors_global$sizeFactor
deseq_total <- DESeq(deseq_total)

df_inactive <- results(deseq_inactive) %>%
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Status = "Inactive")

df_active <- results(deseq_active) %>%
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Status = "Active")

df_total <- results(deseq_total) %>%
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Status=  "Total")

rbind(df_inactive, rbind(df_active, df_total)) %>%
  ggplot(aes(x = baseMean, y = log2FoldChange, col = padj < 0.1)) + geom_point() + scale_x_log10() + 
    ggrepel::geom_text_repel(aes(label = Gene)) + xlab("Mean Expression (log normalized counts)") + ylab("log2 (7d dox / Control)") + 
    theme_paper() + facet_wrap(~Status) + scale_color_manual(values = c("black", "red"))
ggsave("../Figures/FiguresAstrocytes/astrocyte_escape_deseq_analysis.pdf", width = 9, height = 6)

# Visualize also as a heatmap: 

df_combined <- rbind(df_inactive, rbind(df_active, df_total))
df_combined$genomic_position <- start(rowRanges(data[df_combined$Gene, ]))

df_combined %>% 
  ggplot(aes(x = 1, y = reorder(Gene, genomic_position), fill = log2FoldChange)) + geom_tile(col = "black") + facet_wrap(~Status) + 
    theme_paper() + scale_fill_gradient2() + geom_text(aes(label = ifelse(padj < 0.1, ".", "")), size = 10, nudge_y = 0.7) + 
    theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
ggsave("../Figures/FiguresAstrocytes/astrocyte_escape_deseq_analysis_heatmap.pdf", width = 9, height = 30)

```

```{r }

data_here <- data_with_autosomes[,data_with_autosomes$Condition %in% c("Astro_no_dox3", "Astro_dox3")]
data_here <- data_here[!rownames(data_here) %in% genes_out, ]

data_for_sf <- data_with_autosomes[,colnames(data_here)]

size_factors_global <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts(data_for_sf), colData = DataFrame(Cond = rep(1, 4)), design = ~0))
size_factors_x <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts_active(data_here) + counts_inactive(data_here), colData = DataFrame(Cond = rep(1, 4)), design = ~0))

deseq_total <- DESeqDataSetFromMatrix(countData = counts(data_here), colData = colData(data_here), design = ~ Condition)
deseq_total$sizeFactor <- size_factors_global$sizeFactor
deseq_total <- DESeq(deseq_total)

df_total <- results(deseq_total) %>%
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Status=  "Total")

df_total %>%
  mutate(chr = as.character(seqnames(rowRanges(data_with_autosomes)[.$Gene]))) %>%
  dplyr::filter(chr != "X") %>%
  ggplot(aes(x = baseMean, y = log2FoldChange, col = padj < 0.1)) + geom_point(size = .2) + scale_x_log10() + 
    xlab("Mean Expression (log normalized counts)") + ylab("log2 (7d dox / Control)") + 
    theme_paper() + scale_color_manual(values = c("black", "red"))
ggsave("./Figures/FiguresAstrocytes/astrocyte_autosomal_deseq_analysis_3d.pdf", width = 9, height = 6)

results(deseq_total) %>% data.frame() %>% rownames_to_column("Gene") %>%
  mutate(chr = as.character(seqnames(rowRanges(data_with_autosomes)[.$Gene]))) %>%
  arrange(padj) %>% 
  dplyr::filter(chr != "X") %>%
write.csv("./Figures/FiguresAstrocytes/results/autosomal_degs_3d_astrocytes.csv")

data_here <- data_with_autosomes[,data_with_autosomes$Condition %in% c("Astro_no_dox7", "Astro_dox7")]
data_here <- data_here[!rownames(data_here) %in% genes_out, ]

data_for_sf <- data_with_autosomes[,colnames(data_here)]

size_factors_global <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts(data_for_sf), colData = DataFrame(Cond = rep(1, 4)), design = ~0))
size_factors_x <- DESeq2::estimateSizeFactors(DESeqDataSetFromMatrix(counts_active(data_here) + counts_inactive(data_here), colData = DataFrame(Cond = rep(1, 4)), design = ~0))

deseq_total <- DESeqDataSetFromMatrix(countData = counts(data_here), colData = colData(data_here), design = ~ Condition)
deseq_total$sizeFactor <- size_factors_global$sizeFactor
deseq_total <- DESeq(deseq_total)

df_total <- results(deseq_total) %>%
  data.frame() %>%
  rownames_to_column("Gene") %>%
  add_column(Status=  "Total")

df_total %>%
  mutate(chr = as.character(seqnames(rowRanges(data_with_autosomes)[.$Gene]))) %>%
  dplyr::filter(chr != "X") %>%
  ggplot(aes(x = baseMean, y = log2FoldChange, col = padj < 0.1)) + geom_point(size = .2) + scale_x_log10() + 
    xlab("Mean Expression (log normalized counts)") + ylab("log2 (7d dox / Control)") + 
    theme_paper() + scale_color_manual(values = c("black", "red"))
ggsave("./Figures/FiguresAstrocytes/astrocyte_autosomal_deseq_analysis_7d.pdf", width = 9, height = 6)

saveRDS(df_plot, "./ProcessedData/autosomal_de_analysis_astrocytes_7d_dox.rds")

# ggsave("./FiguresIllustrator/FigS1/escape_deseq_analysis.pdf", width = 9, height = 6)

results(deseq_total) %>% data.frame() %>% rownames_to_column("Gene") %>%
  mutate(chr = as.character(seqnames(rowRanges(data_with_autosomes)[.$Gene]))) %>%
  arrange(padj) %>% 
  dplyr::filter(chr != "X") %>%
write.csv("./Figures/FiguresAstrocytes/results/autosomal_degs_7d_astrocytes.csv")

results(deseq_total) %>% data.frame() %>% rownames_to_column("Gene") %>%
  mutate(chr = as.character(seqnames(rowRanges(data_with_autosomes)[.$Gene]))) %>%
  arrange(padj) %>% 
  dplyr::filter(chr != "X")

results(deseq_total) %>% data.frame() %>% rownames_to_column("Gene") %>%
  mutate(chr = as.character(seqnames(rowRanges(data_with_autosomes)[.$Gene]))) %>% 
  mutate(start = as.numeric(start(rowRanges(data_with_autosomes)[.$Gene]))) %>%
  mutate(stop = as.numeric(end(rowRanges(data_with_autosomes)[.$Gene]))) -> res_test

res_test %>% 
  dplyr::filter(!is.na(padj)) %>%
  ggplot(aes(x = start, y = -log10(padj))) + geom_point() + facet_wrap(~chr)

calculate_gene_distances <- function(data, max_dist = 1e5){
  # calculate how often genes have another gene close by (atmost max_dist away)
  data %>% 
    group_by(chr) %>%
    mutate(
      close_gene = unlist(lapply(start, function(x){
        ifelse(any(start - x < max_dist & start - x > 0), "close", "far")
      }))
    ) %>%
    group_by(close_gene, .drop = F) %>% 
    summarize(n = n()) %>%
    pivot_wider(values_from = n, names_from = close_gene) %>%
    mutate(ratio = close / (close + far))
}

set.seed(1233)

lapply(1:10, function(i){
  res_test %>% 
    mutate(padj = sample(padj)) %>%
    dplyr::filter(chr == "X") %>%
    dplyr::filter(padj < 0.1) %>%
    calculate_gene_distances() %>% 
    add_column(it = i)
}) %>% do.call("rbind", .) -> x_ratio_shuff 

lapply(1:10, function(i){
  res_test %>% 
    mutate(padj = sample(padj)) %>%
    dplyr::filter(chr != "X") %>%
    dplyr::filter(padj < 0.1) %>%
    calculate_gene_distances() %>% 
    add_column(it = i)
}) %>% do.call("rbind", .) -> auto_ratio_shuff

res_test %>% 
  dplyr::filter(chr == "X") %>%
  dplyr::filter(padj < 0.1) %>%
  calculate_gene_distances() %>% 
  add_column(it = 1) -> x_ratio

res_test %>% 
  dplyr::filter(chr != "X") %>%
  dplyr::filter(padj < 0.1) %>%
  calculate_gene_distances() %>%
  add_column(it = 1) -> auto_ratio

 list(
    cbind("cond" = "x_ratio_shuff", x_ratio_shuff),
    cbind("cond" = "auto_ratio_shuff", auto_ratio_shuff),
    cbind("cond" = "x_ratio", x_ratio),
    cbind("cond" = "auto_ratio", auto_ratio)
  ) %>% do.call("rbind", .) %>%
  mutate(condition = factor(cond, levels = c("x_ratio", "x_ratio_shuff", "auto_ratio", "auto_ratio_shuff"))) %>%
  group_by(condition) %>%
  summarize(mean = mean(ratio), sd = sd(ratio)) %>% 
  ggplot(aes(x = condition, y = mean)) + geom_bar(stat = "identity") + coord_flip() + geom_errorbar(aes(ymin = mean - sd, ymax = mean + sd), width = .1) + 
    xlab("") + ylab("Fraction of genes with DE neighbour") + theme_paper()
 ggsave("../Figures/FiguresAstrocytes/astrocyte_degs_clustering.pdf", width = 9, height = 6)

```
